// @gibme/starlink/dist/enterprise/index.js
"use strict";
// Copyright (c) 2024, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StarlinkAPI = exports.UserTerminal = exports.ServiceLine = exports.RouterConfig = exports.Router = exports.Account = void 0;
const base_api_1 = __importDefault(require("./api/base_api"));
const account_1 = __importDefault(require("./models/account"));
var account_2 = require("./models/account");
Object.defineProperty(exports, "Account", { enumerable: true, get: function () { return account_2.Account; } });
var router_1 = require("./models/router");
Object.defineProperty(exports, "Router", { enumerable: true, get: function () { return router_1.Router; } });
var router_config_1 = require("./models/router_config");
Object.defineProperty(exports, "RouterConfig", { enumerable: true, get: function () { return router_config_1.RouterConfig; } });
var service_line_1 = require("./models/service_line");
Object.defineProperty(exports, "ServiceLine", { enumerable: true, get: function () { return service_line_1.ServiceLine; } });
var user_terminal_1 = require("./models/user_terminal");
Object.defineProperty(exports, "UserTerminal", { enumerable: true, get: function () { return user_terminal_1.UserTerminal; } });
class StarlinkAPI extends base_api_1.default {
    /**
     * Fetches accounts of the current authenticated user
     *
     * @param options
     */
    async fetch_accounts(options = {
        all: true,
        limit: 50,
        page: 0,
        regionCode: []
    }) {
        var _a, _b, _c;
        (_a = options.limit) !== null && _a !== void 0 ? _a : (options.limit = 50);
        (_b = options.page) !== null && _b !== void 0 ? _b : (options.page = 0);
        (_c = options.all) !== null && _c !== void 0 ? _c : (options.all = true);
        let response;
        if (!options.all) {
            response = await this.get('/enterprise/v1/accounts', options);
            return response.content.results.map(account => new account_1.default(this.client_id, this.client_secret, account));
        }
        else {
            const results = [];
            let page = 0;
            do {
                response = await this.get('/enterprise/v1/accounts', Object.assign(Object.assign({}, options), { limit: 100, page: page++ }));
                results.push(...response.content.results.map(account => new account_1.default(this.client_id, this.client_secret, account)));
            } while (!response.content.isLastPage);
            if (results.length !== response.content.totalCount) {
                throw new Error('Could not fetch all results');
            }
            return results;
        }
    }
}
exports.default = StarlinkAPI;
exports.StarlinkAPI = StarlinkAPI;
