// @gibme/starlink/dist/enterprise/models/account.js
"use strict";
// Copyright (c) 2024, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Account = void 0;
const base_api_1 = __importDefault(require("../api/base_api"));
const router_1 = __importDefault(require("./router"));
const router_config_1 = __importDefault(require("./router_config"));
const service_line_1 = __importDefault(require("./service_line"));
class Account extends base_api_1.default {
    constructor(client_id, client_secret, account) {
        super(client_id, client_secret);
        this.account = account;
    }
    /**
     * Transforms the API response for a Realtime Data Tracking request into an actual response
     *
     * @param record
     * @private
     */
    transformRealtimeDataTracking(record) {
        var _a;
        return Object.assign(Object.assign({}, record), { billingCycles: record.billingCycles !== null
                ? record.billingCycles.map(cycle => {
                    return Object.assign(Object.assign({}, cycle), { dailyDataUsage: cycle.dailyDataUsage !== null
                            ? cycle.dailyDataUsage.map(usage => {
                                return Object.assign(Object.assign({}, usage), { date: new Date(usage.date) });
                            })
                            : null, endDate: new Date(cycle.endDate), startDate: new Date(cycle.startDate) });
                })
                : null, endDate: new Date(record.endDate), lastUpdated: record.lastUpdated !== null ? new Date(record.lastUpdated) : null, servicePlan: Object.assign(Object.assign({}, record.servicePlan), { activeFrom: record.servicePlan.activeFrom !== null
                    ? new Date(record.servicePlan.activeFrom)
                    : null, dataPoolUsage: record.servicePlan.dataPoolUsage
                    ? Object.assign(Object.assign({}, record.servicePlan.dataPoolUsage), { dataBlocks: ((_a = record.servicePlan.dataPoolUsage.dataBlocks) === null || _a === void 0 ? void 0 : _a.map(elem => {
                            return Object.assign(Object.assign({}, elem), { expirationDateUtc: new Date(elem.expirationDateUtc), startDateUtc: new Date(elem.startDateUtc) });
                        })) || [], lastUpdated: new Date(record.servicePlan.dataPoolUsage.lastUpdated) }) : undefined, overageLineDeactivatedDate: record.servicePlan.overageLineDeactivatedDate !== null
                    ? new Date(record.servicePlan.overageLineDeactivatedDate)
                    : null, subscriptionActiveFrom: record.servicePlan.subscriptionActiveFrom !== null
                    ? new Date(record.servicePlan.subscriptionActiveFrom)
                    : null, subscriptionEndDate: record.servicePlan.subscriptionEndDate !== null
                    ? new Date(record.servicePlan.subscriptionEndDate)
                    : null }), startDate: new Date(record.startDate) });
    }
    /**
     * Transforms the API response for a Subscription request into an actual response
     *
     * @param record
     * @private
     */
    transformSubscription(record) {
        return Object.assign(Object.assign({}, record), { endDate: record.endDate !== null ? new Date(record.endDate) : null, normalizedStartDate: record.normalizedStartDate !== null
                ? new Date(record.normalizedStartDate)
                : null, serviceEndDate: record.serviceEndDate !== null ? new Date(record.serviceEndDate) : null, startDate: record.startDate !== null ? new Date(record.startDate) : null });
    }
    /**
     * Adds a user terminal to the account
     *
     * @param deviceId UTID, kit serial number, or dish serial number.
     */
    async add_user_terminal(deviceId) {
        try {
            await this.post(`/enterprise/v1/account/${this.accountNumber}/user-terminals/${deviceId}`);
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    /**
     * Checks the available capacity for the given latitude and longitude
     *
     * @param latitude
     * @param longitude
     */
    async check_capacity(latitude, longitude) {
        const response = await this.post(`/enterprise/v1/account/${this.accountNumber}/addresses/check-capacity`, {
            latitude,
            longitude
        });
        return response.content.availableCapacity;
    }
    /**
     * Create an address within the account
     *
     * @param address
     */
    async create_address(address) {
        const response = await this.post(`/enterprise/v1/account/${this.accountNumber}/addresses`, address);
        return response.content;
    }
    /**
     * Creates a new router configuration on the account
     *
     * @param nickname
     * @param routerConfig
     */
    async create_router_config(nickname, routerConfig) {
        const response = await this.post(`/enterprise/v1/account/${this.accountNumber}/routers/configs`, {
            nickname,
            routerConfigJson: JSON.stringify(routerConfig)
        });
        const _a = response.content, { routerConfigJson } = _a, content = __rest(_a, ["routerConfigJson"]);
        const _routerConfig = JSON.parse(routerConfigJson);
        return new router_config_1.default(this.client_id, this.client_secret, Object.assign(Object.assign({}, content), { routerConfig: _routerConfig }));
    }
    /**
     * Creates a new service line
     *
     * @param addressReferenceId Address Reference ID to associate with the service line.
     * @param productReferenceId Subscription Product ID to associate with the service line.
     */
    async create_service_line(addressReferenceId, productReferenceId) {
        const response = await this.post(`/enterprise/v1/account/${this.accountNumber}/service-lines`, {
            addressReferenceId,
            productReferenceId
        });
        const record = response.content;
        return new service_line_1.default(this.client_id, this.client_secret, Object.assign(Object.assign({}, record), { accountNumber: this.accountNumber, endDate: record.endDate !== null ? new Date(record.endDate) : null, startDate: record.startDate !== null ? new Date(record.startDate) : null }));
    }
    /**
     * Fetches an address by its reference ID
     *
     * @param addressReferenceId
     */
    async fetch_address(addressReferenceId) {
        const response = await this.get(`/enterprise/v1/account/${this.accountNumber}/addresses/${addressReferenceId}`);
        return response.content;
    }
    /**
     * Fetches all the addresses listed on the account
     *
     * @param options
     */
    async fetch_addresses(options = {
        all: true,
        limit: 50,
        page: 0
    }) {
        var _a, _b, _c;
        (_a = options.limit) !== null && _a !== void 0 ? _a : (options.limit = 50);
        (_b = options.page) !== null && _b !== void 0 ? _b : (options.page = 0);
        (_c = options.all) !== null && _c !== void 0 ? _c : (options.all = true);
        let response;
        if (!options.all) {
            response = await this.get(`/enterprise/v1/account/${this.accountNumber}/addresses`, options);
            return response.content.results;
        }
        else {
            const results = [];
            let page = 0;
            do {
                response = await this.get(`/enterprise/v1/account/${this.accountNumber}/addresses`, Object.assign(Object.assign({}, options), { limit: 100, page: page++ }));
                results.push(...response.content.results);
            } while (!response.content.isLastPage);
            if (results.length !== response.content.totalCount) {
                throw new Error('Could not fetch all results');
            }
            return results;
        }
    }
    /**
     * Fetches the available products for the account
     */
    async fetch_available_products(options = { all: true, limit: 500, page: 0 }) {
        var _a, _b, _c;
        (_a = options.limit) !== null && _a !== void 0 ? _a : (options.limit = 500);
        (_b = options.page) !== null && _b !== void 0 ? _b : (options.page = 0);
        (_c = options.all) !== null && _c !== void 0 ? _c : (options.all = true);
        let response;
        if (!options.all) {
            response = await this.get(`/enterprise/v1/account/${this.accountNumber}/service-lines/available-products`, Object.assign(Object.assign({}, options), { pageIndex: options.page, pageLimit: options.limit }));
            return response.content.results;
        }
        else {
            const results = [];
            let pageIndex = 0;
            do {
                response = await this.get(`/enterprise/v1/account/${this.accountNumber}/service-lines/available-products`, Object.assign(Object.assign({}, options), { pageIndex: pageIndex++, pageLimit: 500 }));
                results.push(...response.content.results);
            } while (!response.content.isLastPage);
            if (results.length !== response.content.totalCount) {
                throw new Error('Could not fetch all results');
            }
            return results;
        }
    }
    /**
     * Fetches realtime data tracking information for the service lines under this account
     *
     * @param options
     */
    async fetch_realtime_data_tracking(options = {
        all: true,
        page: 0
    }) {
        var _a, _b;
        (_a = options.page) !== null && _a !== void 0 ? _a : (options.page = 0);
        (_b = options.all) !== null && _b !== void 0 ? _b : (options.all = true);
        let response;
        if (!options.all) {
            response = await this.post(`/enterprise/v1/accounts/${this.accountNumber}/billing-cycles/query`, Object.assign(Object.assign({}, options), { pageIndex: options.page }));
            return response.content.results.map(record => this.transformRealtimeDataTracking(record));
        }
        else {
            const results = [];
            let pageIndex = 0;
            do {
                response = await this.post(`/enterprise/v1/accounts/${this.accountNumber}/billing-cycles/query`, Object.assign(Object.assign({}, options), { pageIndex: pageIndex++ }));
                results.push(...response.content.results.map(record => this.transformRealtimeDataTracking(record)));
            } while (!response.content.isLastPage);
            if (results.length !== response.content.totalCount) {
                throw new Error('Could not fetch all results');
            }
            return results;
        }
    }
    /**
     * Fetches a router
     *
     * @param routerId
     */
    async fetch_router(routerId) {
        const response = await this.get(`/enterprise/v1/account/${this.accountNumber}/routers/${routerId}`);
        return new router_1.default(this.client_id, this.client_secret, response.content);
    }
    /**
     * Fetches the router configuration specified
     *
     * @param configId
     */
    async fetch_router_config(configId) {
        const response = await this.get(`/enterprise/v1/account/${this.accountNumber}/routers/configs/${configId}`);
        const _a = response.content, { routerConfigJson } = _a, content = __rest(_a, ["routerConfigJson"]);
        const routerConfig = JSON.parse(routerConfigJson);
        return new router_config_1.default(this.client_id, this.client_secret, Object.assign(Object.assign({}, content), { routerConfig }));
    }
    /**
     * Fetches all router configs on the account
     *
     * @param options
     */
    async fetch_router_configs(options = {
        all: true,
        page: 0
    }) {
        var _a, _b;
        (_a = options.page) !== null && _a !== void 0 ? _a : (options.page = 0);
        (_b = options.all) !== null && _b !== void 0 ? _b : (options.all = true);
        let response;
        if (!options.all) {
            response = await this.get(`/enterprise/v1/account/${this.accountNumber}/routers/configs`, options);
            return response.content.results.map(record => {
                const { routerConfigJson } = record, rest = __rest(record, ["routerConfigJson"]);
                const routerConfig = JSON.parse(routerConfigJson);
                return new router_config_1.default(this.client_id, this.client_secret, Object.assign(Object.assign({}, rest), { routerConfig }));
            });
        }
        else {
            const results = [];
            let page = 0;
            do {
                response = await this.get(`/enterprise/v1/account/${this.accountNumber}/routers/configs`, { page: page++ });
                results.push(...response.content.results.map(record => {
                    const { routerConfigJson } = record, rest = __rest(record, ["routerConfigJson"]);
                    const routerConfig = JSON.parse(routerConfigJson);
                    return new router_config_1.default(this.client_id, this.client_secret, Object.assign(Object.assign({}, rest), { routerConfig }));
                }));
            } while (!response.content.isLastPage);
            if (results.length !== response.content.totalCount) {
                throw new Error('Could not fetch all results');
            }
            return results;
        }
    }
    /**
     * Fetches a service line in the account
     *
     * @param serviceLineNumber The Service Line Number for the service line
     */
    async fetch_service_line(serviceLineNumber) {
        const response = await this.get(`/enterprise/v1/account/${this.accountNumber}/service-lines/${serviceLineNumber}`);
        const record = response.content;
        return new service_line_1.default(this.client_id, this.client_secret, Object.assign(Object.assign({}, record), { accountNumber: this.accountNumber, endDate: record.endDate !== null ? new Date(record.endDate) : null, startDate: record.startDate !== null ? new Date(record.startDate) : null }));
    }
    /**
     * Fetch the services lines in the account
     *
     * @param options
     */
    async fetch_service_lines(options = {
        all: true,
        limit: 50,
        page: 0
    }) {
        var _a, _b, _c;
        (_a = options.limit) !== null && _a !== void 0 ? _a : (options.limit = 50);
        (_b = options.page) !== null && _b !== void 0 ? _b : (options.page = 0);
        (_c = options.all) !== null && _c !== void 0 ? _c : (options.all = true);
        let response;
        if (!options.all) {
            response = await this.get(`/enterprise/v1/account/${this.accountNumber}/service-lines`, options);
            return response.content.results.map(record => new service_line_1.default(this.client_id, this.client_secret, Object.assign(Object.assign({}, record), { accountNumber: this.accountNumber, endDate: record.endDate !== null ? new Date(record.endDate) : null, startDate: record.startDate !== null ? new Date(record.startDate) : null })));
        }
        else {
            const results = [];
            let page = 0;
            do {
                response = await this.get(`/enterprise/v1/account/${this.accountNumber}/service-lines`, Object.assign(Object.assign({}, options), { limit: 100, page: page++ }));
                results.push(...response.content.results.map(record => new service_line_1.default(this.client_id, this.client_secret, Object.assign(Object.assign({}, record), { accountNumber: this.accountNumber, endDate: record.endDate !== null ? new Date(record.endDate) : null, startDate: record.startDate !== null ? new Date(record.startDate) : null }))));
            } while (!response.content.isLastPage);
            if (results.length !== response.content.totalCount) {
                throw new Error('Could not fetch all results');
            }
            return results;
        }
    }
    /**
     * Fetches the user terminals for an account
     *
     * @param options
     */
    async fetch_user_terminals(options = {
        all: true,
        limit: 50,
        page: 0
    }) {
        var _a, _b, _c;
        (_a = options.limit) !== null && _a !== void 0 ? _a : (options.limit = 50);
        (_b = options.page) !== null && _b !== void 0 ? _b : (options.page = 0);
        (_c = options.all) !== null && _c !== void 0 ? _c : (options.all = true);
        let response;
        if (!options.all) {
            response = await this.get(`/enterprise/v1/account/${this.accountNumber}/user-terminals`, options);
            return response.content.results.map(record => {
                return Object.assign(Object.assign({}, record), { accountNumber: this.accountNumber });
            });
        }
        else {
            const results = [];
            let page = 0;
            do {
                response = await this.get(`/enterprise/v1/account/${this.accountNumber}/user-terminals`, Object.assign(Object.assign({}, options), { limit: 100, page: page++ }));
                results.push(...response.content.results.map(record => {
                    return Object.assign(Object.assign({}, record), { accountNumber: this.accountNumber });
                }));
            } while (!response.content.isLastPage);
            if (results.length !== response.content.totalCount) {
                throw new Error('Could not fetch all results');
            }
            return results;
        }
    }
    /**
     * Remove a service line. This is equivalent to cancelling service for the address.
     *
     * @param serviceLineNumber
     * @param options
     */
    async remove_service_line(serviceLineNumber, options = {}) {
        try {
            await this.delete(`/enterprise/v1/account/${this.accountNumber}/service-lines/${serviceLineNumber}`, options);
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    /**
     * Removes a cold user terminal from the account
     *
     * @param deviceId UTID, kit serial number, or dish serial number.
     */
    async remove_user_terminal(deviceId) {
        try {
            await this.delete(`/enterprise/v1/account/${this.accountNumber}/user-terminals/${deviceId}`);
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    /**
     * The Telemetry API is designed for users that have their own data infrastructure to monitor
     * Starlink devices remotely.
     *
     * This feature enables the ability to perform analysis in near real-time to characterize Starlink device
     * performance relative to ongoing operations.
     *
     * The data in the stream has a time-based retention policy. This is configured to be a fraction of a day (8 hours).
     * The implication of this is that if your consuming process goes offline and does not resume for a period of time
     * longer than this retention policy then you will observe loss of data.
     *
     * Even though this is a JSON HTTP API, it behaves like a data stream. It is important to understand that a
     * consumer's position is tracked and advanced through each successful response transmission. If the consumer
     * process crashes in the middle of processing a response batch, unprocessed entries will not be present in the
     * subsequent request!
     *
     * Each service account advances through the data stream independently. Creating multiple service accounts is a
     * convenient way to consume the data multiple times if, for example, you have multiple environments
     * (e.g. development and production).
     *
     * @param batchSize The maximum number of entries to return in the response (default 100).
     * @param maxLingerMs The maximum number of milliseconds to collect entries for (default 100).
     */
    async telemetry(batchSize = 100, maxLingerMs = 100) {
        const response = await this.post('/telemetry/stream/v1/telemetry', {
            accountNumber: this.accountNumber,
            batchSize,
            maxLingerMs
        });
        const results = {
            Router: [],
            UserTerminal: [],
            UserTerminalDataUsage: [],
            IpAllocs: []
        };
        for (const key of Object.keys(response.metadata.enums.DeviceType)) {
            const alertKeys = response.metadata.enums.AlertsByDeviceType[key] || {};
            const columnKeys = response.data.columnNamesByDeviceType[key] || [];
            const type = response.metadata.enums.DeviceType[key];
            results[type] = [];
            const values = response.data.values.filter(([value]) => value === key);
            for (const value of values) {
                const record = {};
                for (let i = 1; i < columnKeys.length; i++) {
                    const column = columnKeys[i];
                    record[column] = value[i];
                }
                if (record.ActiveAlerts && Array.isArray(record.ActiveAlerts)) {
                    if (record.ActiveAlerts.length !== 0) {
                        const alerts = [];
                        for (const alert of record.ActiveAlerts) {
                            alerts.push(alertKeys[alert.toString()] || alert.toString());
                        }
                        record.ActiveAlerts = alerts;
                    }
                }
                results[type].push(record);
            }
        }
        return results;
    }
    /**
     * Converts the instance to a string (JSON encoded)
     */
    toString() {
        return JSON.stringify(this.account);
    }
    /**
     * Updates an address within the account
     *
     * @param address
     */
    async update_address(address) {
        const response = await this.put(`/enterprise/v1/account/${this.accountNumber}/addresses`, address);
        return response.content;
    }
    /**
     * Updates the account default router configuration to the ID specified
     *
     * @param configId
     */
    async update_default_router_config(configId) {
        const response = await this.put(`/enterprise/v1/accounts/${this.accountNumber}/update-default-router-config`, { configId });
        const account = response.content.results
            .filter(elem => elem.accountNumber === this.accountNumber)
            .shift();
        if (account) {
            this.account = account;
            return true;
        }
        return false;
    }
    /**
     * The name of the account
     */
    get accountName() {
        return this.account.accountName || undefined;
    }
    /**
     * The Account Number. Example: ACC-511274-31364-54
     */
    get accountNumber() {
        return this.account.accountNumber;
    }
    /**
     * Default router config on the account
     */
    get defaultRouterConfigId() {
        return this.account.defaultRouterConfigId || undefined;
    }
    /**
     * The region code of the account. Example: US
     */
    get regionCode() {
        return this.account.regionCode;
    }
}
exports.default = Account;
exports.Account = Account;
