// @gibme/starlink/dist/enterprise/models/router.js
"use strict";
// Copyright (c) 2024, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Router = void 0;
const base_api_1 = __importDefault(require("../api/base_api"));
class Router extends base_api_1.default {
    constructor(client_id, client_secret, router) {
        super(client_id, client_secret);
        this.router = router;
    }
    /**
     * Assign a configuration to the router
     *
     * @param configId
     */
    async assign_config(configId) {
        try {
            await this.put(`/enterprise/v1/account/${this.accountNumber}/routers/${this.routerId}/config`, configId);
            this.router.configId = configId;
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    /**
     * Remove the configuration assignment of the router
     */
    async remove_config() {
        try {
            await this.delete(`/enterprise/v1/account/${this.accountNumber}/routers/${this.routerId}/config`);
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    /**
     * Converts the instance to a string (JSON encoded)
     */
    toString() {
        return JSON.stringify(this.router);
    }
    get accountNumber() {
        return this.router.accountNumber;
    }
    get configId() {
        return this.router.configId;
    }
    get directLinkToDish() {
        return this.router.directLinkToDish;
    }
    get hardwareVersion() {
        return this.router.hardwareVersion;
    }
    get lastBonded() {
        return this.router.lastBonded !== null
            ? new Date(this.router.lastBonded)
            : this.router.lastBonded;
    }
    get nickname() {
        return this.router.nickname;
    }
    get routerId() {
        return this.router.routerId;
    }
    get userTerminalId() {
        return this.router.userTerminalId;
    }
}
exports.default = Router;
exports.Router = Router;
