// @gibme/starlink/dist/enterprise/models/router_config.js
"use strict";
// Copyright (c) 2024, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RouterConfig = void 0;
const base_api_1 = __importDefault(require("../api/base_api"));
class RouterConfig extends base_api_1.default {
    constructor(client_id, client_secret, routerConfig) {
        super(client_id, client_secret);
        this.routerConfig = routerConfig;
    }
    /**
     * Update the router configuration
     *
     * Any router assigned to this configuration will immediately receive the updated if they are online.
     * If it is not online, the router will receive the update when it comes online.
     */
    async save() {
        try {
            const response = await this.put(`/enterprise/v1/account/${this.accountNumber}/routers/configs/${this.configId}`, {
                nickname: this.nickname,
                routerConfigJson: JSON.stringify(this.config)
            });
            const _a = response.content, { routerConfigJson } = _a, content = __rest(_a, ["routerConfigJson"]);
            const routerConfig = JSON.parse(routerConfigJson);
            this.routerConfig = Object.assign(Object.assign({}, content), { routerConfig });
            return true;
        }
        catch (_b) {
            return false;
        }
    }
    /**
     * Converts the instance to a string (JSON encoded)
     */
    toString() {
        return JSON.stringify(this.routerConfig);
    }
    get accountNumber() {
        return this.routerConfig.accountNumber;
    }
    get config() {
        return this.routerConfig.routerConfig;
    }
    set config(value) {
        this.routerConfig.routerConfig = value;
    }
    get configId() {
        return this.routerConfig.configId;
    }
    get nickname() {
        return this.routerConfig.nickname;
    }
    set nickname(value) {
        this.routerConfig.nickname = value;
    }
}
exports.default = RouterConfig;
exports.RouterConfig = RouterConfig;
