// @gibme/starlink/dist/enterprise/models/service_line.js
"use strict";
// Copyright (c) 2024, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceLine = void 0;
const base_api_1 = __importDefault(require("../api/base_api"));
class ServiceLine extends base_api_1.default {
    constructor(client_id, client_secret, serviceLine) {
        super(client_id, client_secret);
        this.serviceLine = serviceLine;
        this.nickname_updated = false;
        this.product_updated = false;
        this.public_ip_updated = false;
    }
    /**
     * Adds the specified terminal to this service line
     *
     * @param terminalOrId
     */
    async add_terminal(terminalOrId) {
        try {
            let url = `/enterprise/v1/account/${this.accountNumber}/user-terminals/`;
            if (typeof terminalOrId === 'string') {
                url += `${terminalOrId}/${this.serviceLineNumber}`;
            }
            else {
                url += `${terminalOrId.userTerminalId}/${this.serviceLineNumber}`;
            }
            await this.post(url);
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    /**
     * Fetches all daily data usage by asset for the last 1 month(s) plus the current cycle
     *
     * @param options
     */
    async fetch_daily_usage(options = {}) {
        var _a, _b;
        const response = await this.get(`/enterprise/v1/account/${this.accountNumber}/service-lines/${this.serviceLineNumber}/billing-cycle/all`, options);
        const record = response.content;
        return Object.assign(Object.assign({}, record), { billingCycles: ((_a = record.billingCycles) === null || _a === void 0 ? void 0 : _a.map(cycle => {
                var _a;
                return Object.assign(Object.assign({}, cycle), { dailyDataUsages: ((_a = cycle.dailyDataUsages) === null || _a === void 0 ? void 0 : _a.map(usage => {
                        return Object.assign(Object.assign({}, usage), { date: new Date(usage.date) });
                    })) || null, endDate: new Date(cycle.endDate), startDate: new Date(cycle.startDate) });
            })) || null, endDate: new Date(record.endDate), lastUpdated: record.lastUpdated !== null ? new Date(record.lastUpdated) : null, servicePlan: Object.assign(Object.assign({}, record.servicePlan), { activeFrom: record.servicePlan.activeFrom !== null ? new Date(record.servicePlan.activeFrom) : null, dataPoolUsage: record.servicePlan.dataPoolUsage
                    ? Object.assign(Object.assign({}, record.servicePlan.dataPoolUsage), { dataBlocks: ((_b = record.servicePlan.dataPoolUsage.dataBlocks) === null || _b === void 0 ? void 0 : _b.map(elem => {
                            return Object.assign(Object.assign({}, elem), { expirationDateUtc: new Date(elem.expirationDateUtc), startDateUtc: new Date(elem.startDateUtc) });
                        })) || [], lastUpdated: new Date(record.servicePlan.dataPoolUsage.lastUpdated) }) : undefined, overageLineDeactivatedDate: record.servicePlan.overageLineDeactivatedDate !== null
                    ? new Date(record.servicePlan.overageLineDeactivatedDate)
                    : null, subscriptionActiveFrom: record.servicePlan.subscriptionActiveFrom !== null
                    ? new Date(record.servicePlan.subscriptionActiveFrom)
                    : null, subscriptionEndDate: record.servicePlan.subscriptionEndDate !== null
                    ? new Date(record.servicePlan.subscriptionEndDate)
                    : null }), startDate: new Date(record.startDate) });
    }
    /**
     * Fetches all historical partial subscription periods by asset for the last 1 month(s) plus the current cycle.
     * Does not include the incomplete period corresponding to the service line's current product
     */
    async fetch_partial_periods() {
        const response = await this.get(`/enterprise/v1/account/${this.accountNumber}/` +
            `service-lines/${this.serviceLineNumber}/billing-cycle/partial-periods`);
        return response.content.map(record => {
            return Object.assign(Object.assign({}, record), { periodEnd: new Date(record.periodEnd), periodStart: new Date(record.periodStart) });
        });
    }
    /**
     * Opt into product for subscription
     */
    async opt_in() {
        const response = await this.post(`/enterprise/v1/account/${this.accountNumber}/service-lines/${this.serviceLineNumber}/opt-in`);
        const record = response.content;
        return Object.assign(Object.assign({}, record), { activatedDate: new Date(record.activatedDate), deactivatedDate: record.deactivatedDate !== null ? new Date(record.deactivatedDate) : null });
    }
    /**
     * Opt out of product for subscription
     */
    async opt_out() {
        const response = await this.delete(`/enterprise/v1/account/${this.accountNumber}/service-lines/${this.serviceLineNumber}/opt-in`);
        const record = response.content;
        return Object.assign(Object.assign({}, record), { activatedDate: new Date(record.activatedDate), deactivatedDate: record.deactivatedDate !== null ? new Date(record.deactivatedDate) : null });
    }
    /**
     * Removes the specified terminal from this service line
     *
     * @param terminalOrId
     */
    async remove_terminal(terminalOrId) {
        try {
            let url = `/enterprise/v1/account/${this.accountNumber}/user-terminals/`;
            if (typeof terminalOrId === 'string') {
                url += `${terminalOrId}/${this.serviceLineNumber}`;
            }
            else {
                url += `${terminalOrId.userTerminalId}/${this.serviceLineNumber}`;
            }
            await this.delete(url);
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    /**
     * Attempts to save any changes to the `nickname`, `productReferenceId`, or `publicIp`
     */
    async save() {
        let response;
        if (this.nickname_updated) {
            try {
                response = await this.put(`/enterprise/v1/account/${this.accountNumber}/service-lines/${this.serviceLineNumber}/nickname`, { nickname: this.nickname });
                this.nickname_updated = false;
            }
            catch (_a) {
                return false;
            }
        }
        if (this.product_updated) {
            try {
                response = await this.put(
                // eslint-disable-next-line max-len
                `/enterprise/v1/account/${this.accountNumber}/service-lines/${this.serviceLineNumber}/product/${this.productReferenceId}`);
                this.product_updated = false;
            }
            catch (_b) {
                return false;
            }
        }
        if (this.public_ip_updated) {
            try {
                response = await this.put(`/enterprise/v1/account/${this.accountNumber}/service-lines/${this.serviceLineNumber}/public-ip`, { publicIp: this.publicIp });
                this.public_ip_updated = false;
            }
            catch (_c) {
                return false;
            }
        }
        if (response) {
            const record = response.content;
            this.serviceLine = Object.assign(Object.assign({}, response.content), { accountNumber: this.accountNumber, endDate: record.endDate !== null ? new Date(record.endDate) : null, startDate: record.startDate !== null ? new Date(record.startDate) : null });
            return true;
        }
        return false;
    }
    /**
     * Converts the instance to a string (JSON encoded)
     */
    toString() {
        return JSON.stringify(this.serviceLine);
    }
    get object() {
        return this.serviceLine;
    }
    get accountNumber() {
        return this.serviceLine.accountNumber;
    }
    get active() {
        return this.serviceLine.active;
    }
    get addressReferenceId() {
        return this.serviceLine.addressReferenceId;
    }
    get aviationMetadata() {
        return this.serviceLine.aviationMetadata;
    }
    get delayedProductId() {
        return this.serviceLine.delayedProductId;
    }
    get endDate() {
        return this.serviceLine.endDate;
    }
    get nickname() {
        return this.serviceLine.nickname;
    }
    set nickname(value) {
        if (value !== this.nickname) {
            this.nickname_updated = true;
        }
        this.serviceLine.nickname = value;
    }
    get optInProductId() {
        return this.serviceLine.optInProductId;
    }
    get productReferenceId() {
        return this.serviceLine.productReferenceId;
    }
    set productReferenceId(value) {
        if (value !== this.productReferenceId) {
            this.product_updated = true;
        }
        this.serviceLine.productReferenceId = value;
    }
    get publicIp() {
        return this.serviceLine.publicIp;
    }
    set publicIp(value) {
        if (value !== this.publicIp) {
            this.public_ip_updated = true;
        }
        this.serviceLine.publicIp = value;
    }
    get serviceLineNumber() {
        return this.serviceLine.serviceLineNumber;
    }
    get startDate() {
        return this.serviceLine.startDate;
    }
}
exports.default = ServiceLine;
exports.ServiceLine = ServiceLine;
