// @gibme/starlink/dist/enterprise/models/user_terminal.js
"use strict";
// Copyright (c) 2024, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserTerminal = void 0;
const base_api_1 = __importDefault(require("../api/base_api"));
const router_1 = __importDefault(require("./router"));
class UserTerminal extends base_api_1.default {
    constructor(client_id, client_secret, userTerminal) {
        super(client_id, client_secret);
        this.userTerminal = userTerminal;
    }
    /**
     * Adds the terminal to the service line specified
     *
     * @param serviceLineOrNumber
     */
    async add_to_service_line(serviceLineOrNumber) {
        try {
            let url = `/enterprise/v1/account/${this.accountNumber}/user-terminals/${this.userTerminalId}/`;
            if (typeof serviceLineOrNumber === 'string') {
                url += serviceLineOrNumber;
            }
            else {
                url += serviceLineOrNumber.serviceLineNumber;
            }
            await this.post(url);
            this.userTerminal.serviceLineNumber = typeof serviceLineOrNumber === 'string'
                ? serviceLineOrNumber
                : serviceLineOrNumber.serviceLineNumber;
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    /**
     * Removes the terminal from it's assigned service line
     */
    async remove_from_service_line() {
        try {
            await this.delete(`/enterprise/v1/account/${this.accountNumber}` +
                `/user-terminals/${this.userTerminalId}/${this.serviceLineNumber}`);
            this.userTerminal.serviceLineNumber = null;
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    /**
     * Converts the instance to a string (JSON encoded)
     */
    toString() {
        return JSON.stringify(this.userTerminal);
    }
    get object() {
        return this.userTerminal;
    }
    get accountNumber() {
        return this.userTerminal.accountNumber;
    }
    get active() {
        return this.userTerminal.active;
    }
    get dishSerialNumber() {
        return this.userTerminal.dishSerialNumber;
    }
    get kitSerialNumber() {
        return this.userTerminal.kitSerialNumber;
    }
    get nickname() {
        return this.userTerminal.nickname;
    }
    get routers() {
        return this.userTerminal.routers.map(router => new router_1.default(this.client_id, this.client_secret, router));
    }
    get serviceLineNumber() {
        return this.userTerminal.serviceLineNumber;
    }
    get userTerminalId() {
        return this.userTerminal.userTerminalId;
    }
}
exports.default = UserTerminal;
exports.UserTerminal = UserTerminal;
