// @gibme/starlink/dist/local/api/grpc_api.js
"use strict";
// Copyright (c) 2024, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GRPCApi = exports.DeviceDefinition = void 0;
const abort_controller_1 = __importDefault(require("abort-controller"));
const nice_grpc_1 = require("nice-grpc");
const service_1 = require("../../protobuf/spacex/api/device/service");
Object.defineProperty(exports, "DeviceDefinition", { enumerable: true, get: function () { return service_1.DeviceDefinition; } });
class GRPCApi {
    constructor(host, port, timeout) {
        this.host = host;
        this.port = port;
        this.timeout = timeout;
        this.channel = (0, nice_grpc_1.createChannel)(`${host}:${port}`);
        this.client = (0, nice_grpc_1.createClient)(service_1.DeviceDefinition, this.channel);
    }
    /**
     * Closes the underlying gRPC channel and releases resources
     */
    close() {
        this.channel.close();
    }
    /**
     * Performs a handle request against the GRPC server
     *
     * @param request
     * @param timeout
     * @protected
     */
    async handle(request, timeout = this.timeout) {
        const controller = new abort_controller_1.default();
        let _timeout;
        if (timeout) {
            _timeout = setTimeout(() => controller.abort(), timeout);
        }
        const response = await this.client.handle(request, { signal: controller.signal });
        if (_timeout) {
            clearTimeout(_timeout);
        }
        return response;
    }
}
exports.default = GRPCApi;
exports.GRPCApi = GRPCApi;
