// @gibme/starlink/dist/enterprise/api/base_api.js
"use strict";
// Copyright (c) 2024, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const memory_1 = __importDefault(require("@gibme/cache/memory"));
const fetch_1 = __importDefault(require("@gibme/fetch"));
/** @ignore */
const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));
/**
 * Provides a base-level abstraction for interacting with the Starlink Enterprise API
 */
class BaseAPI {
    constructor(client_id, client_secret) {
        this.client_id = client_id;
        this.client_secret = client_secret;
        this.baseUrl = 'https://web-api.starlink.com';
    }
    /**
     * Disconnects the internal rate-limit cache and releases its resources
     */
    static async disconnect() {
        await BaseAPI.cache.disconnect();
    }
    /**
     * Attempts to authenticate with the Starlink Enterprise API
     * If it succeeds, the token is stored internally for later use
     *
     * @private
     */
    async authenticate() {
        const response = await fetch_1.default.post('https://www.starlink.com/api/auth/connect/token', {
            formData: {
                client_id: this.client_id,
                client_secret: this.client_secret,
                grant_type: 'client_credentials'
            }
        });
        if (!response.ok)
            return false;
        const token = await response.json();
        if (token.access_token) {
            BaseAPI.token = token;
            return true;
        }
        delete BaseAPI.token;
        return false;
    }
    /**
     * Performs an HTTP request against the Starlink Enterprise API
     *
     * @param method
     * @param endpoint
     * @param params
     * @param payload
     * @param is_retry
     * @private
     */
    async execute(method, endpoint, params = {}, payload, is_retry = false) {
        var _a, _b;
        method = method.toUpperCase();
        // if we don't have a token, and we cannot authenticate to get one, fail
        if (!BaseAPI.token && !await this.authenticate()) {
            throw new Error('Could not authenticate with API');
        }
        // if we are close to exceeding the rate limit, back off and wait 100ms
        if (await BaseAPI.requestCount() >= BaseAPI.requestLimit - 5) {
            await sleep(100);
            return this.execute(method, endpoint, params, payload, is_retry);
        }
        // sets a key in the cache for the request, so that it can be counted towards our rate limit
        await BaseAPI.cache.set({
            date: Date.now(),
            endpoint,
            method,
            params,
            payload
        }, {});
        // construct our headers
        const headers = {
            Accept: 'application/json',
            Authorization: `${(_a = BaseAPI.token) === null || _a === void 0 ? void 0 : _a.token_type} ${(_b = BaseAPI.token) === null || _b === void 0 ? void 0 : _b.access_token}`
        };
        // construct any query string parameters
        const qs = new URLSearchParams();
        for (const key in params) {
            qs.set(key, params[key]);
        }
        // perform the fetch request
        const response = await (0, fetch_1.default)(`${this.baseUrl}${endpoint}?${qs.toString()}`, {
            headers,
            json: (method === 'PATCH' || method === 'POST' || method === 'PUT')
                ? payload
                : undefined,
            method
        });
        if (!response.ok) {
            /**
             * If we received a 401 and this is our first attempt, chances are our token expired,
             * and we will attempt to re-authenticate and then retry this request. If authentication
             * fails, or we've already retried, we'll continue with our normal path and throw an error
             */
            if (response.status === 401 && !is_retry) {
                if (await this.authenticate()) {
                    return this.execute(method, endpoint, params, payload, true);
                }
            }
            throw new Error(`${response.url} [${response.status}] ${response.statusText}`);
        }
        return await response.json();
    }
    static async requestCount() {
        return (await BaseAPI.cache.keys()).length;
    }
    /**
     * Performs an HTTP delete request against the Starlink Enterprise API
     *
     * @param endpoint
     * @param params
     * @protected
     */
    async delete(endpoint, params = {}) {
        return this.execute('DELETE', endpoint, params);
    }
    /**
     * Performs an HTTP get request against the Starlink Enterprise API
     *
     * @param endpoint
     * @param params
     * @protected
     */
    async get(endpoint, params = {}) {
        return this.execute('GET', endpoint, params);
    }
    /**
     * Performs an HTTP post request against the Starlink Enterprise API
     *
     * @param endpoint
     * @param payload
     * @protected
     */
    async post(endpoint, payload) {
        return this.execute('POST', endpoint, {}, payload);
    }
    /**
     * Performs an HTTP put request against the Starlink Enterprise API
     *
     * @param endpoint
     * @param payload
     * @protected
     */
    async put(endpoint, payload) {
        return this.execute('PUT', endpoint, {}, payload);
    }
}
/**
 * We are limited to 250 requests per minute from a single IP address
 * this cache helps to track this and automatically times out old
 * requests
 */
BaseAPI.cache = new memory_1.default({
    checkperiod: 1,
    stdTTL: 60
});
BaseAPI.requestLimit = 250;
exports.default = BaseAPI;
