// @gibme/starlink/dist/local/dishy.js
"use strict";
// Copyright (c) 2024, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dishy = void 0;
const grpc_api_1 = __importDefault(require("./api/grpc_api"));
class Dishy extends grpc_api_1.default {
    constructor(host = '192.168.100.1', port = 9200, timeout) {
        super(host, port, timeout);
    }
    /**
     * Fetches diagnostic information from a dishy
     */
    async fetch_diagnostics() {
        const response = await this.handle({ getDiagnostics: {} });
        if (!response.dishGetDiagnostics) {
            throw new Error(`No diagnostics returned from ${this.host}:${this.port}`);
        }
        return response.dishGetDiagnostics;
    }
    /**
     * Fetches history information from a dishy
     */
    async fetch_history() {
        const response = await this.handle({ getHistory: {} });
        if (!response.dishGetHistory) {
            throw new Error(`No history returned from ${this.host}:${this.port}`);
        }
        return response.dishGetHistory;
    }
    /**
     * Fetches location information from a dishy
     *
     * Note: Location information must be enabled in the dishy settings
     */
    async fetch_location() {
        const response = await this.handle({ getLocation: {} });
        if (!response.getLocation) {
            throw new Error(`No location returned from ${this.host}:${this.port}`);
        }
        return response.getLocation;
    }
    /**
     * Fetches obstruction map information from a dishy
     */
    async fetch_obstruction_map() {
        const response = await this.handle({ dishGetObstructionMap: {} });
        if (!response.dishGetObstructionMap) {
            throw new Error(`No obstruction map returned from ${this.host}:${this.port}`);
        }
        return response.dishGetObstructionMap;
    }
    /**
     * Fetches status information from a dishy
     */
    async fetch_status() {
        const response = await this.handle({ getStatus: {} });
        if (!response.dishGetStatus) {
            throw new Error(`No status returned from ${this.host}:${this.port}`);
        }
        return response.dishGetStatus;
    }
    /**
     * Issues a reboot request to a dishy
     */
    async reboot() {
        try {
            await this.handle({ reboot: {} });
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    /**
     * Issues a stow request to a dishy
     *
     * Note: The dishy must support stow/unstow operations
     */
    async stow() {
        try {
            await this.handle({ dishStow: { unstow: false } });
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    /**
     * Issues an unstow request to a dishy
     *
     * Note: The dishy must support stow/unstow operations
     */
    async unstow() {
        try {
            await this.handle({ dishStow: { unstow: true } });
            return true;
        }
        catch (_a) {
            return false;
        }
    }
}
exports.default = Dishy;
exports.Dishy = Dishy;
