// @gibme/starlink/dist/protobuf/spacex/api/device/common.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.11.6
//   protoc               v3.21.12
// source: spacex/api/device/common.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.ECEFPosition = exports.LLAPosition = exports.WifiNetworkInterface_ThermalStatus = exports.WifiNetworkInterface = exports.EthernetNetworkInterface = exports.NetworkInterface_TxStats = exports.NetworkInterface_RxStats = exports.NetworkInterface = exports.ChallengeResponse = exports.AuthenticateRequest = exports.BondingChallenge = exports.PingResult = exports.PingTarget = exports.GetNextIdResponse = exports.GetNextIdRequest = exports.SignedData = exports.DeviceState = exports.DeviceInfo = exports.EthernetNetworkInterface_Duplex = exports.protobufPackage = void 0;
exports.ethernetNetworkInterface_DuplexFromJSON = ethernetNetworkInterface_DuplexFromJSON;
exports.ethernetNetworkInterface_DuplexToJSON = ethernetNetworkInterface_DuplexToJSON;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
exports.protobufPackage = "SpaceX.API.Device";
var EthernetNetworkInterface_Duplex;
(function (EthernetNetworkInterface_Duplex) {
    EthernetNetworkInterface_Duplex[EthernetNetworkInterface_Duplex["UNKNOWN"] = 0] = "UNKNOWN";
    EthernetNetworkInterface_Duplex[EthernetNetworkInterface_Duplex["HALF"] = 1] = "HALF";
    EthernetNetworkInterface_Duplex[EthernetNetworkInterface_Duplex["FULL"] = 2] = "FULL";
    EthernetNetworkInterface_Duplex[EthernetNetworkInterface_Duplex["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(EthernetNetworkInterface_Duplex || (exports.EthernetNetworkInterface_Duplex = EthernetNetworkInterface_Duplex = {}));
function ethernetNetworkInterface_DuplexFromJSON(object) {
    switch (object) {
        case 0:
        case "UNKNOWN":
            return EthernetNetworkInterface_Duplex.UNKNOWN;
        case 1:
        case "HALF":
            return EthernetNetworkInterface_Duplex.HALF;
        case 2:
        case "FULL":
            return EthernetNetworkInterface_Duplex.FULL;
        case -1:
        case "UNRECOGNIZED":
        default:
            return EthernetNetworkInterface_Duplex.UNRECOGNIZED;
    }
}
function ethernetNetworkInterface_DuplexToJSON(object) {
    switch (object) {
        case EthernetNetworkInterface_Duplex.UNKNOWN:
            return "UNKNOWN";
        case EthernetNetworkInterface_Duplex.HALF:
            return "HALF";
        case EthernetNetworkInterface_Duplex.FULL:
            return "FULL";
        case EthernetNetworkInterface_Duplex.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBaseDeviceInfo() {
    return {
        id: undefined,
        hardwareVersion: undefined,
        softwareVersion: undefined,
        countryCode: undefined,
        utcOffsetS: undefined,
    };
}
exports.DeviceInfo = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.id !== undefined) {
            writer.uint32(10).string(message.id);
        }
        if (message.hardwareVersion !== undefined) {
            writer.uint32(18).string(message.hardwareVersion);
        }
        if (message.softwareVersion !== undefined) {
            writer.uint32(26).string(message.softwareVersion);
        }
        if (message.countryCode !== undefined) {
            writer.uint32(34).string(message.countryCode);
        }
        if (message.utcOffsetS !== undefined) {
            writer.uint32(40).int32(message.utcOffsetS);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDeviceInfo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.hardwareVersion = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.softwareVersion = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.countryCode = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 40) {
                        break;
                    }
                    message.utcOffsetS = reader.int32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? globalThis.String(object.id) : undefined,
            hardwareVersion: isSet(object.hardwareVersion)
                ? globalThis.String(object.hardwareVersion)
                : isSet(object.hardware_version)
                    ? globalThis.String(object.hardware_version)
                    : undefined,
            softwareVersion: isSet(object.softwareVersion)
                ? globalThis.String(object.softwareVersion)
                : isSet(object.software_version)
                    ? globalThis.String(object.software_version)
                    : undefined,
            countryCode: isSet(object.countryCode)
                ? globalThis.String(object.countryCode)
                : isSet(object.country_code)
                    ? globalThis.String(object.country_code)
                    : undefined,
            utcOffsetS: isSet(object.utcOffsetS)
                ? globalThis.Number(object.utcOffsetS)
                : isSet(object.utc_offset_s)
                    ? globalThis.Number(object.utc_offset_s)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== undefined) {
            obj.id = message.id;
        }
        if (message.hardwareVersion !== undefined) {
            obj.hardwareVersion = message.hardwareVersion;
        }
        if (message.softwareVersion !== undefined) {
            obj.softwareVersion = message.softwareVersion;
        }
        if (message.countryCode !== undefined) {
            obj.countryCode = message.countryCode;
        }
        if (message.utcOffsetS !== undefined) {
            obj.utcOffsetS = Math.round(message.utcOffsetS);
        }
        return obj;
    },
    create(base) {
        return exports.DeviceInfo.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e;
        const message = createBaseDeviceInfo();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : undefined;
        message.hardwareVersion = (_b = object.hardwareVersion) !== null && _b !== void 0 ? _b : undefined;
        message.softwareVersion = (_c = object.softwareVersion) !== null && _c !== void 0 ? _c : undefined;
        message.countryCode = (_d = object.countryCode) !== null && _d !== void 0 ? _d : undefined;
        message.utcOffsetS = (_e = object.utcOffsetS) !== null && _e !== void 0 ? _e : undefined;
        return message;
    },
};
function createBaseDeviceState() {
    return { uptimeS: undefined };
}
exports.DeviceState = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.uptimeS !== undefined) {
            writer.uint32(8).uint64(message.uptimeS);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDeviceState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.uptimeS = longToNumber(reader.uint64());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            uptimeS: isSet(object.uptimeS)
                ? globalThis.Number(object.uptimeS)
                : isSet(object.uptime_s)
                    ? globalThis.Number(object.uptime_s)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.uptimeS !== undefined) {
            obj.uptimeS = Math.round(message.uptimeS);
        }
        return obj;
    },
    create(base) {
        return exports.DeviceState.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseDeviceState();
        message.uptimeS = (_a = object.uptimeS) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBaseSignedData() {
    return { data: undefined, signature: undefined };
}
exports.SignedData = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.data !== undefined) {
            writer.uint32(10).bytes(message.data);
        }
        if (message.signature !== undefined) {
            writer.uint32(18).bytes(message.signature);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSignedData();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.data = reader.bytes();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.signature = reader.bytes();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            data: isSet(object.data) ? bytesFromBase64(object.data) : undefined,
            signature: isSet(object.signature) ? bytesFromBase64(object.signature) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.data !== undefined) {
            obj.data = base64FromBytes(message.data);
        }
        if (message.signature !== undefined) {
            obj.signature = base64FromBytes(message.signature);
        }
        return obj;
    },
    create(base) {
        return exports.SignedData.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseSignedData();
        message.data = (_a = object.data) !== null && _a !== void 0 ? _a : undefined;
        message.signature = (_b = object.signature) !== null && _b !== void 0 ? _b : undefined;
        return message;
    },
};
function createBaseGetNextIdRequest() {
    return {};
}
exports.GetNextIdRequest = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetNextIdRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.GetNextIdRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseGetNextIdRequest();
        return message;
    },
};
function createBaseGetNextIdResponse() {
    return { id: undefined, epochId: undefined };
}
exports.GetNextIdResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.id !== undefined) {
            writer.uint32(8).uint64(message.id);
        }
        if (message.epochId !== undefined) {
            writer.uint32(16).uint64(message.epochId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetNextIdResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.id = longToNumber(reader.uint64());
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.epochId = longToNumber(reader.uint64());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? globalThis.Number(object.id) : undefined,
            epochId: isSet(object.epochId)
                ? globalThis.Number(object.epochId)
                : isSet(object.epoch_id)
                    ? globalThis.Number(object.epoch_id)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== undefined) {
            obj.id = Math.round(message.id);
        }
        if (message.epochId !== undefined) {
            obj.epochId = Math.round(message.epochId);
        }
        return obj;
    },
    create(base) {
        return exports.GetNextIdResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseGetNextIdResponse();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : undefined;
        message.epochId = (_b = object.epochId) !== null && _b !== void 0 ? _b : undefined;
        return message;
    },
};
function createBasePingTarget() {
    return { service: undefined, location: undefined, address: undefined };
}
exports.PingTarget = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.service !== undefined) {
            writer.uint32(10).string(message.service);
        }
        if (message.location !== undefined) {
            writer.uint32(18).string(message.location);
        }
        if (message.address !== undefined) {
            writer.uint32(26).string(message.address);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePingTarget();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.service = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.location = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.address = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            service: isSet(object.service) ? globalThis.String(object.service) : undefined,
            location: isSet(object.location) ? globalThis.String(object.location) : undefined,
            address: isSet(object.address) ? globalThis.String(object.address) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.service !== undefined) {
            obj.service = message.service;
        }
        if (message.location !== undefined) {
            obj.location = message.location;
        }
        if (message.address !== undefined) {
            obj.address = message.address;
        }
        return obj;
    },
    create(base) {
        return exports.PingTarget.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBasePingTarget();
        message.service = (_a = object.service) !== null && _a !== void 0 ? _a : undefined;
        message.location = (_b = object.location) !== null && _b !== void 0 ? _b : undefined;
        message.address = (_c = object.address) !== null && _c !== void 0 ? _c : undefined;
        return message;
    },
};
function createBasePingResult() {
    return { target: undefined, dropRate: undefined, latencyMs: undefined };
}
exports.PingResult = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.target !== undefined) {
            exports.PingTarget.encode(message.target, writer.uint32(26).fork()).join();
        }
        if (message.dropRate !== undefined) {
            writer.uint32(13).float(message.dropRate);
        }
        if (message.latencyMs !== undefined) {
            writer.uint32(21).float(message.latencyMs);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePingResult();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.target = exports.PingTarget.decode(reader, reader.uint32());
                    continue;
                }
                case 1: {
                    if (tag !== 13) {
                        break;
                    }
                    message.dropRate = reader.float();
                    continue;
                }
                case 2: {
                    if (tag !== 21) {
                        break;
                    }
                    message.latencyMs = reader.float();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            target: isSet(object.target) ? exports.PingTarget.fromJSON(object.target) : undefined,
            dropRate: isSet(object.dropRate) ? globalThis.Number(object.dropRate) : undefined,
            latencyMs: isSet(object.latencyMs) ? globalThis.Number(object.latencyMs) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.target !== undefined) {
            obj.target = exports.PingTarget.toJSON(message.target);
        }
        if (message.dropRate !== undefined) {
            obj.dropRate = message.dropRate;
        }
        if (message.latencyMs !== undefined) {
            obj.latencyMs = message.latencyMs;
        }
        return obj;
    },
    create(base) {
        return exports.PingResult.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBasePingResult();
        message.target = (object.target !== undefined && object.target !== null)
            ? exports.PingTarget.fromPartial(object.target)
            : undefined;
        message.dropRate = (_a = object.dropRate) !== null && _a !== void 0 ? _a : undefined;
        message.latencyMs = (_b = object.latencyMs) !== null && _b !== void 0 ? _b : undefined;
        return message;
    },
};
function createBaseBondingChallenge() {
    return { dishId: undefined, wifiId: undefined, nonce: undefined };
}
exports.BondingChallenge = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.dishId !== undefined) {
            writer.uint32(10).string(message.dishId);
        }
        if (message.wifiId !== undefined) {
            writer.uint32(18).string(message.wifiId);
        }
        if (message.nonce !== undefined) {
            writer.uint32(26).bytes(message.nonce);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseBondingChallenge();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.dishId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.wifiId = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.nonce = reader.bytes();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            dishId: isSet(object.dishId)
                ? globalThis.String(object.dishId)
                : isSet(object.dish_id)
                    ? globalThis.String(object.dish_id)
                    : undefined,
            wifiId: isSet(object.wifiId)
                ? globalThis.String(object.wifiId)
                : isSet(object.wifi_id)
                    ? globalThis.String(object.wifi_id)
                    : undefined,
            nonce: isSet(object.nonce) ? bytesFromBase64(object.nonce) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.dishId !== undefined) {
            obj.dishId = message.dishId;
        }
        if (message.wifiId !== undefined) {
            obj.wifiId = message.wifiId;
        }
        if (message.nonce !== undefined) {
            obj.nonce = base64FromBytes(message.nonce);
        }
        return obj;
    },
    create(base) {
        return exports.BondingChallenge.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseBondingChallenge();
        message.dishId = (_a = object.dishId) !== null && _a !== void 0 ? _a : undefined;
        message.wifiId = (_b = object.wifiId) !== null && _b !== void 0 ? _b : undefined;
        message.nonce = (_c = object.nonce) !== null && _c !== void 0 ? _c : undefined;
        return message;
    },
};
function createBaseAuthenticateRequest() {
    return { challenge: undefined };
}
exports.AuthenticateRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.challenge !== undefined) {
            exports.SignedData.encode(message.challenge, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAuthenticateRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.challenge = exports.SignedData.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { challenge: isSet(object.challenge) ? exports.SignedData.fromJSON(object.challenge) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.challenge !== undefined) {
            obj.challenge = exports.SignedData.toJSON(message.challenge);
        }
        return obj;
    },
    create(base) {
        return exports.AuthenticateRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseAuthenticateRequest();
        message.challenge = (object.challenge !== undefined && object.challenge !== null)
            ? exports.SignedData.fromPartial(object.challenge)
            : undefined;
        return message;
    },
};
function createBaseChallengeResponse() {
    return { signature: undefined, certificateChain: undefined };
}
exports.ChallengeResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.signature !== undefined) {
            writer.uint32(10).bytes(message.signature);
        }
        if (message.certificateChain !== undefined) {
            writer.uint32(18).bytes(message.certificateChain);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseChallengeResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.signature = reader.bytes();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.certificateChain = reader.bytes();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            signature: isSet(object.signature) ? bytesFromBase64(object.signature) : undefined,
            certificateChain: isSet(object.certificateChain)
                ? bytesFromBase64(object.certificateChain)
                : isSet(object.certificate_chain)
                    ? bytesFromBase64(object.certificate_chain)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.signature !== undefined) {
            obj.signature = base64FromBytes(message.signature);
        }
        if (message.certificateChain !== undefined) {
            obj.certificateChain = base64FromBytes(message.certificateChain);
        }
        return obj;
    },
    create(base) {
        return exports.ChallengeResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseChallengeResponse();
        message.signature = (_a = object.signature) !== null && _a !== void 0 ? _a : undefined;
        message.certificateChain = (_b = object.certificateChain) !== null && _b !== void 0 ? _b : undefined;
        return message;
    },
};
function createBaseNetworkInterface() {
    return { name: undefined, rxStats: undefined, txStats: undefined, ethernet: undefined, wifi: undefined };
}
exports.NetworkInterface = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.name !== undefined) {
            writer.uint32(10).string(message.name);
        }
        if (message.rxStats !== undefined) {
            exports.NetworkInterface_RxStats.encode(message.rxStats, writer.uint32(18).fork()).join();
        }
        if (message.txStats !== undefined) {
            exports.NetworkInterface_TxStats.encode(message.txStats, writer.uint32(26).fork()).join();
        }
        if (message.ethernet !== undefined) {
            exports.EthernetNetworkInterface.encode(message.ethernet, writer.uint32(8002).fork()).join();
        }
        if (message.wifi !== undefined) {
            exports.WifiNetworkInterface.encode(message.wifi, writer.uint32(8010).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseNetworkInterface();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.rxStats = exports.NetworkInterface_RxStats.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.txStats = exports.NetworkInterface_TxStats.decode(reader, reader.uint32());
                    continue;
                }
                case 1000: {
                    if (tag !== 8002) {
                        break;
                    }
                    message.ethernet = exports.EthernetNetworkInterface.decode(reader, reader.uint32());
                    continue;
                }
                case 1001: {
                    if (tag !== 8010) {
                        break;
                    }
                    message.wifi = exports.WifiNetworkInterface.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            name: isSet(object.name) ? globalThis.String(object.name) : undefined,
            rxStats: isSet(object.rxStats)
                ? exports.NetworkInterface_RxStats.fromJSON(object.rxStats)
                : isSet(object.rx_stats)
                    ? exports.NetworkInterface_RxStats.fromJSON(object.rx_stats)
                    : undefined,
            txStats: isSet(object.txStats)
                ? exports.NetworkInterface_TxStats.fromJSON(object.txStats)
                : isSet(object.tx_stats)
                    ? exports.NetworkInterface_TxStats.fromJSON(object.tx_stats)
                    : undefined,
            ethernet: isSet(object.ethernet) ? exports.EthernetNetworkInterface.fromJSON(object.ethernet) : undefined,
            wifi: isSet(object.wifi) ? exports.WifiNetworkInterface.fromJSON(object.wifi) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== undefined) {
            obj.name = message.name;
        }
        if (message.rxStats !== undefined) {
            obj.rxStats = exports.NetworkInterface_RxStats.toJSON(message.rxStats);
        }
        if (message.txStats !== undefined) {
            obj.txStats = exports.NetworkInterface_TxStats.toJSON(message.txStats);
        }
        if (message.ethernet !== undefined) {
            obj.ethernet = exports.EthernetNetworkInterface.toJSON(message.ethernet);
        }
        if (message.wifi !== undefined) {
            obj.wifi = exports.WifiNetworkInterface.toJSON(message.wifi);
        }
        return obj;
    },
    create(base) {
        return exports.NetworkInterface.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseNetworkInterface();
        message.name = (_a = object.name) !== null && _a !== void 0 ? _a : undefined;
        message.rxStats = (object.rxStats !== undefined && object.rxStats !== null)
            ? exports.NetworkInterface_RxStats.fromPartial(object.rxStats)
            : undefined;
        message.txStats = (object.txStats !== undefined && object.txStats !== null)
            ? exports.NetworkInterface_TxStats.fromPartial(object.txStats)
            : undefined;
        message.ethernet = (object.ethernet !== undefined && object.ethernet !== null)
            ? exports.EthernetNetworkInterface.fromPartial(object.ethernet)
            : undefined;
        message.wifi = (object.wifi !== undefined && object.wifi !== null)
            ? exports.WifiNetworkInterface.fromPartial(object.wifi)
            : undefined;
        return message;
    },
};
function createBaseNetworkInterface_RxStats() {
    return { bytes: undefined, packets: undefined, frameErrors: undefined };
}
exports.NetworkInterface_RxStats = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.bytes !== undefined) {
            writer.uint32(8).uint64(message.bytes);
        }
        if (message.packets !== undefined) {
            writer.uint32(16).uint64(message.packets);
        }
        if (message.frameErrors !== undefined) {
            writer.uint32(24).uint64(message.frameErrors);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseNetworkInterface_RxStats();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.bytes = longToNumber(reader.uint64());
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.packets = longToNumber(reader.uint64());
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.frameErrors = longToNumber(reader.uint64());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            bytes: isSet(object.bytes) ? globalThis.Number(object.bytes) : undefined,
            packets: isSet(object.packets) ? globalThis.Number(object.packets) : undefined,
            frameErrors: isSet(object.frameErrors)
                ? globalThis.Number(object.frameErrors)
                : isSet(object.frame_errors)
                    ? globalThis.Number(object.frame_errors)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.bytes !== undefined) {
            obj.bytes = Math.round(message.bytes);
        }
        if (message.packets !== undefined) {
            obj.packets = Math.round(message.packets);
        }
        if (message.frameErrors !== undefined) {
            obj.frameErrors = Math.round(message.frameErrors);
        }
        return obj;
    },
    create(base) {
        return exports.NetworkInterface_RxStats.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseNetworkInterface_RxStats();
        message.bytes = (_a = object.bytes) !== null && _a !== void 0 ? _a : undefined;
        message.packets = (_b = object.packets) !== null && _b !== void 0 ? _b : undefined;
        message.frameErrors = (_c = object.frameErrors) !== null && _c !== void 0 ? _c : undefined;
        return message;
    },
};
function createBaseNetworkInterface_TxStats() {
    return { bytes: undefined, packets: undefined };
}
exports.NetworkInterface_TxStats = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.bytes !== undefined) {
            writer.uint32(8).uint64(message.bytes);
        }
        if (message.packets !== undefined) {
            writer.uint32(16).uint64(message.packets);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseNetworkInterface_TxStats();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.bytes = longToNumber(reader.uint64());
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.packets = longToNumber(reader.uint64());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            bytes: isSet(object.bytes) ? globalThis.Number(object.bytes) : undefined,
            packets: isSet(object.packets) ? globalThis.Number(object.packets) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.bytes !== undefined) {
            obj.bytes = Math.round(message.bytes);
        }
        if (message.packets !== undefined) {
            obj.packets = Math.round(message.packets);
        }
        return obj;
    },
    create(base) {
        return exports.NetworkInterface_TxStats.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseNetworkInterface_TxStats();
        message.bytes = (_a = object.bytes) !== null && _a !== void 0 ? _a : undefined;
        message.packets = (_b = object.packets) !== null && _b !== void 0 ? _b : undefined;
        return message;
    },
};
function createBaseEthernetNetworkInterface() {
    return { linkDetected: undefined, speedMbps: undefined, autonegotiationOn: undefined, duplex: undefined };
}
exports.EthernetNetworkInterface = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.linkDetected !== undefined) {
            writer.uint32(8).bool(message.linkDetected);
        }
        if (message.speedMbps !== undefined) {
            writer.uint32(16).uint32(message.speedMbps);
        }
        if (message.autonegotiationOn !== undefined) {
            writer.uint32(24).bool(message.autonegotiationOn);
        }
        if (message.duplex !== undefined) {
            writer.uint32(32).int32(message.duplex);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEthernetNetworkInterface();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.linkDetected = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.speedMbps = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.autonegotiationOn = reader.bool();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.duplex = reader.int32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            linkDetected: isSet(object.linkDetected)
                ? globalThis.Boolean(object.linkDetected)
                : isSet(object.link_detected)
                    ? globalThis.Boolean(object.link_detected)
                    : undefined,
            speedMbps: isSet(object.speedMbps)
                ? globalThis.Number(object.speedMbps)
                : isSet(object.speed_mbps)
                    ? globalThis.Number(object.speed_mbps)
                    : undefined,
            autonegotiationOn: isSet(object.autonegotiationOn)
                ? globalThis.Boolean(object.autonegotiationOn)
                : isSet(object.autonegotiation_on)
                    ? globalThis.Boolean(object.autonegotiation_on)
                    : undefined,
            duplex: isSet(object.duplex) ? ethernetNetworkInterface_DuplexFromJSON(object.duplex) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.linkDetected !== undefined) {
            obj.linkDetected = message.linkDetected;
        }
        if (message.speedMbps !== undefined) {
            obj.speedMbps = Math.round(message.speedMbps);
        }
        if (message.autonegotiationOn !== undefined) {
            obj.autonegotiationOn = message.autonegotiationOn;
        }
        if (message.duplex !== undefined) {
            obj.duplex = ethernetNetworkInterface_DuplexToJSON(message.duplex);
        }
        return obj;
    },
    create(base) {
        return exports.EthernetNetworkInterface.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseEthernetNetworkInterface();
        message.linkDetected = (_a = object.linkDetected) !== null && _a !== void 0 ? _a : undefined;
        message.speedMbps = (_b = object.speedMbps) !== null && _b !== void 0 ? _b : undefined;
        message.autonegotiationOn = (_c = object.autonegotiationOn) !== null && _c !== void 0 ? _c : undefined;
        message.duplex = (_d = object.duplex) !== null && _d !== void 0 ? _d : undefined;
        return message;
    },
};
function createBaseWifiNetworkInterface() {
    return { thermalStatus: undefined };
}
exports.WifiNetworkInterface = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.thermalStatus !== undefined) {
            exports.WifiNetworkInterface_ThermalStatus.encode(message.thermalStatus, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiNetworkInterface();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.thermalStatus = exports.WifiNetworkInterface_ThermalStatus.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            thermalStatus: isSet(object.thermalStatus)
                ? exports.WifiNetworkInterface_ThermalStatus.fromJSON(object.thermalStatus)
                : isSet(object.thermal_status)
                    ? exports.WifiNetworkInterface_ThermalStatus.fromJSON(object.thermal_status)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.thermalStatus !== undefined) {
            obj.thermalStatus = exports.WifiNetworkInterface_ThermalStatus.toJSON(message.thermalStatus);
        }
        return obj;
    },
    create(base) {
        return exports.WifiNetworkInterface.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseWifiNetworkInterface();
        message.thermalStatus = (object.thermalStatus !== undefined && object.thermalStatus !== null)
            ? exports.WifiNetworkInterface_ThermalStatus.fromPartial(object.thermalStatus)
            : undefined;
        return message;
    },
};
function createBaseWifiNetworkInterface_ThermalStatus() {
    return { level: undefined, temp: undefined };
}
exports.WifiNetworkInterface_ThermalStatus = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.level !== undefined) {
            writer.uint32(8).uint32(message.level);
        }
        if (message.temp !== undefined) {
            writer.uint32(16).uint32(message.temp);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiNetworkInterface_ThermalStatus();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.level = reader.uint32();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.temp = reader.uint32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            level: isSet(object.level) ? globalThis.Number(object.level) : undefined,
            temp: isSet(object.temp) ? globalThis.Number(object.temp) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.level !== undefined) {
            obj.level = Math.round(message.level);
        }
        if (message.temp !== undefined) {
            obj.temp = Math.round(message.temp);
        }
        return obj;
    },
    create(base) {
        return exports.WifiNetworkInterface_ThermalStatus.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseWifiNetworkInterface_ThermalStatus();
        message.level = (_a = object.level) !== null && _a !== void 0 ? _a : undefined;
        message.temp = (_b = object.temp) !== null && _b !== void 0 ? _b : undefined;
        return message;
    },
};
function createBaseLLAPosition() {
    return { lat: undefined, lon: undefined, alt: undefined };
}
exports.LLAPosition = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.lat !== undefined) {
            writer.uint32(9).double(message.lat);
        }
        if (message.lon !== undefined) {
            writer.uint32(17).double(message.lon);
        }
        if (message.alt !== undefined) {
            writer.uint32(25).double(message.alt);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseLLAPosition();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 9) {
                        break;
                    }
                    message.lat = reader.double();
                    continue;
                }
                case 2: {
                    if (tag !== 17) {
                        break;
                    }
                    message.lon = reader.double();
                    continue;
                }
                case 3: {
                    if (tag !== 25) {
                        break;
                    }
                    message.alt = reader.double();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            lat: isSet(object.lat) ? globalThis.Number(object.lat) : undefined,
            lon: isSet(object.lon) ? globalThis.Number(object.lon) : undefined,
            alt: isSet(object.alt) ? globalThis.Number(object.alt) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.lat !== undefined) {
            obj.lat = message.lat;
        }
        if (message.lon !== undefined) {
            obj.lon = message.lon;
        }
        if (message.alt !== undefined) {
            obj.alt = message.alt;
        }
        return obj;
    },
    create(base) {
        return exports.LLAPosition.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseLLAPosition();
        message.lat = (_a = object.lat) !== null && _a !== void 0 ? _a : undefined;
        message.lon = (_b = object.lon) !== null && _b !== void 0 ? _b : undefined;
        message.alt = (_c = object.alt) !== null && _c !== void 0 ? _c : undefined;
        return message;
    },
};
function createBaseECEFPosition() {
    return { x: undefined, y: undefined, z: undefined };
}
exports.ECEFPosition = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.x !== undefined) {
            writer.uint32(9).double(message.x);
        }
        if (message.y !== undefined) {
            writer.uint32(17).double(message.y);
        }
        if (message.z !== undefined) {
            writer.uint32(25).double(message.z);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseECEFPosition();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 9) {
                        break;
                    }
                    message.x = reader.double();
                    continue;
                }
                case 2: {
                    if (tag !== 17) {
                        break;
                    }
                    message.y = reader.double();
                    continue;
                }
                case 3: {
                    if (tag !== 25) {
                        break;
                    }
                    message.z = reader.double();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            x: isSet(object.x) ? globalThis.Number(object.x) : undefined,
            y: isSet(object.y) ? globalThis.Number(object.y) : undefined,
            z: isSet(object.z) ? globalThis.Number(object.z) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.x !== undefined) {
            obj.x = message.x;
        }
        if (message.y !== undefined) {
            obj.y = message.y;
        }
        if (message.z !== undefined) {
            obj.z = message.z;
        }
        return obj;
    },
    create(base) {
        return exports.ECEFPosition.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseECEFPosition();
        message.x = (_a = object.x) !== null && _a !== void 0 ? _a : undefined;
        message.y = (_b = object.y) !== null && _b !== void 0 ? _b : undefined;
        message.z = (_c = object.z) !== null && _c !== void 0 ? _c : undefined;
        return message;
    },
};
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function longToNumber(int64) {
    const num = globalThis.Number(int64.toString());
    if (num > globalThis.Number.MAX_SAFE_INTEGER) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    if (num < globalThis.Number.MIN_SAFE_INTEGER) {
        throw new globalThis.Error("Value is smaller than Number.MIN_SAFE_INTEGER");
    }
    return num;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
