// @gibme/starlink/dist/protobuf/spacex/api/device/command.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.11.6
//   protoc               v3.21.12
// source: spacex/api/device/command.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublicKey = exports.Capability = exports.protobufPackage = void 0;
exports.capabilityFromJSON = capabilityFromJSON;
exports.capabilityToJSON = capabilityToJSON;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
exports.protobufPackage = "SpaceX.API.Device";
var Capability;
(function (Capability) {
    Capability[Capability["READ"] = 0] = "READ";
    Capability[Capability["WRITE"] = 1] = "WRITE";
    Capability[Capability["DEBUG"] = 2] = "DEBUG";
    Capability[Capability["ADMIN"] = 3] = "ADMIN";
    Capability[Capability["SETUP"] = 4] = "SETUP";
    Capability[Capability["SET_SKU"] = 5] = "SET_SKU";
    Capability[Capability["REFRESH"] = 6] = "REFRESH";
    Capability[Capability["READ_PRIVATE"] = 7] = "READ_PRIVATE";
    Capability[Capability["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(Capability || (exports.Capability = Capability = {}));
function capabilityFromJSON(object) {
    switch (object) {
        case 0:
        case "READ":
            return Capability.READ;
        case 1:
        case "WRITE":
            return Capability.WRITE;
        case 2:
        case "DEBUG":
            return Capability.DEBUG;
        case 3:
        case "ADMIN":
            return Capability.ADMIN;
        case 4:
        case "SETUP":
            return Capability.SETUP;
        case 5:
        case "SET_SKU":
            return Capability.SET_SKU;
        case 6:
        case "REFRESH":
            return Capability.REFRESH;
        case 7:
        case "READ_PRIVATE":
            return Capability.READ_PRIVATE;
        case -1:
        case "UNRECOGNIZED":
        default:
            return Capability.UNRECOGNIZED;
    }
}
function capabilityToJSON(object) {
    switch (object) {
        case Capability.READ:
            return "READ";
        case Capability.WRITE:
            return "WRITE";
        case Capability.DEBUG:
            return "DEBUG";
        case Capability.ADMIN:
            return "ADMIN";
        case Capability.SETUP:
            return "SETUP";
        case Capability.SET_SKU:
            return "SET_SKU";
        case Capability.REFRESH:
            return "REFRESH";
        case Capability.READ_PRIVATE:
            return "READ_PRIVATE";
        case Capability.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBasePublicKey() {
    return { key: undefined, capabilities: [] };
}
exports.PublicKey = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.key !== undefined) {
            writer.uint32(10).string(message.key);
        }
        writer.uint32(18).fork();
        for (const v of message.capabilities) {
            writer.int32(v);
        }
        writer.join();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePublicKey();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                }
                case 2: {
                    if (tag === 16) {
                        message.capabilities.push(reader.int32());
                        continue;
                    }
                    if (tag === 18) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.capabilities.push(reader.int32());
                        }
                        continue;
                    }
                    break;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : undefined,
            capabilities: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.capabilities)
                ? object.capabilities.map((e) => capabilityFromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.key !== undefined) {
            obj.key = message.key;
        }
        if ((_a = message.capabilities) === null || _a === void 0 ? void 0 : _a.length) {
            obj.capabilities = message.capabilities.map((e) => capabilityToJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.PublicKey.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBasePublicKey();
        message.key = (_a = object.key) !== null && _a !== void 0 ? _a : undefined;
        message.capabilities = ((_b = object.capabilities) === null || _b === void 0 ? void 0 : _b.map((e) => e)) || [];
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
