// @gibme/starlink/dist/protobuf/spacex/api/device/dish.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.11.6
//   protoc               v3.21.12
// source: spacex/api/device/dish.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.DishGetDiagnosticsResponse_AlignmentStats = exports.DishGetDiagnosticsResponse_Location = exports.DishGetDiagnosticsResponse_Alerts = exports.DishGetDiagnosticsResponse = exports.DishAuthenticateResponse = exports.DishObstructionStats = exports.DishAlerts = exports.DishGetObstructionMapResponse = exports.DishGetObstructionMapRequest = exports.DishGetStatusResponse = exports.DishGetHistoryResponse = exports.DishGetContextResponse = exports.DishGetContextRequest = exports.DishStowResponse = exports.DishStowRequest = exports.DishGetDiagnosticsResponse_TestResultCode = exports.DishGetDiagnosticsResponse_DisablementCode = exports.DishGetDiagnosticsResponse_TestResult = exports.DishState = exports.protobufPackage = void 0;
exports.dishStateFromJSON = dishStateFromJSON;
exports.dishStateToJSON = dishStateToJSON;
exports.dishGetDiagnosticsResponse_TestResultFromJSON = dishGetDiagnosticsResponse_TestResultFromJSON;
exports.dishGetDiagnosticsResponse_TestResultToJSON = dishGetDiagnosticsResponse_TestResultToJSON;
exports.dishGetDiagnosticsResponse_DisablementCodeFromJSON = dishGetDiagnosticsResponse_DisablementCodeFromJSON;
exports.dishGetDiagnosticsResponse_DisablementCodeToJSON = dishGetDiagnosticsResponse_DisablementCodeToJSON;
exports.dishGetDiagnosticsResponse_TestResultCodeFromJSON = dishGetDiagnosticsResponse_TestResultCodeFromJSON;
exports.dishGetDiagnosticsResponse_TestResultCodeToJSON = dishGetDiagnosticsResponse_TestResultCodeToJSON;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const common_1 = require("./common");
exports.protobufPackage = "SpaceX.API.Device";
var DishState;
(function (DishState) {
    DishState[DishState["UNKNOWN"] = 0] = "UNKNOWN";
    DishState[DishState["CONNECTED"] = 1] = "CONNECTED";
    DishState[DishState["SEARCHING"] = 2] = "SEARCHING";
    DishState[DishState["BOOTING"] = 3] = "BOOTING";
    DishState[DishState["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(DishState || (exports.DishState = DishState = {}));
function dishStateFromJSON(object) {
    switch (object) {
        case 0:
        case "UNKNOWN":
            return DishState.UNKNOWN;
        case 1:
        case "CONNECTED":
            return DishState.CONNECTED;
        case 2:
        case "SEARCHING":
            return DishState.SEARCHING;
        case 3:
        case "BOOTING":
            return DishState.BOOTING;
        case -1:
        case "UNRECOGNIZED":
        default:
            return DishState.UNRECOGNIZED;
    }
}
function dishStateToJSON(object) {
    switch (object) {
        case DishState.UNKNOWN:
            return "UNKNOWN";
        case DishState.CONNECTED:
            return "CONNECTED";
        case DishState.SEARCHING:
            return "SEARCHING";
        case DishState.BOOTING:
            return "BOOTING";
        case DishState.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
/** / Result of the hardware self test that the UT runs on boot. */
var DishGetDiagnosticsResponse_TestResult;
(function (DishGetDiagnosticsResponse_TestResult) {
    /** NO_RESULT - / No hardware test has been run. */
    DishGetDiagnosticsResponse_TestResult[DishGetDiagnosticsResponse_TestResult["NO_RESULT"] = 0] = "NO_RESULT";
    /** PASSED - / The hardware test passed. */
    DishGetDiagnosticsResponse_TestResult[DishGetDiagnosticsResponse_TestResult["PASSED"] = 1] = "PASSED";
    /** FAILED - / The hardware test failed. Something is wrong. */
    DishGetDiagnosticsResponse_TestResult[DishGetDiagnosticsResponse_TestResult["FAILED"] = 2] = "FAILED";
    DishGetDiagnosticsResponse_TestResult[DishGetDiagnosticsResponse_TestResult["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(DishGetDiagnosticsResponse_TestResult || (exports.DishGetDiagnosticsResponse_TestResult = DishGetDiagnosticsResponse_TestResult = {}));
function dishGetDiagnosticsResponse_TestResultFromJSON(object) {
    switch (object) {
        case 0:
        case "NO_RESULT":
            return DishGetDiagnosticsResponse_TestResult.NO_RESULT;
        case 1:
        case "PASSED":
            return DishGetDiagnosticsResponse_TestResult.PASSED;
        case 2:
        case "FAILED":
            return DishGetDiagnosticsResponse_TestResult.FAILED;
        case -1:
        case "UNRECOGNIZED":
        default:
            return DishGetDiagnosticsResponse_TestResult.UNRECOGNIZED;
    }
}
function dishGetDiagnosticsResponse_TestResultToJSON(object) {
    switch (object) {
        case DishGetDiagnosticsResponse_TestResult.NO_RESULT:
            return "NO_RESULT";
        case DishGetDiagnosticsResponse_TestResult.PASSED:
            return "PASSED";
        case DishGetDiagnosticsResponse_TestResult.FAILED:
            return "FAILED";
        case DishGetDiagnosticsResponse_TestResult.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
/**
 * / The dish may be disabled for a variety of reasons.
 * / This code may be unknown while the dish is loading state
 * / of if the dish does not support this metric.
 */
var DishGetDiagnosticsResponse_DisablementCode;
(function (DishGetDiagnosticsResponse_DisablementCode) {
    /** UNKNOWN - Unknown state. */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["UNKNOWN"] = 0] = "UNKNOWN";
    /** OKAY - / There is no disablement, things are working fine. */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["OKAY"] = 1] = "OKAY";
    /** NO_ACTIVE_ACCOUNT - / No active account, or expired subscription for this user. */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["NO_ACTIVE_ACCOUNT"] = 2] = "NO_ACTIVE_ACCOUNT";
    /** TOO_FAR_FROM_SERVICE_ADDRESS - / User is too far from their service address. */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["TOO_FAR_FROM_SERVICE_ADDRESS"] = 3] = "TOO_FAR_FROM_SERVICE_ADDRESS";
    /** IN_OCEAN - / User is in the Ocean and is not allowed to be. */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["IN_OCEAN"] = 4] = "IN_OCEAN";
    /** INVALID_COUNTRY - / In a country the user isn't allowed to get service in. */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["INVALID_COUNTRY"] = 5] = "INVALID_COUNTRY";
    /** BLOCKED_COUNTRY - / The user is in a country that is blocked for general service. */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["BLOCKED_COUNTRY"] = 6] = "BLOCKED_COUNTRY";
    /** DATA_OVERAGE_SANDBOX_POLICY - / The user has exceeded the data usage quota and sandbox policy has been enforced. */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["DATA_OVERAGE_SANDBOX_POLICY"] = 7] = "DATA_OVERAGE_SANDBOX_POLICY";
    /** CELL_IS_DISABLED - / The user is in a disabled cell */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["CELL_IS_DISABLED"] = 8] = "CELL_IS_DISABLED";
    /** UNLICENSED_COUNTRY - / The user is in an unlicensed country */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["UNLICENSED_COUNTRY"] = 9] = "UNLICENSED_COUNTRY";
    /** ROAM_RESTRICTED - / The user is roam restricted, controlled from GCP via policy feature value */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["ROAM_RESTRICTED"] = 10] = "ROAM_RESTRICTED";
    /** UNKNOWN_LOCATION - / The user's location is unknown to perform geo-restriction */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["UNKNOWN_LOCATION"] = 11] = "UNKNOWN_LOCATION";
    /** ACCOUNT_DISABLED - / The user needs to log in to get more data about their restriction */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["ACCOUNT_DISABLED"] = 12] = "ACCOUNT_DISABLED";
    /** UNSUPPORTED_VERSION - The user's terminal is running an old version that is unsupported on the network */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["UNSUPPORTED_VERSION"] = 13] = "UNSUPPORTED_VERSION";
    /** MOVING_TO_FAST_FOR_POLICY - / The user's terminal is moving faster than their policy allows for service */
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["MOVING_TO_FAST_FOR_POLICY"] = 14] = "MOVING_TO_FAST_FOR_POLICY";
    DishGetDiagnosticsResponse_DisablementCode[DishGetDiagnosticsResponse_DisablementCode["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(DishGetDiagnosticsResponse_DisablementCode || (exports.DishGetDiagnosticsResponse_DisablementCode = DishGetDiagnosticsResponse_DisablementCode = {}));
function dishGetDiagnosticsResponse_DisablementCodeFromJSON(object) {
    switch (object) {
        case 0:
        case "UNKNOWN":
            return DishGetDiagnosticsResponse_DisablementCode.UNKNOWN;
        case 1:
        case "OKAY":
            return DishGetDiagnosticsResponse_DisablementCode.OKAY;
        case 2:
        case "NO_ACTIVE_ACCOUNT":
            return DishGetDiagnosticsResponse_DisablementCode.NO_ACTIVE_ACCOUNT;
        case 3:
        case "TOO_FAR_FROM_SERVICE_ADDRESS":
            return DishGetDiagnosticsResponse_DisablementCode.TOO_FAR_FROM_SERVICE_ADDRESS;
        case 4:
        case "IN_OCEAN":
            return DishGetDiagnosticsResponse_DisablementCode.IN_OCEAN;
        case 5:
        case "INVALID_COUNTRY":
            return DishGetDiagnosticsResponse_DisablementCode.INVALID_COUNTRY;
        case 6:
        case "BLOCKED_COUNTRY":
            return DishGetDiagnosticsResponse_DisablementCode.BLOCKED_COUNTRY;
        case 7:
        case "DATA_OVERAGE_SANDBOX_POLICY":
            return DishGetDiagnosticsResponse_DisablementCode.DATA_OVERAGE_SANDBOX_POLICY;
        case 8:
        case "CELL_IS_DISABLED":
            return DishGetDiagnosticsResponse_DisablementCode.CELL_IS_DISABLED;
        case 9:
        case "UNLICENSED_COUNTRY":
            return DishGetDiagnosticsResponse_DisablementCode.UNLICENSED_COUNTRY;
        case 10:
        case "ROAM_RESTRICTED":
            return DishGetDiagnosticsResponse_DisablementCode.ROAM_RESTRICTED;
        case 11:
        case "UNKNOWN_LOCATION":
            return DishGetDiagnosticsResponse_DisablementCode.UNKNOWN_LOCATION;
        case 12:
        case "ACCOUNT_DISABLED":
            return DishGetDiagnosticsResponse_DisablementCode.ACCOUNT_DISABLED;
        case 13:
        case "UNSUPPORTED_VERSION":
            return DishGetDiagnosticsResponse_DisablementCode.UNSUPPORTED_VERSION;
        case 14:
        case "MOVING_TO_FAST_FOR_POLICY":
            return DishGetDiagnosticsResponse_DisablementCode.MOVING_TO_FAST_FOR_POLICY;
        case -1:
        case "UNRECOGNIZED":
        default:
            return DishGetDiagnosticsResponse_DisablementCode.UNRECOGNIZED;
    }
}
function dishGetDiagnosticsResponse_DisablementCodeToJSON(object) {
    switch (object) {
        case DishGetDiagnosticsResponse_DisablementCode.UNKNOWN:
            return "UNKNOWN";
        case DishGetDiagnosticsResponse_DisablementCode.OKAY:
            return "OKAY";
        case DishGetDiagnosticsResponse_DisablementCode.NO_ACTIVE_ACCOUNT:
            return "NO_ACTIVE_ACCOUNT";
        case DishGetDiagnosticsResponse_DisablementCode.TOO_FAR_FROM_SERVICE_ADDRESS:
            return "TOO_FAR_FROM_SERVICE_ADDRESS";
        case DishGetDiagnosticsResponse_DisablementCode.IN_OCEAN:
            return "IN_OCEAN";
        case DishGetDiagnosticsResponse_DisablementCode.INVALID_COUNTRY:
            return "INVALID_COUNTRY";
        case DishGetDiagnosticsResponse_DisablementCode.BLOCKED_COUNTRY:
            return "BLOCKED_COUNTRY";
        case DishGetDiagnosticsResponse_DisablementCode.DATA_OVERAGE_SANDBOX_POLICY:
            return "DATA_OVERAGE_SANDBOX_POLICY";
        case DishGetDiagnosticsResponse_DisablementCode.CELL_IS_DISABLED:
            return "CELL_IS_DISABLED";
        case DishGetDiagnosticsResponse_DisablementCode.UNLICENSED_COUNTRY:
            return "UNLICENSED_COUNTRY";
        case DishGetDiagnosticsResponse_DisablementCode.ROAM_RESTRICTED:
            return "ROAM_RESTRICTED";
        case DishGetDiagnosticsResponse_DisablementCode.UNKNOWN_LOCATION:
            return "UNKNOWN_LOCATION";
        case DishGetDiagnosticsResponse_DisablementCode.ACCOUNT_DISABLED:
            return "ACCOUNT_DISABLED";
        case DishGetDiagnosticsResponse_DisablementCode.UNSUPPORTED_VERSION:
            return "UNSUPPORTED_VERSION";
        case DishGetDiagnosticsResponse_DisablementCode.MOVING_TO_FAST_FOR_POLICY:
            return "MOVING_TO_FAST_FOR_POLICY";
        case DishGetDiagnosticsResponse_DisablementCode.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
var DishGetDiagnosticsResponse_TestResultCode;
(function (DishGetDiagnosticsResponse_TestResultCode) {
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["GENERAL"] = 0] = "GENERAL";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["BOOT_UP"] = 1] = "BOOT_UP";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["CPU_VOLTAGE"] = 2] = "CPU_VOLTAGE";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["DBF_AAP_CS"] = 3] = "DBF_AAP_CS";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["DBF_NUM_FEMS"] = 4] = "DBF_NUM_FEMS";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["DBF_READ_ERRORS"] = 5] = "DBF_READ_ERRORS";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["DBF_T_DIE_0"] = 6] = "DBF_T_DIE_0";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["DBF_T_DIE_1"] = 7] = "DBF_T_DIE_1";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["DBF_T_DIE_0_VALID"] = 8] = "DBF_T_DIE_0_VALID";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["DBF_T_DIE_1_VALID"] = 9] = "DBF_T_DIE_1_VALID";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["ETH_PRIME"] = 10] = "ETH_PRIME";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["EIRP"] = 11] = "EIRP";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["FEM_CUT"] = 12] = "FEM_CUT";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["FUSE_AVS"] = 13] = "FUSE_AVS";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["GPS"] = 14] = "GPS";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["IMU"] = 15] = "IMU";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["PHY"] = 16] = "PHY";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["SCP_ERROR"] = 17] = "SCP_ERROR";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["TEMPERATURE"] = 18] = "TEMPERATURE";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["VTSENS"] = 19] = "VTSENS";
    DishGetDiagnosticsResponse_TestResultCode[DishGetDiagnosticsResponse_TestResultCode["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(DishGetDiagnosticsResponse_TestResultCode || (exports.DishGetDiagnosticsResponse_TestResultCode = DishGetDiagnosticsResponse_TestResultCode = {}));
function dishGetDiagnosticsResponse_TestResultCodeFromJSON(object) {
    switch (object) {
        case 0:
        case "GENERAL":
            return DishGetDiagnosticsResponse_TestResultCode.GENERAL;
        case 1:
        case "BOOT_UP":
            return DishGetDiagnosticsResponse_TestResultCode.BOOT_UP;
        case 2:
        case "CPU_VOLTAGE":
            return DishGetDiagnosticsResponse_TestResultCode.CPU_VOLTAGE;
        case 3:
        case "DBF_AAP_CS":
            return DishGetDiagnosticsResponse_TestResultCode.DBF_AAP_CS;
        case 4:
        case "DBF_NUM_FEMS":
            return DishGetDiagnosticsResponse_TestResultCode.DBF_NUM_FEMS;
        case 5:
        case "DBF_READ_ERRORS":
            return DishGetDiagnosticsResponse_TestResultCode.DBF_READ_ERRORS;
        case 6:
        case "DBF_T_DIE_0":
            return DishGetDiagnosticsResponse_TestResultCode.DBF_T_DIE_0;
        case 7:
        case "DBF_T_DIE_1":
            return DishGetDiagnosticsResponse_TestResultCode.DBF_T_DIE_1;
        case 8:
        case "DBF_T_DIE_0_VALID":
            return DishGetDiagnosticsResponse_TestResultCode.DBF_T_DIE_0_VALID;
        case 9:
        case "DBF_T_DIE_1_VALID":
            return DishGetDiagnosticsResponse_TestResultCode.DBF_T_DIE_1_VALID;
        case 10:
        case "ETH_PRIME":
            return DishGetDiagnosticsResponse_TestResultCode.ETH_PRIME;
        case 11:
        case "EIRP":
            return DishGetDiagnosticsResponse_TestResultCode.EIRP;
        case 12:
        case "FEM_CUT":
            return DishGetDiagnosticsResponse_TestResultCode.FEM_CUT;
        case 13:
        case "FUSE_AVS":
            return DishGetDiagnosticsResponse_TestResultCode.FUSE_AVS;
        case 14:
        case "GPS":
            return DishGetDiagnosticsResponse_TestResultCode.GPS;
        case 15:
        case "IMU":
            return DishGetDiagnosticsResponse_TestResultCode.IMU;
        case 16:
        case "PHY":
            return DishGetDiagnosticsResponse_TestResultCode.PHY;
        case 17:
        case "SCP_ERROR":
            return DishGetDiagnosticsResponse_TestResultCode.SCP_ERROR;
        case 18:
        case "TEMPERATURE":
            return DishGetDiagnosticsResponse_TestResultCode.TEMPERATURE;
        case 19:
        case "VTSENS":
            return DishGetDiagnosticsResponse_TestResultCode.VTSENS;
        case -1:
        case "UNRECOGNIZED":
        default:
            return DishGetDiagnosticsResponse_TestResultCode.UNRECOGNIZED;
    }
}
function dishGetDiagnosticsResponse_TestResultCodeToJSON(object) {
    switch (object) {
        case DishGetDiagnosticsResponse_TestResultCode.GENERAL:
            return "GENERAL";
        case DishGetDiagnosticsResponse_TestResultCode.BOOT_UP:
            return "BOOT_UP";
        case DishGetDiagnosticsResponse_TestResultCode.CPU_VOLTAGE:
            return "CPU_VOLTAGE";
        case DishGetDiagnosticsResponse_TestResultCode.DBF_AAP_CS:
            return "DBF_AAP_CS";
        case DishGetDiagnosticsResponse_TestResultCode.DBF_NUM_FEMS:
            return "DBF_NUM_FEMS";
        case DishGetDiagnosticsResponse_TestResultCode.DBF_READ_ERRORS:
            return "DBF_READ_ERRORS";
        case DishGetDiagnosticsResponse_TestResultCode.DBF_T_DIE_0:
            return "DBF_T_DIE_0";
        case DishGetDiagnosticsResponse_TestResultCode.DBF_T_DIE_1:
            return "DBF_T_DIE_1";
        case DishGetDiagnosticsResponse_TestResultCode.DBF_T_DIE_0_VALID:
            return "DBF_T_DIE_0_VALID";
        case DishGetDiagnosticsResponse_TestResultCode.DBF_T_DIE_1_VALID:
            return "DBF_T_DIE_1_VALID";
        case DishGetDiagnosticsResponse_TestResultCode.ETH_PRIME:
            return "ETH_PRIME";
        case DishGetDiagnosticsResponse_TestResultCode.EIRP:
            return "EIRP";
        case DishGetDiagnosticsResponse_TestResultCode.FEM_CUT:
            return "FEM_CUT";
        case DishGetDiagnosticsResponse_TestResultCode.FUSE_AVS:
            return "FUSE_AVS";
        case DishGetDiagnosticsResponse_TestResultCode.GPS:
            return "GPS";
        case DishGetDiagnosticsResponse_TestResultCode.IMU:
            return "IMU";
        case DishGetDiagnosticsResponse_TestResultCode.PHY:
            return "PHY";
        case DishGetDiagnosticsResponse_TestResultCode.SCP_ERROR:
            return "SCP_ERROR";
        case DishGetDiagnosticsResponse_TestResultCode.TEMPERATURE:
            return "TEMPERATURE";
        case DishGetDiagnosticsResponse_TestResultCode.VTSENS:
            return "VTSENS";
        case DishGetDiagnosticsResponse_TestResultCode.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBaseDishStowRequest() {
    return { unstow: undefined };
}
exports.DishStowRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.unstow !== undefined) {
            writer.uint32(8).bool(message.unstow);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishStowRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.unstow = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { unstow: isSet(object.unstow) ? globalThis.Boolean(object.unstow) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.unstow !== undefined) {
            obj.unstow = message.unstow;
        }
        return obj;
    },
    create(base) {
        return exports.DishStowRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseDishStowRequest();
        message.unstow = (_a = object.unstow) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBaseDishStowResponse() {
    return {};
}
exports.DishStowResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishStowResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.DishStowResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseDishStowResponse();
        return message;
    },
};
function createBaseDishGetContextRequest() {
    return {};
}
exports.DishGetContextRequest = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishGetContextRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.DishGetContextRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseDishGetContextRequest();
        return message;
    },
};
function createBaseDishGetContextResponse() {
    return {
        deviceInfo: undefined,
        deviceState: undefined,
        obstructionFraction: undefined,
        obstructionValidS: undefined,
        cellId: undefined,
        popRackId: undefined,
        initialSatelliteId: undefined,
        initialGatewayId: undefined,
        onBackupBeam: undefined,
        secondsToSlotEnd: undefined,
    };
}
exports.DishGetContextResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.deviceInfo !== undefined) {
            common_1.DeviceInfo.encode(message.deviceInfo, writer.uint32(10).fork()).join();
        }
        if (message.deviceState !== undefined) {
            common_1.DeviceState.encode(message.deviceState, writer.uint32(58).fork()).join();
        }
        if (message.obstructionFraction !== undefined) {
            writer.uint32(21).float(message.obstructionFraction);
        }
        if (message.obstructionValidS !== undefined) {
            writer.uint32(29).float(message.obstructionValidS);
        }
        if (message.cellId !== undefined) {
            writer.uint32(32).uint32(message.cellId);
        }
        if (message.popRackId !== undefined) {
            writer.uint32(40).uint32(message.popRackId);
        }
        if (message.initialSatelliteId !== undefined) {
            writer.uint32(64).uint32(message.initialSatelliteId);
        }
        if (message.initialGatewayId !== undefined) {
            writer.uint32(72).uint32(message.initialGatewayId);
        }
        if (message.onBackupBeam !== undefined) {
            writer.uint32(80).bool(message.onBackupBeam);
        }
        if (message.secondsToSlotEnd !== undefined) {
            writer.uint32(53).float(message.secondsToSlotEnd);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishGetContextResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.deviceInfo = common_1.DeviceInfo.decode(reader, reader.uint32());
                    continue;
                }
                case 7: {
                    if (tag !== 58) {
                        break;
                    }
                    message.deviceState = common_1.DeviceState.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 21) {
                        break;
                    }
                    message.obstructionFraction = reader.float();
                    continue;
                }
                case 3: {
                    if (tag !== 29) {
                        break;
                    }
                    message.obstructionValidS = reader.float();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.cellId = reader.uint32();
                    continue;
                }
                case 5: {
                    if (tag !== 40) {
                        break;
                    }
                    message.popRackId = reader.uint32();
                    continue;
                }
                case 8: {
                    if (tag !== 64) {
                        break;
                    }
                    message.initialSatelliteId = reader.uint32();
                    continue;
                }
                case 9: {
                    if (tag !== 72) {
                        break;
                    }
                    message.initialGatewayId = reader.uint32();
                    continue;
                }
                case 10: {
                    if (tag !== 80) {
                        break;
                    }
                    message.onBackupBeam = reader.bool();
                    continue;
                }
                case 6: {
                    if (tag !== 53) {
                        break;
                    }
                    message.secondsToSlotEnd = reader.float();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            deviceInfo: isSet(object.deviceInfo)
                ? common_1.DeviceInfo.fromJSON(object.deviceInfo)
                : isSet(object.device_info)
                    ? common_1.DeviceInfo.fromJSON(object.device_info)
                    : undefined,
            deviceState: isSet(object.deviceState)
                ? common_1.DeviceState.fromJSON(object.deviceState)
                : isSet(object.device_state)
                    ? common_1.DeviceState.fromJSON(object.device_state)
                    : undefined,
            obstructionFraction: isSet(object.obstructionFraction)
                ? globalThis.Number(object.obstructionFraction)
                : isSet(object.obstruction_fraction)
                    ? globalThis.Number(object.obstruction_fraction)
                    : undefined,
            obstructionValidS: isSet(object.obstructionValidS)
                ? globalThis.Number(object.obstructionValidS)
                : isSet(object.obstruction_valid_s)
                    ? globalThis.Number(object.obstruction_valid_s)
                    : undefined,
            cellId: isSet(object.cellId)
                ? globalThis.Number(object.cellId)
                : isSet(object.cell_id)
                    ? globalThis.Number(object.cell_id)
                    : undefined,
            popRackId: isSet(object.popRackId)
                ? globalThis.Number(object.popRackId)
                : isSet(object.pop_rack_id)
                    ? globalThis.Number(object.pop_rack_id)
                    : undefined,
            initialSatelliteId: isSet(object.initialSatelliteId)
                ? globalThis.Number(object.initialSatelliteId)
                : isSet(object.initial_satellite_id)
                    ? globalThis.Number(object.initial_satellite_id)
                    : undefined,
            initialGatewayId: isSet(object.initialGatewayId)
                ? globalThis.Number(object.initialGatewayId)
                : isSet(object.initial_gateway_id)
                    ? globalThis.Number(object.initial_gateway_id)
                    : undefined,
            onBackupBeam: isSet(object.onBackupBeam)
                ? globalThis.Boolean(object.onBackupBeam)
                : isSet(object.on_backup_beam)
                    ? globalThis.Boolean(object.on_backup_beam)
                    : undefined,
            secondsToSlotEnd: isSet(object.secondsToSlotEnd)
                ? globalThis.Number(object.secondsToSlotEnd)
                : isSet(object.seconds_to_slot_end)
                    ? globalThis.Number(object.seconds_to_slot_end)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.deviceInfo !== undefined) {
            obj.deviceInfo = common_1.DeviceInfo.toJSON(message.deviceInfo);
        }
        if (message.deviceState !== undefined) {
            obj.deviceState = common_1.DeviceState.toJSON(message.deviceState);
        }
        if (message.obstructionFraction !== undefined) {
            obj.obstructionFraction = message.obstructionFraction;
        }
        if (message.obstructionValidS !== undefined) {
            obj.obstructionValidS = message.obstructionValidS;
        }
        if (message.cellId !== undefined) {
            obj.cellId = Math.round(message.cellId);
        }
        if (message.popRackId !== undefined) {
            obj.popRackId = Math.round(message.popRackId);
        }
        if (message.initialSatelliteId !== undefined) {
            obj.initialSatelliteId = Math.round(message.initialSatelliteId);
        }
        if (message.initialGatewayId !== undefined) {
            obj.initialGatewayId = Math.round(message.initialGatewayId);
        }
        if (message.onBackupBeam !== undefined) {
            obj.onBackupBeam = message.onBackupBeam;
        }
        if (message.secondsToSlotEnd !== undefined) {
            obj.secondsToSlotEnd = message.secondsToSlotEnd;
        }
        return obj;
    },
    create(base) {
        return exports.DishGetContextResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const message = createBaseDishGetContextResponse();
        message.deviceInfo = (object.deviceInfo !== undefined && object.deviceInfo !== null)
            ? common_1.DeviceInfo.fromPartial(object.deviceInfo)
            : undefined;
        message.deviceState = (object.deviceState !== undefined && object.deviceState !== null)
            ? common_1.DeviceState.fromPartial(object.deviceState)
            : undefined;
        message.obstructionFraction = (_a = object.obstructionFraction) !== null && _a !== void 0 ? _a : undefined;
        message.obstructionValidS = (_b = object.obstructionValidS) !== null && _b !== void 0 ? _b : undefined;
        message.cellId = (_c = object.cellId) !== null && _c !== void 0 ? _c : undefined;
        message.popRackId = (_d = object.popRackId) !== null && _d !== void 0 ? _d : undefined;
        message.initialSatelliteId = (_e = object.initialSatelliteId) !== null && _e !== void 0 ? _e : undefined;
        message.initialGatewayId = (_f = object.initialGatewayId) !== null && _f !== void 0 ? _f : undefined;
        message.onBackupBeam = (_g = object.onBackupBeam) !== null && _g !== void 0 ? _g : undefined;
        message.secondsToSlotEnd = (_h = object.secondsToSlotEnd) !== null && _h !== void 0 ? _h : undefined;
        return message;
    },
};
function createBaseDishGetHistoryResponse() {
    return {
        current: undefined,
        popPingDropRate: [],
        popPingLatencyMs: [],
        downlinkThroughputBps: [],
        uplinkThroughputBps: [],
        snr: [],
        scheduled: [],
        obstructed: [],
    };
}
exports.DishGetHistoryResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.current !== undefined) {
            writer.uint32(8).uint64(message.current);
        }
        writer.uint32(8010).fork();
        for (const v of message.popPingDropRate) {
            writer.float(v);
        }
        writer.join();
        writer.uint32(8018).fork();
        for (const v of message.popPingLatencyMs) {
            writer.float(v);
        }
        writer.join();
        writer.uint32(8026).fork();
        for (const v of message.downlinkThroughputBps) {
            writer.float(v);
        }
        writer.join();
        writer.uint32(8034).fork();
        for (const v of message.uplinkThroughputBps) {
            writer.float(v);
        }
        writer.join();
        writer.uint32(8042).fork();
        for (const v of message.snr) {
            writer.float(v);
        }
        writer.join();
        writer.uint32(8050).fork();
        for (const v of message.scheduled) {
            writer.bool(v);
        }
        writer.join();
        writer.uint32(8058).fork();
        for (const v of message.obstructed) {
            writer.bool(v);
        }
        writer.join();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishGetHistoryResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.current = longToNumber(reader.uint64());
                    continue;
                }
                case 1001: {
                    if (tag === 8013) {
                        message.popPingDropRate.push(reader.float());
                        continue;
                    }
                    if (tag === 8010) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.popPingDropRate.push(reader.float());
                        }
                        continue;
                    }
                    break;
                }
                case 1002: {
                    if (tag === 8021) {
                        message.popPingLatencyMs.push(reader.float());
                        continue;
                    }
                    if (tag === 8018) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.popPingLatencyMs.push(reader.float());
                        }
                        continue;
                    }
                    break;
                }
                case 1003: {
                    if (tag === 8029) {
                        message.downlinkThroughputBps.push(reader.float());
                        continue;
                    }
                    if (tag === 8026) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.downlinkThroughputBps.push(reader.float());
                        }
                        continue;
                    }
                    break;
                }
                case 1004: {
                    if (tag === 8037) {
                        message.uplinkThroughputBps.push(reader.float());
                        continue;
                    }
                    if (tag === 8034) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.uplinkThroughputBps.push(reader.float());
                        }
                        continue;
                    }
                    break;
                }
                case 1005: {
                    if (tag === 8045) {
                        message.snr.push(reader.float());
                        continue;
                    }
                    if (tag === 8042) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.snr.push(reader.float());
                        }
                        continue;
                    }
                    break;
                }
                case 1006: {
                    if (tag === 8048) {
                        message.scheduled.push(reader.bool());
                        continue;
                    }
                    if (tag === 8050) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.scheduled.push(reader.bool());
                        }
                        continue;
                    }
                    break;
                }
                case 1007: {
                    if (tag === 8056) {
                        message.obstructed.push(reader.bool());
                        continue;
                    }
                    if (tag === 8058) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.obstructed.push(reader.bool());
                        }
                        continue;
                    }
                    break;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            current: isSet(object.current) ? globalThis.Number(object.current) : undefined,
            popPingDropRate: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.popPingDropRate)
                ? object.popPingDropRate.map((e) => globalThis.Number(e))
                : globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.pop_ping_drop_rate)
                    ? object.pop_ping_drop_rate.map((e) => globalThis.Number(e))
                    : [],
            popPingLatencyMs: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.popPingLatencyMs)
                ? object.popPingLatencyMs.map((e) => globalThis.Number(e))
                : globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.pop_ping_latency_ms)
                    ? object.pop_ping_latency_ms.map((e) => globalThis.Number(e))
                    : [],
            downlinkThroughputBps: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.downlinkThroughputBps)
                ? object.downlinkThroughputBps.map((e) => globalThis.Number(e))
                : globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.downlink_throughput_bps)
                    ? object.downlink_throughput_bps.map((e) => globalThis.Number(e))
                    : [],
            uplinkThroughputBps: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.uplinkThroughputBps)
                ? object.uplinkThroughputBps.map((e) => globalThis.Number(e))
                : globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.uplink_throughput_bps)
                    ? object.uplink_throughput_bps.map((e) => globalThis.Number(e))
                    : [],
            snr: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.snr)
                ? object.snr.map((e) => globalThis.Number(e))
                : [],
            scheduled: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.scheduled)
                ? object.scheduled.map((e) => globalThis.Boolean(e))
                : [],
            obstructed: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.obstructed)
                ? object.obstructed.map((e) => globalThis.Boolean(e))
                : [],
        };
    },
    toJSON(message) {
        var _a, _b, _c, _d, _e, _f, _g;
        const obj = {};
        if (message.current !== undefined) {
            obj.current = Math.round(message.current);
        }
        if ((_a = message.popPingDropRate) === null || _a === void 0 ? void 0 : _a.length) {
            obj.popPingDropRate = message.popPingDropRate;
        }
        if ((_b = message.popPingLatencyMs) === null || _b === void 0 ? void 0 : _b.length) {
            obj.popPingLatencyMs = message.popPingLatencyMs;
        }
        if ((_c = message.downlinkThroughputBps) === null || _c === void 0 ? void 0 : _c.length) {
            obj.downlinkThroughputBps = message.downlinkThroughputBps;
        }
        if ((_d = message.uplinkThroughputBps) === null || _d === void 0 ? void 0 : _d.length) {
            obj.uplinkThroughputBps = message.uplinkThroughputBps;
        }
        if ((_e = message.snr) === null || _e === void 0 ? void 0 : _e.length) {
            obj.snr = message.snr;
        }
        if ((_f = message.scheduled) === null || _f === void 0 ? void 0 : _f.length) {
            obj.scheduled = message.scheduled;
        }
        if ((_g = message.obstructed) === null || _g === void 0 ? void 0 : _g.length) {
            obj.obstructed = message.obstructed;
        }
        return obj;
    },
    create(base) {
        return exports.DishGetHistoryResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const message = createBaseDishGetHistoryResponse();
        message.current = (_a = object.current) !== null && _a !== void 0 ? _a : undefined;
        message.popPingDropRate = ((_b = object.popPingDropRate) === null || _b === void 0 ? void 0 : _b.map((e) => e)) || [];
        message.popPingLatencyMs = ((_c = object.popPingLatencyMs) === null || _c === void 0 ? void 0 : _c.map((e) => e)) || [];
        message.downlinkThroughputBps = ((_d = object.downlinkThroughputBps) === null || _d === void 0 ? void 0 : _d.map((e) => e)) || [];
        message.uplinkThroughputBps = ((_e = object.uplinkThroughputBps) === null || _e === void 0 ? void 0 : _e.map((e) => e)) || [];
        message.snr = ((_f = object.snr) === null || _f === void 0 ? void 0 : _f.map((e) => e)) || [];
        message.scheduled = ((_g = object.scheduled) === null || _g === void 0 ? void 0 : _g.map((e) => e)) || [];
        message.obstructed = ((_h = object.obstructed) === null || _h === void 0 ? void 0 : _h.map((e) => e)) || [];
        return message;
    },
};
function createBaseDishGetStatusResponse() {
    return {
        deviceInfo: undefined,
        deviceState: undefined,
        state: undefined,
        alerts: undefined,
        snr: undefined,
        secondsToFirstNonemptySlot: undefined,
        popPingDropRate: undefined,
        downlinkThroughputBps: undefined,
        uplinkThroughputBps: undefined,
        popPingLatencyMs: undefined,
        obstructionStats: undefined,
        stowRequested: undefined,
    };
}
exports.DishGetStatusResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.deviceInfo !== undefined) {
            common_1.DeviceInfo.encode(message.deviceInfo, writer.uint32(10).fork()).join();
        }
        if (message.deviceState !== undefined) {
            common_1.DeviceState.encode(message.deviceState, writer.uint32(18).fork()).join();
        }
        if (message.state !== undefined) {
            writer.uint32(8048).int32(message.state);
        }
        if (message.alerts !== undefined) {
            exports.DishAlerts.encode(message.alerts, writer.uint32(8042).fork()).join();
        }
        if (message.snr !== undefined) {
            writer.uint32(8013).float(message.snr);
        }
        if (message.secondsToFirstNonemptySlot !== undefined) {
            writer.uint32(8021).float(message.secondsToFirstNonemptySlot);
        }
        if (message.popPingDropRate !== undefined) {
            writer.uint32(8029).float(message.popPingDropRate);
        }
        if (message.downlinkThroughputBps !== undefined) {
            writer.uint32(8061).float(message.downlinkThroughputBps);
        }
        if (message.uplinkThroughputBps !== undefined) {
            writer.uint32(8069).float(message.uplinkThroughputBps);
        }
        if (message.popPingLatencyMs !== undefined) {
            writer.uint32(8077).float(message.popPingLatencyMs);
        }
        if (message.obstructionStats !== undefined) {
            exports.DishObstructionStats.encode(message.obstructionStats, writer.uint32(8034).fork()).join();
        }
        if (message.stowRequested !== undefined) {
            writer.uint32(8080).bool(message.stowRequested);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishGetStatusResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.deviceInfo = common_1.DeviceInfo.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.deviceState = common_1.DeviceState.decode(reader, reader.uint32());
                    continue;
                }
                case 1006: {
                    if (tag !== 8048) {
                        break;
                    }
                    message.state = reader.int32();
                    continue;
                }
                case 1005: {
                    if (tag !== 8042) {
                        break;
                    }
                    message.alerts = exports.DishAlerts.decode(reader, reader.uint32());
                    continue;
                }
                case 1001: {
                    if (tag !== 8013) {
                        break;
                    }
                    message.snr = reader.float();
                    continue;
                }
                case 1002: {
                    if (tag !== 8021) {
                        break;
                    }
                    message.secondsToFirstNonemptySlot = reader.float();
                    continue;
                }
                case 1003: {
                    if (tag !== 8029) {
                        break;
                    }
                    message.popPingDropRate = reader.float();
                    continue;
                }
                case 1007: {
                    if (tag !== 8061) {
                        break;
                    }
                    message.downlinkThroughputBps = reader.float();
                    continue;
                }
                case 1008: {
                    if (tag !== 8069) {
                        break;
                    }
                    message.uplinkThroughputBps = reader.float();
                    continue;
                }
                case 1009: {
                    if (tag !== 8077) {
                        break;
                    }
                    message.popPingLatencyMs = reader.float();
                    continue;
                }
                case 1004: {
                    if (tag !== 8034) {
                        break;
                    }
                    message.obstructionStats = exports.DishObstructionStats.decode(reader, reader.uint32());
                    continue;
                }
                case 1010: {
                    if (tag !== 8080) {
                        break;
                    }
                    message.stowRequested = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            deviceInfo: isSet(object.deviceInfo)
                ? common_1.DeviceInfo.fromJSON(object.deviceInfo)
                : isSet(object.device_info)
                    ? common_1.DeviceInfo.fromJSON(object.device_info)
                    : undefined,
            deviceState: isSet(object.deviceState)
                ? common_1.DeviceState.fromJSON(object.deviceState)
                : isSet(object.device_state)
                    ? common_1.DeviceState.fromJSON(object.device_state)
                    : undefined,
            state: isSet(object.state) ? dishStateFromJSON(object.state) : undefined,
            alerts: isSet(object.alerts) ? exports.DishAlerts.fromJSON(object.alerts) : undefined,
            snr: isSet(object.snr) ? globalThis.Number(object.snr) : undefined,
            secondsToFirstNonemptySlot: isSet(object.secondsToFirstNonemptySlot)
                ? globalThis.Number(object.secondsToFirstNonemptySlot)
                : isSet(object.seconds_to_first_nonempty_slot)
                    ? globalThis.Number(object.seconds_to_first_nonempty_slot)
                    : undefined,
            popPingDropRate: isSet(object.popPingDropRate)
                ? globalThis.Number(object.popPingDropRate)
                : isSet(object.pop_ping_drop_rate)
                    ? globalThis.Number(object.pop_ping_drop_rate)
                    : undefined,
            downlinkThroughputBps: isSet(object.downlinkThroughputBps)
                ? globalThis.Number(object.downlinkThroughputBps)
                : isSet(object.downlink_throughput_bps)
                    ? globalThis.Number(object.downlink_throughput_bps)
                    : undefined,
            uplinkThroughputBps: isSet(object.uplinkThroughputBps)
                ? globalThis.Number(object.uplinkThroughputBps)
                : isSet(object.uplink_throughput_bps)
                    ? globalThis.Number(object.uplink_throughput_bps)
                    : undefined,
            popPingLatencyMs: isSet(object.popPingLatencyMs)
                ? globalThis.Number(object.popPingLatencyMs)
                : isSet(object.pop_ping_latency_ms)
                    ? globalThis.Number(object.pop_ping_latency_ms)
                    : undefined,
            obstructionStats: isSet(object.obstructionStats)
                ? exports.DishObstructionStats.fromJSON(object.obstructionStats)
                : isSet(object.obstruction_stats)
                    ? exports.DishObstructionStats.fromJSON(object.obstruction_stats)
                    : undefined,
            stowRequested: isSet(object.stowRequested)
                ? globalThis.Boolean(object.stowRequested)
                : isSet(object.stow_requested)
                    ? globalThis.Boolean(object.stow_requested)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.deviceInfo !== undefined) {
            obj.deviceInfo = common_1.DeviceInfo.toJSON(message.deviceInfo);
        }
        if (message.deviceState !== undefined) {
            obj.deviceState = common_1.DeviceState.toJSON(message.deviceState);
        }
        if (message.state !== undefined) {
            obj.state = dishStateToJSON(message.state);
        }
        if (message.alerts !== undefined) {
            obj.alerts = exports.DishAlerts.toJSON(message.alerts);
        }
        if (message.snr !== undefined) {
            obj.snr = message.snr;
        }
        if (message.secondsToFirstNonemptySlot !== undefined) {
            obj.secondsToFirstNonemptySlot = message.secondsToFirstNonemptySlot;
        }
        if (message.popPingDropRate !== undefined) {
            obj.popPingDropRate = message.popPingDropRate;
        }
        if (message.downlinkThroughputBps !== undefined) {
            obj.downlinkThroughputBps = message.downlinkThroughputBps;
        }
        if (message.uplinkThroughputBps !== undefined) {
            obj.uplinkThroughputBps = message.uplinkThroughputBps;
        }
        if (message.popPingLatencyMs !== undefined) {
            obj.popPingLatencyMs = message.popPingLatencyMs;
        }
        if (message.obstructionStats !== undefined) {
            obj.obstructionStats = exports.DishObstructionStats.toJSON(message.obstructionStats);
        }
        if (message.stowRequested !== undefined) {
            obj.stowRequested = message.stowRequested;
        }
        return obj;
    },
    create(base) {
        return exports.DishGetStatusResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const message = createBaseDishGetStatusResponse();
        message.deviceInfo = (object.deviceInfo !== undefined && object.deviceInfo !== null)
            ? common_1.DeviceInfo.fromPartial(object.deviceInfo)
            : undefined;
        message.deviceState = (object.deviceState !== undefined && object.deviceState !== null)
            ? common_1.DeviceState.fromPartial(object.deviceState)
            : undefined;
        message.state = (_a = object.state) !== null && _a !== void 0 ? _a : undefined;
        message.alerts = (object.alerts !== undefined && object.alerts !== null)
            ? exports.DishAlerts.fromPartial(object.alerts)
            : undefined;
        message.snr = (_b = object.snr) !== null && _b !== void 0 ? _b : undefined;
        message.secondsToFirstNonemptySlot = (_c = object.secondsToFirstNonemptySlot) !== null && _c !== void 0 ? _c : undefined;
        message.popPingDropRate = (_d = object.popPingDropRate) !== null && _d !== void 0 ? _d : undefined;
        message.downlinkThroughputBps = (_e = object.downlinkThroughputBps) !== null && _e !== void 0 ? _e : undefined;
        message.uplinkThroughputBps = (_f = object.uplinkThroughputBps) !== null && _f !== void 0 ? _f : undefined;
        message.popPingLatencyMs = (_g = object.popPingLatencyMs) !== null && _g !== void 0 ? _g : undefined;
        message.obstructionStats = (object.obstructionStats !== undefined && object.obstructionStats !== null)
            ? exports.DishObstructionStats.fromPartial(object.obstructionStats)
            : undefined;
        message.stowRequested = (_h = object.stowRequested) !== null && _h !== void 0 ? _h : undefined;
        return message;
    },
};
function createBaseDishGetObstructionMapRequest() {
    return {};
}
exports.DishGetObstructionMapRequest = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishGetObstructionMapRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.DishGetObstructionMapRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseDishGetObstructionMapRequest();
        return message;
    },
};
function createBaseDishGetObstructionMapResponse() {
    return { numRows: undefined, numCols: undefined, snr: [] };
}
exports.DishGetObstructionMapResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.numRows !== undefined) {
            writer.uint32(8).uint32(message.numRows);
        }
        if (message.numCols !== undefined) {
            writer.uint32(16).uint32(message.numCols);
        }
        writer.uint32(26).fork();
        for (const v of message.snr) {
            writer.float(v);
        }
        writer.join();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishGetObstructionMapResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.numRows = reader.uint32();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.numCols = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag === 29) {
                        message.snr.push(reader.float());
                        continue;
                    }
                    if (tag === 26) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.snr.push(reader.float());
                        }
                        continue;
                    }
                    break;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            numRows: isSet(object.numRows)
                ? globalThis.Number(object.numRows)
                : isSet(object.num_rows)
                    ? globalThis.Number(object.num_rows)
                    : undefined,
            numCols: isSet(object.numCols)
                ? globalThis.Number(object.numCols)
                : isSet(object.num_cols)
                    ? globalThis.Number(object.num_cols)
                    : undefined,
            snr: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.snr) ? object.snr.map((e) => globalThis.Number(e)) : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.numRows !== undefined) {
            obj.numRows = Math.round(message.numRows);
        }
        if (message.numCols !== undefined) {
            obj.numCols = Math.round(message.numCols);
        }
        if ((_a = message.snr) === null || _a === void 0 ? void 0 : _a.length) {
            obj.snr = message.snr;
        }
        return obj;
    },
    create(base) {
        return exports.DishGetObstructionMapResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseDishGetObstructionMapResponse();
        message.numRows = (_a = object.numRows) !== null && _a !== void 0 ? _a : undefined;
        message.numCols = (_b = object.numCols) !== null && _b !== void 0 ? _b : undefined;
        message.snr = ((_c = object.snr) === null || _c === void 0 ? void 0 : _c.map((e) => e)) || [];
        return message;
    },
};
function createBaseDishAlerts() {
    return {
        motorsStuck: undefined,
        thermalThrottle: undefined,
        thermalShutdown: undefined,
        mastNotNearVertical: undefined,
        unexpectedLocation: undefined,
        slowEthernetSpeeds: undefined,
    };
}
exports.DishAlerts = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.motorsStuck !== undefined) {
            writer.uint32(8).bool(message.motorsStuck);
        }
        if (message.thermalThrottle !== undefined) {
            writer.uint32(24).bool(message.thermalThrottle);
        }
        if (message.thermalShutdown !== undefined) {
            writer.uint32(16).bool(message.thermalShutdown);
        }
        if (message.mastNotNearVertical !== undefined) {
            writer.uint32(40).bool(message.mastNotNearVertical);
        }
        if (message.unexpectedLocation !== undefined) {
            writer.uint32(32).bool(message.unexpectedLocation);
        }
        if (message.slowEthernetSpeeds !== undefined) {
            writer.uint32(48).bool(message.slowEthernetSpeeds);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishAlerts();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.motorsStuck = reader.bool();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.thermalThrottle = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.thermalShutdown = reader.bool();
                    continue;
                }
                case 5: {
                    if (tag !== 40) {
                        break;
                    }
                    message.mastNotNearVertical = reader.bool();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.unexpectedLocation = reader.bool();
                    continue;
                }
                case 6: {
                    if (tag !== 48) {
                        break;
                    }
                    message.slowEthernetSpeeds = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            motorsStuck: isSet(object.motorsStuck)
                ? globalThis.Boolean(object.motorsStuck)
                : isSet(object.motors_stuck)
                    ? globalThis.Boolean(object.motors_stuck)
                    : undefined,
            thermalThrottle: isSet(object.thermalThrottle)
                ? globalThis.Boolean(object.thermalThrottle)
                : isSet(object.thermal_throttle)
                    ? globalThis.Boolean(object.thermal_throttle)
                    : undefined,
            thermalShutdown: isSet(object.thermalShutdown)
                ? globalThis.Boolean(object.thermalShutdown)
                : isSet(object.thermal_shutdown)
                    ? globalThis.Boolean(object.thermal_shutdown)
                    : undefined,
            mastNotNearVertical: isSet(object.mastNotNearVertical)
                ? globalThis.Boolean(object.mastNotNearVertical)
                : isSet(object.mast_not_near_vertical)
                    ? globalThis.Boolean(object.mast_not_near_vertical)
                    : undefined,
            unexpectedLocation: isSet(object.unexpectedLocation)
                ? globalThis.Boolean(object.unexpectedLocation)
                : isSet(object.unexpected_location)
                    ? globalThis.Boolean(object.unexpected_location)
                    : undefined,
            slowEthernetSpeeds: isSet(object.slowEthernetSpeeds)
                ? globalThis.Boolean(object.slowEthernetSpeeds)
                : isSet(object.slow_ethernet_speeds)
                    ? globalThis.Boolean(object.slow_ethernet_speeds)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.motorsStuck !== undefined) {
            obj.motorsStuck = message.motorsStuck;
        }
        if (message.thermalThrottle !== undefined) {
            obj.thermalThrottle = message.thermalThrottle;
        }
        if (message.thermalShutdown !== undefined) {
            obj.thermalShutdown = message.thermalShutdown;
        }
        if (message.mastNotNearVertical !== undefined) {
            obj.mastNotNearVertical = message.mastNotNearVertical;
        }
        if (message.unexpectedLocation !== undefined) {
            obj.unexpectedLocation = message.unexpectedLocation;
        }
        if (message.slowEthernetSpeeds !== undefined) {
            obj.slowEthernetSpeeds = message.slowEthernetSpeeds;
        }
        return obj;
    },
    create(base) {
        return exports.DishAlerts.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f;
        const message = createBaseDishAlerts();
        message.motorsStuck = (_a = object.motorsStuck) !== null && _a !== void 0 ? _a : undefined;
        message.thermalThrottle = (_b = object.thermalThrottle) !== null && _b !== void 0 ? _b : undefined;
        message.thermalShutdown = (_c = object.thermalShutdown) !== null && _c !== void 0 ? _c : undefined;
        message.mastNotNearVertical = (_d = object.mastNotNearVertical) !== null && _d !== void 0 ? _d : undefined;
        message.unexpectedLocation = (_e = object.unexpectedLocation) !== null && _e !== void 0 ? _e : undefined;
        message.slowEthernetSpeeds = (_f = object.slowEthernetSpeeds) !== null && _f !== void 0 ? _f : undefined;
        return message;
    },
};
function createBaseDishObstructionStats() {
    return {
        currentlyObstructed: undefined,
        fractionObstructed: undefined,
        last24hObstructedS: undefined,
        validS: undefined,
        wedgeFractionObstructed: [],
        wedgeAbsFractionObstructed: [],
    };
}
exports.DishObstructionStats = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.currentlyObstructed !== undefined) {
            writer.uint32(40).bool(message.currentlyObstructed);
        }
        if (message.fractionObstructed !== undefined) {
            writer.uint32(13).float(message.fractionObstructed);
        }
        if (message.last24hObstructedS !== undefined) {
            writer.uint32(8053).float(message.last24hObstructedS);
        }
        if (message.validS !== undefined) {
            writer.uint32(37).float(message.validS);
        }
        writer.uint32(18).fork();
        for (const v of message.wedgeFractionObstructed) {
            writer.float(v);
        }
        writer.join();
        writer.uint32(26).fork();
        for (const v of message.wedgeAbsFractionObstructed) {
            writer.float(v);
        }
        writer.join();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishObstructionStats();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 5: {
                    if (tag !== 40) {
                        break;
                    }
                    message.currentlyObstructed = reader.bool();
                    continue;
                }
                case 1: {
                    if (tag !== 13) {
                        break;
                    }
                    message.fractionObstructed = reader.float();
                    continue;
                }
                case 1006: {
                    if (tag !== 8053) {
                        break;
                    }
                    message.last24hObstructedS = reader.float();
                    continue;
                }
                case 4: {
                    if (tag !== 37) {
                        break;
                    }
                    message.validS = reader.float();
                    continue;
                }
                case 2: {
                    if (tag === 21) {
                        message.wedgeFractionObstructed.push(reader.float());
                        continue;
                    }
                    if (tag === 18) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.wedgeFractionObstructed.push(reader.float());
                        }
                        continue;
                    }
                    break;
                }
                case 3: {
                    if (tag === 29) {
                        message.wedgeAbsFractionObstructed.push(reader.float());
                        continue;
                    }
                    if (tag === 26) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.wedgeAbsFractionObstructed.push(reader.float());
                        }
                        continue;
                    }
                    break;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            currentlyObstructed: isSet(object.currentlyObstructed)
                ? globalThis.Boolean(object.currentlyObstructed)
                : isSet(object.currently_obstructed)
                    ? globalThis.Boolean(object.currently_obstructed)
                    : undefined,
            fractionObstructed: isSet(object.fractionObstructed)
                ? globalThis.Number(object.fractionObstructed)
                : isSet(object.fraction_obstructed)
                    ? globalThis.Number(object.fraction_obstructed)
                    : undefined,
            last24hObstructedS: isSet(object.last24hObstructedS)
                ? globalThis.Number(object.last24hObstructedS)
                : isSet(object.last_24h_obstructed_s)
                    ? globalThis.Number(object.last_24h_obstructed_s)
                    : undefined,
            validS: isSet(object.validS)
                ? globalThis.Number(object.validS)
                : isSet(object.valid_s)
                    ? globalThis.Number(object.valid_s)
                    : undefined,
            wedgeFractionObstructed: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.wedgeFractionObstructed)
                ? object.wedgeFractionObstructed.map((e) => globalThis.Number(e))
                : globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.wedge_fraction_obstructed)
                    ? object.wedge_fraction_obstructed.map((e) => globalThis.Number(e))
                    : [],
            wedgeAbsFractionObstructed: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.wedgeAbsFractionObstructed)
                ? object.wedgeAbsFractionObstructed.map((e) => globalThis.Number(e))
                : globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.wedge_abs_fraction_obstructed)
                    ? object.wedge_abs_fraction_obstructed.map((e) => globalThis.Number(e))
                    : [],
        };
    },
    toJSON(message) {
        var _a, _b;
        const obj = {};
        if (message.currentlyObstructed !== undefined) {
            obj.currentlyObstructed = message.currentlyObstructed;
        }
        if (message.fractionObstructed !== undefined) {
            obj.fractionObstructed = message.fractionObstructed;
        }
        if (message.last24hObstructedS !== undefined) {
            obj.last24hObstructedS = message.last24hObstructedS;
        }
        if (message.validS !== undefined) {
            obj.validS = message.validS;
        }
        if ((_a = message.wedgeFractionObstructed) === null || _a === void 0 ? void 0 : _a.length) {
            obj.wedgeFractionObstructed = message.wedgeFractionObstructed;
        }
        if ((_b = message.wedgeAbsFractionObstructed) === null || _b === void 0 ? void 0 : _b.length) {
            obj.wedgeAbsFractionObstructed = message.wedgeAbsFractionObstructed;
        }
        return obj;
    },
    create(base) {
        return exports.DishObstructionStats.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f;
        const message = createBaseDishObstructionStats();
        message.currentlyObstructed = (_a = object.currentlyObstructed) !== null && _a !== void 0 ? _a : undefined;
        message.fractionObstructed = (_b = object.fractionObstructed) !== null && _b !== void 0 ? _b : undefined;
        message.last24hObstructedS = (_c = object.last24hObstructedS) !== null && _c !== void 0 ? _c : undefined;
        message.validS = (_d = object.validS) !== null && _d !== void 0 ? _d : undefined;
        message.wedgeFractionObstructed = ((_e = object.wedgeFractionObstructed) === null || _e === void 0 ? void 0 : _e.map((e) => e)) || [];
        message.wedgeAbsFractionObstructed = ((_f = object.wedgeAbsFractionObstructed) === null || _f === void 0 ? void 0 : _f.map((e) => e)) || [];
        return message;
    },
};
function createBaseDishAuthenticateResponse() {
    return { dish: undefined };
}
exports.DishAuthenticateResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.dish !== undefined) {
            common_1.ChallengeResponse.encode(message.dish, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishAuthenticateResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.dish = common_1.ChallengeResponse.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { dish: isSet(object.dish) ? common_1.ChallengeResponse.fromJSON(object.dish) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.dish !== undefined) {
            obj.dish = common_1.ChallengeResponse.toJSON(message.dish);
        }
        return obj;
    },
    create(base) {
        return exports.DishAuthenticateResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseDishAuthenticateResponse();
        message.dish = (object.dish !== undefined && object.dish !== null)
            ? common_1.ChallengeResponse.fromPartial(object.dish)
            : undefined;
        return message;
    },
};
function createBaseDishGetDiagnosticsResponse() {
    return {
        id: "",
        hardwareVersion: "",
        softwareVersion: "",
        utcOffsetS: 0,
        hardwareSelfTest: 0,
        alerts: undefined,
        disablementCode: 0,
        location: undefined,
        alignmentStats: undefined,
        stowed: false,
        hardwareSelfTestCodes: [],
    };
}
exports.DishGetDiagnosticsResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.hardwareVersion !== "") {
            writer.uint32(18).string(message.hardwareVersion);
        }
        if (message.softwareVersion !== "") {
            writer.uint32(26).string(message.softwareVersion);
        }
        if (message.utcOffsetS !== 0) {
            writer.uint32(32).int32(message.utcOffsetS);
        }
        if (message.hardwareSelfTest !== 0) {
            writer.uint32(56).int32(message.hardwareSelfTest);
        }
        if (message.alerts !== undefined) {
            exports.DishGetDiagnosticsResponse_Alerts.encode(message.alerts, writer.uint32(42).fork()).join();
        }
        if (message.disablementCode !== 0) {
            writer.uint32(48).int32(message.disablementCode);
        }
        if (message.location !== undefined) {
            exports.DishGetDiagnosticsResponse_Location.encode(message.location, writer.uint32(66).fork()).join();
        }
        if (message.alignmentStats !== undefined) {
            exports.DishGetDiagnosticsResponse_AlignmentStats.encode(message.alignmentStats, writer.uint32(74).fork()).join();
        }
        if (message.stowed !== false) {
            writer.uint32(80).bool(message.stowed);
        }
        writer.uint32(90).fork();
        for (const v of message.hardwareSelfTestCodes) {
            writer.int32(v);
        }
        writer.join();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishGetDiagnosticsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.hardwareVersion = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.softwareVersion = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.utcOffsetS = reader.int32();
                    continue;
                }
                case 7: {
                    if (tag !== 56) {
                        break;
                    }
                    message.hardwareSelfTest = reader.int32();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.alerts = exports.DishGetDiagnosticsResponse_Alerts.decode(reader, reader.uint32());
                    continue;
                }
                case 6: {
                    if (tag !== 48) {
                        break;
                    }
                    message.disablementCode = reader.int32();
                    continue;
                }
                case 8: {
                    if (tag !== 66) {
                        break;
                    }
                    message.location = exports.DishGetDiagnosticsResponse_Location.decode(reader, reader.uint32());
                    continue;
                }
                case 9: {
                    if (tag !== 74) {
                        break;
                    }
                    message.alignmentStats = exports.DishGetDiagnosticsResponse_AlignmentStats.decode(reader, reader.uint32());
                    continue;
                }
                case 10: {
                    if (tag !== 80) {
                        break;
                    }
                    message.stowed = reader.bool();
                    continue;
                }
                case 11: {
                    if (tag === 88) {
                        message.hardwareSelfTestCodes.push(reader.int32());
                        continue;
                    }
                    if (tag === 90) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.hardwareSelfTestCodes.push(reader.int32());
                        }
                        continue;
                    }
                    break;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? globalThis.String(object.id) : "",
            hardwareVersion: isSet(object.hardwareVersion)
                ? globalThis.String(object.hardwareVersion)
                : isSet(object.hardware_version)
                    ? globalThis.String(object.hardware_version)
                    : "",
            softwareVersion: isSet(object.softwareVersion)
                ? globalThis.String(object.softwareVersion)
                : isSet(object.software_version)
                    ? globalThis.String(object.software_version)
                    : "",
            utcOffsetS: isSet(object.utcOffsetS)
                ? globalThis.Number(object.utcOffsetS)
                : isSet(object.utc_offset_s)
                    ? globalThis.Number(object.utc_offset_s)
                    : 0,
            hardwareSelfTest: isSet(object.hardwareSelfTest)
                ? dishGetDiagnosticsResponse_TestResultFromJSON(object.hardwareSelfTest)
                : isSet(object.hardware_self_test)
                    ? dishGetDiagnosticsResponse_TestResultFromJSON(object.hardware_self_test)
                    : 0,
            alerts: isSet(object.alerts) ? exports.DishGetDiagnosticsResponse_Alerts.fromJSON(object.alerts) : undefined,
            disablementCode: isSet(object.disablementCode)
                ? dishGetDiagnosticsResponse_DisablementCodeFromJSON(object.disablementCode)
                : isSet(object.disablement_code)
                    ? dishGetDiagnosticsResponse_DisablementCodeFromJSON(object.disablement_code)
                    : 0,
            location: isSet(object.location) ? exports.DishGetDiagnosticsResponse_Location.fromJSON(object.location) : undefined,
            alignmentStats: isSet(object.alignmentStats)
                ? exports.DishGetDiagnosticsResponse_AlignmentStats.fromJSON(object.alignmentStats)
                : isSet(object.alignment_stats)
                    ? exports.DishGetDiagnosticsResponse_AlignmentStats.fromJSON(object.alignment_stats)
                    : undefined,
            stowed: isSet(object.stowed) ? globalThis.Boolean(object.stowed) : false,
            hardwareSelfTestCodes: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.hardwareSelfTestCodes)
                ? object.hardwareSelfTestCodes.map((e) => dishGetDiagnosticsResponse_TestResultCodeFromJSON(e))
                : globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.hardware_self_test_codes)
                    ? object.hardware_self_test_codes.map((e) => dishGetDiagnosticsResponse_TestResultCodeFromJSON(e))
                    : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.id !== "") {
            obj.id = message.id;
        }
        if (message.hardwareVersion !== "") {
            obj.hardwareVersion = message.hardwareVersion;
        }
        if (message.softwareVersion !== "") {
            obj.softwareVersion = message.softwareVersion;
        }
        if (message.utcOffsetS !== 0) {
            obj.utcOffsetS = Math.round(message.utcOffsetS);
        }
        if (message.hardwareSelfTest !== 0) {
            obj.hardwareSelfTest = dishGetDiagnosticsResponse_TestResultToJSON(message.hardwareSelfTest);
        }
        if (message.alerts !== undefined) {
            obj.alerts = exports.DishGetDiagnosticsResponse_Alerts.toJSON(message.alerts);
        }
        if (message.disablementCode !== 0) {
            obj.disablementCode = dishGetDiagnosticsResponse_DisablementCodeToJSON(message.disablementCode);
        }
        if (message.location !== undefined) {
            obj.location = exports.DishGetDiagnosticsResponse_Location.toJSON(message.location);
        }
        if (message.alignmentStats !== undefined) {
            obj.alignmentStats = exports.DishGetDiagnosticsResponse_AlignmentStats.toJSON(message.alignmentStats);
        }
        if (message.stowed !== false) {
            obj.stowed = message.stowed;
        }
        if ((_a = message.hardwareSelfTestCodes) === null || _a === void 0 ? void 0 : _a.length) {
            obj.hardwareSelfTestCodes = message.hardwareSelfTestCodes.map((e) => dishGetDiagnosticsResponse_TestResultCodeToJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.DishGetDiagnosticsResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const message = createBaseDishGetDiagnosticsResponse();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : "";
        message.hardwareVersion = (_b = object.hardwareVersion) !== null && _b !== void 0 ? _b : "";
        message.softwareVersion = (_c = object.softwareVersion) !== null && _c !== void 0 ? _c : "";
        message.utcOffsetS = (_d = object.utcOffsetS) !== null && _d !== void 0 ? _d : 0;
        message.hardwareSelfTest = (_e = object.hardwareSelfTest) !== null && _e !== void 0 ? _e : 0;
        message.alerts = (object.alerts !== undefined && object.alerts !== null)
            ? exports.DishGetDiagnosticsResponse_Alerts.fromPartial(object.alerts)
            : undefined;
        message.disablementCode = (_f = object.disablementCode) !== null && _f !== void 0 ? _f : 0;
        message.location = (object.location !== undefined && object.location !== null)
            ? exports.DishGetDiagnosticsResponse_Location.fromPartial(object.location)
            : undefined;
        message.alignmentStats = (object.alignmentStats !== undefined && object.alignmentStats !== null)
            ? exports.DishGetDiagnosticsResponse_AlignmentStats.fromPartial(object.alignmentStats)
            : undefined;
        message.stowed = (_g = object.stowed) !== null && _g !== void 0 ? _g : false;
        message.hardwareSelfTestCodes = ((_h = object.hardwareSelfTestCodes) === null || _h === void 0 ? void 0 : _h.map((e) => e)) || [];
        return message;
    },
};
function createBaseDishGetDiagnosticsResponse_Alerts() {
    return {
        dishIsHeating: false,
        dishThermalThrottle: false,
        dishThermalShutdown: false,
        powerSupplyThermalThrottle: false,
        motorsStuck: false,
        mastNotNearVertical: false,
        slowEthernetSpeeds: false,
        softwareInstallPending: false,
        movingTooFastForPolicy: false,
        obstructed: false,
    };
}
exports.DishGetDiagnosticsResponse_Alerts = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.dishIsHeating !== false) {
            writer.uint32(8).bool(message.dishIsHeating);
        }
        if (message.dishThermalThrottle !== false) {
            writer.uint32(16).bool(message.dishThermalThrottle);
        }
        if (message.dishThermalShutdown !== false) {
            writer.uint32(24).bool(message.dishThermalShutdown);
        }
        if (message.powerSupplyThermalThrottle !== false) {
            writer.uint32(32).bool(message.powerSupplyThermalThrottle);
        }
        if (message.motorsStuck !== false) {
            writer.uint32(40).bool(message.motorsStuck);
        }
        if (message.mastNotNearVertical !== false) {
            writer.uint32(48).bool(message.mastNotNearVertical);
        }
        if (message.slowEthernetSpeeds !== false) {
            writer.uint32(56).bool(message.slowEthernetSpeeds);
        }
        if (message.softwareInstallPending !== false) {
            writer.uint32(64).bool(message.softwareInstallPending);
        }
        if (message.movingTooFastForPolicy !== false) {
            writer.uint32(72).bool(message.movingTooFastForPolicy);
        }
        if (message.obstructed !== false) {
            writer.uint32(80).bool(message.obstructed);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishGetDiagnosticsResponse_Alerts();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.dishIsHeating = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.dishThermalThrottle = reader.bool();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.dishThermalShutdown = reader.bool();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.powerSupplyThermalThrottle = reader.bool();
                    continue;
                }
                case 5: {
                    if (tag !== 40) {
                        break;
                    }
                    message.motorsStuck = reader.bool();
                    continue;
                }
                case 6: {
                    if (tag !== 48) {
                        break;
                    }
                    message.mastNotNearVertical = reader.bool();
                    continue;
                }
                case 7: {
                    if (tag !== 56) {
                        break;
                    }
                    message.slowEthernetSpeeds = reader.bool();
                    continue;
                }
                case 8: {
                    if (tag !== 64) {
                        break;
                    }
                    message.softwareInstallPending = reader.bool();
                    continue;
                }
                case 9: {
                    if (tag !== 72) {
                        break;
                    }
                    message.movingTooFastForPolicy = reader.bool();
                    continue;
                }
                case 10: {
                    if (tag !== 80) {
                        break;
                    }
                    message.obstructed = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            dishIsHeating: isSet(object.dishIsHeating)
                ? globalThis.Boolean(object.dishIsHeating)
                : isSet(object.dish_is_heating)
                    ? globalThis.Boolean(object.dish_is_heating)
                    : false,
            dishThermalThrottle: isSet(object.dishThermalThrottle)
                ? globalThis.Boolean(object.dishThermalThrottle)
                : isSet(object.dish_thermal_throttle)
                    ? globalThis.Boolean(object.dish_thermal_throttle)
                    : false,
            dishThermalShutdown: isSet(object.dishThermalShutdown)
                ? globalThis.Boolean(object.dishThermalShutdown)
                : isSet(object.dish_thermal_shutdown)
                    ? globalThis.Boolean(object.dish_thermal_shutdown)
                    : false,
            powerSupplyThermalThrottle: isSet(object.powerSupplyThermalThrottle)
                ? globalThis.Boolean(object.powerSupplyThermalThrottle)
                : isSet(object.power_supply_thermal_throttle)
                    ? globalThis.Boolean(object.power_supply_thermal_throttle)
                    : false,
            motorsStuck: isSet(object.motorsStuck)
                ? globalThis.Boolean(object.motorsStuck)
                : isSet(object.motors_stuck)
                    ? globalThis.Boolean(object.motors_stuck)
                    : false,
            mastNotNearVertical: isSet(object.mastNotNearVertical)
                ? globalThis.Boolean(object.mastNotNearVertical)
                : isSet(object.mast_not_near_vertical)
                    ? globalThis.Boolean(object.mast_not_near_vertical)
                    : false,
            slowEthernetSpeeds: isSet(object.slowEthernetSpeeds)
                ? globalThis.Boolean(object.slowEthernetSpeeds)
                : isSet(object.slow_ethernet_speeds)
                    ? globalThis.Boolean(object.slow_ethernet_speeds)
                    : false,
            softwareInstallPending: isSet(object.softwareInstallPending)
                ? globalThis.Boolean(object.softwareInstallPending)
                : isSet(object.software_install_pending)
                    ? globalThis.Boolean(object.software_install_pending)
                    : false,
            movingTooFastForPolicy: isSet(object.movingTooFastForPolicy)
                ? globalThis.Boolean(object.movingTooFastForPolicy)
                : isSet(object.moving_too_fast_for_policy)
                    ? globalThis.Boolean(object.moving_too_fast_for_policy)
                    : false,
            obstructed: isSet(object.obstructed) ? globalThis.Boolean(object.obstructed) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.dishIsHeating !== false) {
            obj.dishIsHeating = message.dishIsHeating;
        }
        if (message.dishThermalThrottle !== false) {
            obj.dishThermalThrottle = message.dishThermalThrottle;
        }
        if (message.dishThermalShutdown !== false) {
            obj.dishThermalShutdown = message.dishThermalShutdown;
        }
        if (message.powerSupplyThermalThrottle !== false) {
            obj.powerSupplyThermalThrottle = message.powerSupplyThermalThrottle;
        }
        if (message.motorsStuck !== false) {
            obj.motorsStuck = message.motorsStuck;
        }
        if (message.mastNotNearVertical !== false) {
            obj.mastNotNearVertical = message.mastNotNearVertical;
        }
        if (message.slowEthernetSpeeds !== false) {
            obj.slowEthernetSpeeds = message.slowEthernetSpeeds;
        }
        if (message.softwareInstallPending !== false) {
            obj.softwareInstallPending = message.softwareInstallPending;
        }
        if (message.movingTooFastForPolicy !== false) {
            obj.movingTooFastForPolicy = message.movingTooFastForPolicy;
        }
        if (message.obstructed !== false) {
            obj.obstructed = message.obstructed;
        }
        return obj;
    },
    create(base) {
        return exports.DishGetDiagnosticsResponse_Alerts.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        const message = createBaseDishGetDiagnosticsResponse_Alerts();
        message.dishIsHeating = (_a = object.dishIsHeating) !== null && _a !== void 0 ? _a : false;
        message.dishThermalThrottle = (_b = object.dishThermalThrottle) !== null && _b !== void 0 ? _b : false;
        message.dishThermalShutdown = (_c = object.dishThermalShutdown) !== null && _c !== void 0 ? _c : false;
        message.powerSupplyThermalThrottle = (_d = object.powerSupplyThermalThrottle) !== null && _d !== void 0 ? _d : false;
        message.motorsStuck = (_e = object.motorsStuck) !== null && _e !== void 0 ? _e : false;
        message.mastNotNearVertical = (_f = object.mastNotNearVertical) !== null && _f !== void 0 ? _f : false;
        message.slowEthernetSpeeds = (_g = object.slowEthernetSpeeds) !== null && _g !== void 0 ? _g : false;
        message.softwareInstallPending = (_h = object.softwareInstallPending) !== null && _h !== void 0 ? _h : false;
        message.movingTooFastForPolicy = (_j = object.movingTooFastForPolicy) !== null && _j !== void 0 ? _j : false;
        message.obstructed = (_k = object.obstructed) !== null && _k !== void 0 ? _k : false;
        return message;
    },
};
function createBaseDishGetDiagnosticsResponse_Location() {
    return {
        enabled: false,
        latitude: 0,
        longitude: 0,
        altitudeMeters: 0,
        uncertaintyMetersValid: false,
        uncertaintyMeters: 0,
        gpsTimeS: 0,
    };
}
exports.DishGetDiagnosticsResponse_Location = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.enabled !== false) {
            writer.uint32(8).bool(message.enabled);
        }
        if (message.latitude !== 0) {
            writer.uint32(17).double(message.latitude);
        }
        if (message.longitude !== 0) {
            writer.uint32(25).double(message.longitude);
        }
        if (message.altitudeMeters !== 0) {
            writer.uint32(33).double(message.altitudeMeters);
        }
        if (message.uncertaintyMetersValid !== false) {
            writer.uint32(40).bool(message.uncertaintyMetersValid);
        }
        if (message.uncertaintyMeters !== 0) {
            writer.uint32(49).double(message.uncertaintyMeters);
        }
        if (message.gpsTimeS !== 0) {
            writer.uint32(57).double(message.gpsTimeS);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishGetDiagnosticsResponse_Location();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.enabled = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 17) {
                        break;
                    }
                    message.latitude = reader.double();
                    continue;
                }
                case 3: {
                    if (tag !== 25) {
                        break;
                    }
                    message.longitude = reader.double();
                    continue;
                }
                case 4: {
                    if (tag !== 33) {
                        break;
                    }
                    message.altitudeMeters = reader.double();
                    continue;
                }
                case 5: {
                    if (tag !== 40) {
                        break;
                    }
                    message.uncertaintyMetersValid = reader.bool();
                    continue;
                }
                case 6: {
                    if (tag !== 49) {
                        break;
                    }
                    message.uncertaintyMeters = reader.double();
                    continue;
                }
                case 7: {
                    if (tag !== 57) {
                        break;
                    }
                    message.gpsTimeS = reader.double();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            enabled: isSet(object.enabled) ? globalThis.Boolean(object.enabled) : false,
            latitude: isSet(object.latitude) ? globalThis.Number(object.latitude) : 0,
            longitude: isSet(object.longitude) ? globalThis.Number(object.longitude) : 0,
            altitudeMeters: isSet(object.altitudeMeters)
                ? globalThis.Number(object.altitudeMeters)
                : isSet(object.altitude_meters)
                    ? globalThis.Number(object.altitude_meters)
                    : 0,
            uncertaintyMetersValid: isSet(object.uncertaintyMetersValid)
                ? globalThis.Boolean(object.uncertaintyMetersValid)
                : isSet(object.uncertainty_meters_valid)
                    ? globalThis.Boolean(object.uncertainty_meters_valid)
                    : false,
            uncertaintyMeters: isSet(object.uncertaintyMeters)
                ? globalThis.Number(object.uncertaintyMeters)
                : isSet(object.uncertainty_meters)
                    ? globalThis.Number(object.uncertainty_meters)
                    : 0,
            gpsTimeS: isSet(object.gpsTimeS)
                ? globalThis.Number(object.gpsTimeS)
                : isSet(object.gps_time_s)
                    ? globalThis.Number(object.gps_time_s)
                    : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.enabled !== false) {
            obj.enabled = message.enabled;
        }
        if (message.latitude !== 0) {
            obj.latitude = message.latitude;
        }
        if (message.longitude !== 0) {
            obj.longitude = message.longitude;
        }
        if (message.altitudeMeters !== 0) {
            obj.altitudeMeters = message.altitudeMeters;
        }
        if (message.uncertaintyMetersValid !== false) {
            obj.uncertaintyMetersValid = message.uncertaintyMetersValid;
        }
        if (message.uncertaintyMeters !== 0) {
            obj.uncertaintyMeters = message.uncertaintyMeters;
        }
        if (message.gpsTimeS !== 0) {
            obj.gpsTimeS = message.gpsTimeS;
        }
        return obj;
    },
    create(base) {
        return exports.DishGetDiagnosticsResponse_Location.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g;
        const message = createBaseDishGetDiagnosticsResponse_Location();
        message.enabled = (_a = object.enabled) !== null && _a !== void 0 ? _a : false;
        message.latitude = (_b = object.latitude) !== null && _b !== void 0 ? _b : 0;
        message.longitude = (_c = object.longitude) !== null && _c !== void 0 ? _c : 0;
        message.altitudeMeters = (_d = object.altitudeMeters) !== null && _d !== void 0 ? _d : 0;
        message.uncertaintyMetersValid = (_e = object.uncertaintyMetersValid) !== null && _e !== void 0 ? _e : false;
        message.uncertaintyMeters = (_f = object.uncertaintyMeters) !== null && _f !== void 0 ? _f : 0;
        message.gpsTimeS = (_g = object.gpsTimeS) !== null && _g !== void 0 ? _g : 0;
        return message;
    },
};
function createBaseDishGetDiagnosticsResponse_AlignmentStats() {
    return {
        boresightAzimuthDeg: 0,
        boresightElevationDeg: 0,
        desiredBoresightAzimuthDeg: 0,
        desiredBoresightElevationDeg: 0,
    };
}
exports.DishGetDiagnosticsResponse_AlignmentStats = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.boresightAzimuthDeg !== 0) {
            writer.uint32(13).float(message.boresightAzimuthDeg);
        }
        if (message.boresightElevationDeg !== 0) {
            writer.uint32(21).float(message.boresightElevationDeg);
        }
        if (message.desiredBoresightAzimuthDeg !== 0) {
            writer.uint32(29).float(message.desiredBoresightAzimuthDeg);
        }
        if (message.desiredBoresightElevationDeg !== 0) {
            writer.uint32(37).float(message.desiredBoresightElevationDeg);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDishGetDiagnosticsResponse_AlignmentStats();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 13) {
                        break;
                    }
                    message.boresightAzimuthDeg = reader.float();
                    continue;
                }
                case 2: {
                    if (tag !== 21) {
                        break;
                    }
                    message.boresightElevationDeg = reader.float();
                    continue;
                }
                case 3: {
                    if (tag !== 29) {
                        break;
                    }
                    message.desiredBoresightAzimuthDeg = reader.float();
                    continue;
                }
                case 4: {
                    if (tag !== 37) {
                        break;
                    }
                    message.desiredBoresightElevationDeg = reader.float();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            boresightAzimuthDeg: isSet(object.boresightAzimuthDeg)
                ? globalThis.Number(object.boresightAzimuthDeg)
                : isSet(object.boresight_azimuth_deg)
                    ? globalThis.Number(object.boresight_azimuth_deg)
                    : 0,
            boresightElevationDeg: isSet(object.boresightElevationDeg)
                ? globalThis.Number(object.boresightElevationDeg)
                : isSet(object.boresight_elevation_deg)
                    ? globalThis.Number(object.boresight_elevation_deg)
                    : 0,
            desiredBoresightAzimuthDeg: isSet(object.desiredBoresightAzimuthDeg)
                ? globalThis.Number(object.desiredBoresightAzimuthDeg)
                : isSet(object.desired_boresight_azimuth_deg)
                    ? globalThis.Number(object.desired_boresight_azimuth_deg)
                    : 0,
            desiredBoresightElevationDeg: isSet(object.desiredBoresightElevationDeg)
                ? globalThis.Number(object.desiredBoresightElevationDeg)
                : isSet(object.desired_boresight_elevation_deg)
                    ? globalThis.Number(object.desired_boresight_elevation_deg)
                    : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.boresightAzimuthDeg !== 0) {
            obj.boresightAzimuthDeg = message.boresightAzimuthDeg;
        }
        if (message.boresightElevationDeg !== 0) {
            obj.boresightElevationDeg = message.boresightElevationDeg;
        }
        if (message.desiredBoresightAzimuthDeg !== 0) {
            obj.desiredBoresightAzimuthDeg = message.desiredBoresightAzimuthDeg;
        }
        if (message.desiredBoresightElevationDeg !== 0) {
            obj.desiredBoresightElevationDeg = message.desiredBoresightElevationDeg;
        }
        return obj;
    },
    create(base) {
        return exports.DishGetDiagnosticsResponse_AlignmentStats.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseDishGetDiagnosticsResponse_AlignmentStats();
        message.boresightAzimuthDeg = (_a = object.boresightAzimuthDeg) !== null && _a !== void 0 ? _a : 0;
        message.boresightElevationDeg = (_b = object.boresightElevationDeg) !== null && _b !== void 0 ? _b : 0;
        message.desiredBoresightAzimuthDeg = (_c = object.desiredBoresightAzimuthDeg) !== null && _c !== void 0 ? _c : 0;
        message.desiredBoresightElevationDeg = (_d = object.desiredBoresightElevationDeg) !== null && _d !== void 0 ? _d : 0;
        return message;
    },
};
function longToNumber(int64) {
    const num = globalThis.Number(int64.toString());
    if (num > globalThis.Number.MAX_SAFE_INTEGER) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    if (num < globalThis.Number.MIN_SAFE_INTEGER) {
        throw new globalThis.Error("Value is smaller than Number.MIN_SAFE_INTEGER");
    }
    return num;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
