// @gibme/starlink/dist/protobuf/spacex/api/device/transceiver.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.11.6
//   protoc               v3.21.12
// source: spacex/api/device/transceiver.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransceiverGetTelemetryResponse = exports.TransceiverGetTelemetryRequest = exports.TransceiverFaults = exports.TransceiverGetStatusResponse = exports.TransceiverGetStatusRequest = exports.TransceiverIFLoopbackTestResponse = exports.TransceiverIFLoopbackTestRequest = exports.TransceiverTransmitBlankingState = exports.TransceiverTxRxState = exports.TransceiverModulatorState = exports.protobufPackage = void 0;
exports.transceiverModulatorStateFromJSON = transceiverModulatorStateFromJSON;
exports.transceiverModulatorStateToJSON = transceiverModulatorStateToJSON;
exports.transceiverTxRxStateFromJSON = transceiverTxRxStateFromJSON;
exports.transceiverTxRxStateToJSON = transceiverTxRxStateToJSON;
exports.transceiverTransmitBlankingStateFromJSON = transceiverTransmitBlankingStateFromJSON;
exports.transceiverTransmitBlankingStateToJSON = transceiverTransmitBlankingStateToJSON;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const dish_1 = require("./dish");
exports.protobufPackage = "SpaceX.API.Device";
var TransceiverModulatorState;
(function (TransceiverModulatorState) {
    TransceiverModulatorState[TransceiverModulatorState["MODSTATE_UNKNOWN"] = 0] = "MODSTATE_UNKNOWN";
    TransceiverModulatorState[TransceiverModulatorState["MODSTATE_ENABLED"] = 1] = "MODSTATE_ENABLED";
    TransceiverModulatorState[TransceiverModulatorState["MODSTATE_DISABLED"] = 2] = "MODSTATE_DISABLED";
    TransceiverModulatorState[TransceiverModulatorState["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(TransceiverModulatorState || (exports.TransceiverModulatorState = TransceiverModulatorState = {}));
function transceiverModulatorStateFromJSON(object) {
    switch (object) {
        case 0:
        case "MODSTATE_UNKNOWN":
            return TransceiverModulatorState.MODSTATE_UNKNOWN;
        case 1:
        case "MODSTATE_ENABLED":
            return TransceiverModulatorState.MODSTATE_ENABLED;
        case 2:
        case "MODSTATE_DISABLED":
            return TransceiverModulatorState.MODSTATE_DISABLED;
        case -1:
        case "UNRECOGNIZED":
        default:
            return TransceiverModulatorState.UNRECOGNIZED;
    }
}
function transceiverModulatorStateToJSON(object) {
    switch (object) {
        case TransceiverModulatorState.MODSTATE_UNKNOWN:
            return "MODSTATE_UNKNOWN";
        case TransceiverModulatorState.MODSTATE_ENABLED:
            return "MODSTATE_ENABLED";
        case TransceiverModulatorState.MODSTATE_DISABLED:
            return "MODSTATE_DISABLED";
        case TransceiverModulatorState.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
var TransceiverTxRxState;
(function (TransceiverTxRxState) {
    TransceiverTxRxState[TransceiverTxRxState["TXRX_UNKNOWN"] = 0] = "TXRX_UNKNOWN";
    TransceiverTxRxState[TransceiverTxRxState["TXRX_ENABLED"] = 1] = "TXRX_ENABLED";
    TransceiverTxRxState[TransceiverTxRxState["TXRX_DISABLED"] = 2] = "TXRX_DISABLED";
    TransceiverTxRxState[TransceiverTxRxState["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(TransceiverTxRxState || (exports.TransceiverTxRxState = TransceiverTxRxState = {}));
function transceiverTxRxStateFromJSON(object) {
    switch (object) {
        case 0:
        case "TXRX_UNKNOWN":
            return TransceiverTxRxState.TXRX_UNKNOWN;
        case 1:
        case "TXRX_ENABLED":
            return TransceiverTxRxState.TXRX_ENABLED;
        case 2:
        case "TXRX_DISABLED":
            return TransceiverTxRxState.TXRX_DISABLED;
        case -1:
        case "UNRECOGNIZED":
        default:
            return TransceiverTxRxState.UNRECOGNIZED;
    }
}
function transceiverTxRxStateToJSON(object) {
    switch (object) {
        case TransceiverTxRxState.TXRX_UNKNOWN:
            return "TXRX_UNKNOWN";
        case TransceiverTxRxState.TXRX_ENABLED:
            return "TXRX_ENABLED";
        case TransceiverTxRxState.TXRX_DISABLED:
            return "TXRX_DISABLED";
        case TransceiverTxRxState.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
var TransceiverTransmitBlankingState;
(function (TransceiverTransmitBlankingState) {
    TransceiverTransmitBlankingState[TransceiverTransmitBlankingState["TB_UNKNOWN"] = 0] = "TB_UNKNOWN";
    TransceiverTransmitBlankingState[TransceiverTransmitBlankingState["TB_ENABLED"] = 1] = "TB_ENABLED";
    TransceiverTransmitBlankingState[TransceiverTransmitBlankingState["TB_DISABLED"] = 2] = "TB_DISABLED";
    TransceiverTransmitBlankingState[TransceiverTransmitBlankingState["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(TransceiverTransmitBlankingState || (exports.TransceiverTransmitBlankingState = TransceiverTransmitBlankingState = {}));
function transceiverTransmitBlankingStateFromJSON(object) {
    switch (object) {
        case 0:
        case "TB_UNKNOWN":
            return TransceiverTransmitBlankingState.TB_UNKNOWN;
        case 1:
        case "TB_ENABLED":
            return TransceiverTransmitBlankingState.TB_ENABLED;
        case 2:
        case "TB_DISABLED":
            return TransceiverTransmitBlankingState.TB_DISABLED;
        case -1:
        case "UNRECOGNIZED":
        default:
            return TransceiverTransmitBlankingState.UNRECOGNIZED;
    }
}
function transceiverTransmitBlankingStateToJSON(object) {
    switch (object) {
        case TransceiverTransmitBlankingState.TB_UNKNOWN:
            return "TB_UNKNOWN";
        case TransceiverTransmitBlankingState.TB_ENABLED:
            return "TB_ENABLED";
        case TransceiverTransmitBlankingState.TB_DISABLED:
            return "TB_DISABLED";
        case TransceiverTransmitBlankingState.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBaseTransceiverIFLoopbackTestRequest() {
    return { enableIfLoopback: undefined };
}
exports.TransceiverIFLoopbackTestRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.enableIfLoopback !== undefined) {
            writer.uint32(8).bool(message.enableIfLoopback);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTransceiverIFLoopbackTestRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.enableIfLoopback = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            enableIfLoopback: isSet(object.enableIfLoopback)
                ? globalThis.Boolean(object.enableIfLoopback)
                : isSet(object.enable_if_loopback)
                    ? globalThis.Boolean(object.enable_if_loopback)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.enableIfLoopback !== undefined) {
            obj.enableIfLoopback = message.enableIfLoopback;
        }
        return obj;
    },
    create(base) {
        return exports.TransceiverIFLoopbackTestRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseTransceiverIFLoopbackTestRequest();
        message.enableIfLoopback = (_a = object.enableIfLoopback) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBaseTransceiverIFLoopbackTestResponse() {
    return { berLoopbackTest: undefined, snrLoopbackTest: undefined, rssiLoopbackTest: undefined, pllLock: undefined };
}
exports.TransceiverIFLoopbackTestResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.berLoopbackTest !== undefined) {
            writer.uint32(13).float(message.berLoopbackTest);
        }
        if (message.snrLoopbackTest !== undefined) {
            writer.uint32(21).float(message.snrLoopbackTest);
        }
        if (message.rssiLoopbackTest !== undefined) {
            writer.uint32(29).float(message.rssiLoopbackTest);
        }
        if (message.pllLock !== undefined) {
            writer.uint32(32).bool(message.pllLock);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTransceiverIFLoopbackTestResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 13) {
                        break;
                    }
                    message.berLoopbackTest = reader.float();
                    continue;
                }
                case 2: {
                    if (tag !== 21) {
                        break;
                    }
                    message.snrLoopbackTest = reader.float();
                    continue;
                }
                case 3: {
                    if (tag !== 29) {
                        break;
                    }
                    message.rssiLoopbackTest = reader.float();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.pllLock = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            berLoopbackTest: isSet(object.berLoopbackTest)
                ? globalThis.Number(object.berLoopbackTest)
                : isSet(object.ber_loopback_test)
                    ? globalThis.Number(object.ber_loopback_test)
                    : undefined,
            snrLoopbackTest: isSet(object.snrLoopbackTest)
                ? globalThis.Number(object.snrLoopbackTest)
                : isSet(object.snr_loopback_test)
                    ? globalThis.Number(object.snr_loopback_test)
                    : undefined,
            rssiLoopbackTest: isSet(object.rssiLoopbackTest)
                ? globalThis.Number(object.rssiLoopbackTest)
                : isSet(object.rssi_loopback_test)
                    ? globalThis.Number(object.rssi_loopback_test)
                    : undefined,
            pllLock: isSet(object.pllLock)
                ? globalThis.Boolean(object.pllLock)
                : isSet(object.pll_lock)
                    ? globalThis.Boolean(object.pll_lock)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.berLoopbackTest !== undefined) {
            obj.berLoopbackTest = message.berLoopbackTest;
        }
        if (message.snrLoopbackTest !== undefined) {
            obj.snrLoopbackTest = message.snrLoopbackTest;
        }
        if (message.rssiLoopbackTest !== undefined) {
            obj.rssiLoopbackTest = message.rssiLoopbackTest;
        }
        if (message.pllLock !== undefined) {
            obj.pllLock = message.pllLock;
        }
        return obj;
    },
    create(base) {
        return exports.TransceiverIFLoopbackTestResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseTransceiverIFLoopbackTestResponse();
        message.berLoopbackTest = (_a = object.berLoopbackTest) !== null && _a !== void 0 ? _a : undefined;
        message.snrLoopbackTest = (_b = object.snrLoopbackTest) !== null && _b !== void 0 ? _b : undefined;
        message.rssiLoopbackTest = (_c = object.rssiLoopbackTest) !== null && _c !== void 0 ? _c : undefined;
        message.pllLock = (_d = object.pllLock) !== null && _d !== void 0 ? _d : undefined;
        return message;
    },
};
function createBaseTransceiverGetStatusRequest() {
    return {};
}
exports.TransceiverGetStatusRequest = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTransceiverGetStatusRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.TransceiverGetStatusRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseTransceiverGetStatusRequest();
        return message;
    },
};
function createBaseTransceiverGetStatusResponse() {
    return {
        modState: undefined,
        demodState: undefined,
        txState: undefined,
        rxState: undefined,
        state: undefined,
        faults: undefined,
        transmitBlankingState: undefined,
        modemAsicTemp: undefined,
        txIfTemp: undefined,
    };
}
exports.TransceiverGetStatusResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.modState !== undefined) {
            writer.uint32(8).int32(message.modState);
        }
        if (message.demodState !== undefined) {
            writer.uint32(16).int32(message.demodState);
        }
        if (message.txState !== undefined) {
            writer.uint32(24).int32(message.txState);
        }
        if (message.rxState !== undefined) {
            writer.uint32(32).int32(message.rxState);
        }
        if (message.state !== undefined) {
            writer.uint32(8048).int32(message.state);
        }
        if (message.faults !== undefined) {
            exports.TransceiverFaults.encode(message.faults, writer.uint32(8058).fork()).join();
        }
        if (message.transmitBlankingState !== undefined) {
            writer.uint32(8064).int32(message.transmitBlankingState);
        }
        if (message.modemAsicTemp !== undefined) {
            writer.uint32(8077).float(message.modemAsicTemp);
        }
        if (message.txIfTemp !== undefined) {
            writer.uint32(8085).float(message.txIfTemp);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTransceiverGetStatusResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.modState = reader.int32();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.demodState = reader.int32();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.txState = reader.int32();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.rxState = reader.int32();
                    continue;
                }
                case 1006: {
                    if (tag !== 8048) {
                        break;
                    }
                    message.state = reader.int32();
                    continue;
                }
                case 1007: {
                    if (tag !== 8058) {
                        break;
                    }
                    message.faults = exports.TransceiverFaults.decode(reader, reader.uint32());
                    continue;
                }
                case 1008: {
                    if (tag !== 8064) {
                        break;
                    }
                    message.transmitBlankingState = reader.int32();
                    continue;
                }
                case 1009: {
                    if (tag !== 8077) {
                        break;
                    }
                    message.modemAsicTemp = reader.float();
                    continue;
                }
                case 1010: {
                    if (tag !== 8085) {
                        break;
                    }
                    message.txIfTemp = reader.float();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            modState: isSet(object.modState)
                ? transceiverModulatorStateFromJSON(object.modState)
                : isSet(object.mod_state)
                    ? transceiverModulatorStateFromJSON(object.mod_state)
                    : undefined,
            demodState: isSet(object.demodState)
                ? transceiverModulatorStateFromJSON(object.demodState)
                : isSet(object.demod_state)
                    ? transceiverModulatorStateFromJSON(object.demod_state)
                    : undefined,
            txState: isSet(object.txState)
                ? transceiverTxRxStateFromJSON(object.txState)
                : isSet(object.tx_state)
                    ? transceiverTxRxStateFromJSON(object.tx_state)
                    : undefined,
            rxState: isSet(object.rxState)
                ? transceiverTxRxStateFromJSON(object.rxState)
                : isSet(object.rx_state)
                    ? transceiverTxRxStateFromJSON(object.rx_state)
                    : undefined,
            state: isSet(object.state) ? (0, dish_1.dishStateFromJSON)(object.state) : undefined,
            faults: isSet(object.faults) ? exports.TransceiverFaults.fromJSON(object.faults) : undefined,
            transmitBlankingState: isSet(object.transmitBlankingState)
                ? transceiverTransmitBlankingStateFromJSON(object.transmitBlankingState)
                : isSet(object.transmit_blanking_state)
                    ? transceiverTransmitBlankingStateFromJSON(object.transmit_blanking_state)
                    : undefined,
            modemAsicTemp: isSet(object.modemAsicTemp)
                ? globalThis.Number(object.modemAsicTemp)
                : isSet(object.modem_asic_temp)
                    ? globalThis.Number(object.modem_asic_temp)
                    : undefined,
            txIfTemp: isSet(object.txIfTemp)
                ? globalThis.Number(object.txIfTemp)
                : isSet(object.tx_if_temp)
                    ? globalThis.Number(object.tx_if_temp)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.modState !== undefined) {
            obj.modState = transceiverModulatorStateToJSON(message.modState);
        }
        if (message.demodState !== undefined) {
            obj.demodState = transceiverModulatorStateToJSON(message.demodState);
        }
        if (message.txState !== undefined) {
            obj.txState = transceiverTxRxStateToJSON(message.txState);
        }
        if (message.rxState !== undefined) {
            obj.rxState = transceiverTxRxStateToJSON(message.rxState);
        }
        if (message.state !== undefined) {
            obj.state = (0, dish_1.dishStateToJSON)(message.state);
        }
        if (message.faults !== undefined) {
            obj.faults = exports.TransceiverFaults.toJSON(message.faults);
        }
        if (message.transmitBlankingState !== undefined) {
            obj.transmitBlankingState = transceiverTransmitBlankingStateToJSON(message.transmitBlankingState);
        }
        if (message.modemAsicTemp !== undefined) {
            obj.modemAsicTemp = message.modemAsicTemp;
        }
        if (message.txIfTemp !== undefined) {
            obj.txIfTemp = message.txIfTemp;
        }
        return obj;
    },
    create(base) {
        return exports.TransceiverGetStatusResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const message = createBaseTransceiverGetStatusResponse();
        message.modState = (_a = object.modState) !== null && _a !== void 0 ? _a : undefined;
        message.demodState = (_b = object.demodState) !== null && _b !== void 0 ? _b : undefined;
        message.txState = (_c = object.txState) !== null && _c !== void 0 ? _c : undefined;
        message.rxState = (_d = object.rxState) !== null && _d !== void 0 ? _d : undefined;
        message.state = (_e = object.state) !== null && _e !== void 0 ? _e : undefined;
        message.faults = (object.faults !== undefined && object.faults !== null)
            ? exports.TransceiverFaults.fromPartial(object.faults)
            : undefined;
        message.transmitBlankingState = (_f = object.transmitBlankingState) !== null && _f !== void 0 ? _f : undefined;
        message.modemAsicTemp = (_g = object.modemAsicTemp) !== null && _g !== void 0 ? _g : undefined;
        message.txIfTemp = (_h = object.txIfTemp) !== null && _h !== void 0 ? _h : undefined;
        return message;
    },
};
function createBaseTransceiverFaults() {
    return { overTempModemAsicFault: undefined, overTempPcbaFault: undefined, dcVoltageFault: undefined };
}
exports.TransceiverFaults = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.overTempModemAsicFault !== undefined) {
            writer.uint32(8).bool(message.overTempModemAsicFault);
        }
        if (message.overTempPcbaFault !== undefined) {
            writer.uint32(16).bool(message.overTempPcbaFault);
        }
        if (message.dcVoltageFault !== undefined) {
            writer.uint32(24).bool(message.dcVoltageFault);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTransceiverFaults();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.overTempModemAsicFault = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.overTempPcbaFault = reader.bool();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.dcVoltageFault = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            overTempModemAsicFault: isSet(object.overTempModemAsicFault)
                ? globalThis.Boolean(object.overTempModemAsicFault)
                : isSet(object.over_temp_modem_asic_fault)
                    ? globalThis.Boolean(object.over_temp_modem_asic_fault)
                    : undefined,
            overTempPcbaFault: isSet(object.overTempPcbaFault)
                ? globalThis.Boolean(object.overTempPcbaFault)
                : isSet(object.over_temp_pcba_fault)
                    ? globalThis.Boolean(object.over_temp_pcba_fault)
                    : undefined,
            dcVoltageFault: isSet(object.dcVoltageFault)
                ? globalThis.Boolean(object.dcVoltageFault)
                : isSet(object.dc_voltage_fault)
                    ? globalThis.Boolean(object.dc_voltage_fault)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.overTempModemAsicFault !== undefined) {
            obj.overTempModemAsicFault = message.overTempModemAsicFault;
        }
        if (message.overTempPcbaFault !== undefined) {
            obj.overTempPcbaFault = message.overTempPcbaFault;
        }
        if (message.dcVoltageFault !== undefined) {
            obj.dcVoltageFault = message.dcVoltageFault;
        }
        return obj;
    },
    create(base) {
        return exports.TransceiverFaults.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseTransceiverFaults();
        message.overTempModemAsicFault = (_a = object.overTempModemAsicFault) !== null && _a !== void 0 ? _a : undefined;
        message.overTempPcbaFault = (_b = object.overTempPcbaFault) !== null && _b !== void 0 ? _b : undefined;
        message.dcVoltageFault = (_c = object.dcVoltageFault) !== null && _c !== void 0 ? _c : undefined;
        return message;
    },
};
function createBaseTransceiverGetTelemetryRequest() {
    return {};
}
exports.TransceiverGetTelemetryRequest = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTransceiverGetTelemetryRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.TransceiverGetTelemetryRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseTransceiverGetTelemetryRequest();
        return message;
    },
};
function createBaseTransceiverGetTelemetryResponse() {
    return {
        antennaPointingMode: undefined,
        antennaPitch: undefined,
        antennaRoll: undefined,
        antennaRxTheta: undefined,
        antennaTrueHeading: undefined,
        rxChannel: undefined,
        currentCellId: undefined,
        secondsUntilSlotEnd: undefined,
        wbRssiPeakMagDb: undefined,
        popPingDropRate: undefined,
        snrDb: undefined,
        l1SnrAvgDb: undefined,
        l1SnrMinDb: undefined,
        l1SnrMaxDb: undefined,
        lmacSatelliteId: undefined,
        targetSatelliteId: undefined,
        grantMcs: undefined,
        grantSymbolsAvg: undefined,
        dedGrant: undefined,
        mobilityProactiveSlotChange: undefined,
        mobilityReactiveSlotChange: undefined,
        rfpTotalSynFailed: undefined,
        numOutOfSeq: undefined,
        numUlmapDrop: undefined,
        currentSecondsOfSchedule: undefined,
        sendLabelSwitchToGroundFailedCalls: undefined,
        emaVelocityX: undefined,
        emaVelocityY: undefined,
        emaVelocityZ: undefined,
        ceRssiDb: undefined,
    };
}
exports.TransceiverGetTelemetryResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.antennaPointingMode !== undefined) {
            writer.uint32(8008).uint32(message.antennaPointingMode);
        }
        if (message.antennaPitch !== undefined) {
            writer.uint32(8021).float(message.antennaPitch);
        }
        if (message.antennaRoll !== undefined) {
            writer.uint32(8029).float(message.antennaRoll);
        }
        if (message.antennaRxTheta !== undefined) {
            writer.uint32(8037).float(message.antennaRxTheta);
        }
        if (message.antennaTrueHeading !== undefined) {
            writer.uint32(8045).float(message.antennaTrueHeading);
        }
        if (message.rxChannel !== undefined) {
            writer.uint32(8048).uint32(message.rxChannel);
        }
        if (message.currentCellId !== undefined) {
            writer.uint32(8056).uint32(message.currentCellId);
        }
        if (message.secondsUntilSlotEnd !== undefined) {
            writer.uint32(8069).float(message.secondsUntilSlotEnd);
        }
        if (message.wbRssiPeakMagDb !== undefined) {
            writer.uint32(8077).float(message.wbRssiPeakMagDb);
        }
        if (message.popPingDropRate !== undefined) {
            writer.uint32(8085).float(message.popPingDropRate);
        }
        if (message.snrDb !== undefined) {
            writer.uint32(8093).float(message.snrDb);
        }
        if (message.l1SnrAvgDb !== undefined) {
            writer.uint32(8101).float(message.l1SnrAvgDb);
        }
        if (message.l1SnrMinDb !== undefined) {
            writer.uint32(8109).float(message.l1SnrMinDb);
        }
        if (message.l1SnrMaxDb !== undefined) {
            writer.uint32(8117).float(message.l1SnrMaxDb);
        }
        if (message.lmacSatelliteId !== undefined) {
            writer.uint32(8120).uint32(message.lmacSatelliteId);
        }
        if (message.targetSatelliteId !== undefined) {
            writer.uint32(8128).uint32(message.targetSatelliteId);
        }
        if (message.grantMcs !== undefined) {
            writer.uint32(8136).uint32(message.grantMcs);
        }
        if (message.grantSymbolsAvg !== undefined) {
            writer.uint32(8149).float(message.grantSymbolsAvg);
        }
        if (message.dedGrant !== undefined) {
            writer.uint32(8152).uint32(message.dedGrant);
        }
        if (message.mobilityProactiveSlotChange !== undefined) {
            writer.uint32(8160).uint32(message.mobilityProactiveSlotChange);
        }
        if (message.mobilityReactiveSlotChange !== undefined) {
            writer.uint32(8168).uint32(message.mobilityReactiveSlotChange);
        }
        if (message.rfpTotalSynFailed !== undefined) {
            writer.uint32(8176).uint32(message.rfpTotalSynFailed);
        }
        if (message.numOutOfSeq !== undefined) {
            writer.uint32(8184).uint32(message.numOutOfSeq);
        }
        if (message.numUlmapDrop !== undefined) {
            writer.uint32(8192).uint32(message.numUlmapDrop);
        }
        if (message.currentSecondsOfSchedule !== undefined) {
            writer.uint32(8205).float(message.currentSecondsOfSchedule);
        }
        if (message.sendLabelSwitchToGroundFailedCalls !== undefined) {
            writer.uint32(8208).uint32(message.sendLabelSwitchToGroundFailedCalls);
        }
        if (message.emaVelocityX !== undefined) {
            writer.uint32(8217).double(message.emaVelocityX);
        }
        if (message.emaVelocityY !== undefined) {
            writer.uint32(8225).double(message.emaVelocityY);
        }
        if (message.emaVelocityZ !== undefined) {
            writer.uint32(8233).double(message.emaVelocityZ);
        }
        if (message.ceRssiDb !== undefined) {
            writer.uint32(8245).float(message.ceRssiDb);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTransceiverGetTelemetryResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1001: {
                    if (tag !== 8008) {
                        break;
                    }
                    message.antennaPointingMode = reader.uint32();
                    continue;
                }
                case 1002: {
                    if (tag !== 8021) {
                        break;
                    }
                    message.antennaPitch = reader.float();
                    continue;
                }
                case 1003: {
                    if (tag !== 8029) {
                        break;
                    }
                    message.antennaRoll = reader.float();
                    continue;
                }
                case 1004: {
                    if (tag !== 8037) {
                        break;
                    }
                    message.antennaRxTheta = reader.float();
                    continue;
                }
                case 1005: {
                    if (tag !== 8045) {
                        break;
                    }
                    message.antennaTrueHeading = reader.float();
                    continue;
                }
                case 1006: {
                    if (tag !== 8048) {
                        break;
                    }
                    message.rxChannel = reader.uint32();
                    continue;
                }
                case 1007: {
                    if (tag !== 8056) {
                        break;
                    }
                    message.currentCellId = reader.uint32();
                    continue;
                }
                case 1008: {
                    if (tag !== 8069) {
                        break;
                    }
                    message.secondsUntilSlotEnd = reader.float();
                    continue;
                }
                case 1009: {
                    if (tag !== 8077) {
                        break;
                    }
                    message.wbRssiPeakMagDb = reader.float();
                    continue;
                }
                case 1010: {
                    if (tag !== 8085) {
                        break;
                    }
                    message.popPingDropRate = reader.float();
                    continue;
                }
                case 1011: {
                    if (tag !== 8093) {
                        break;
                    }
                    message.snrDb = reader.float();
                    continue;
                }
                case 1012: {
                    if (tag !== 8101) {
                        break;
                    }
                    message.l1SnrAvgDb = reader.float();
                    continue;
                }
                case 1013: {
                    if (tag !== 8109) {
                        break;
                    }
                    message.l1SnrMinDb = reader.float();
                    continue;
                }
                case 1014: {
                    if (tag !== 8117) {
                        break;
                    }
                    message.l1SnrMaxDb = reader.float();
                    continue;
                }
                case 1015: {
                    if (tag !== 8120) {
                        break;
                    }
                    message.lmacSatelliteId = reader.uint32();
                    continue;
                }
                case 1016: {
                    if (tag !== 8128) {
                        break;
                    }
                    message.targetSatelliteId = reader.uint32();
                    continue;
                }
                case 1017: {
                    if (tag !== 8136) {
                        break;
                    }
                    message.grantMcs = reader.uint32();
                    continue;
                }
                case 1018: {
                    if (tag !== 8149) {
                        break;
                    }
                    message.grantSymbolsAvg = reader.float();
                    continue;
                }
                case 1019: {
                    if (tag !== 8152) {
                        break;
                    }
                    message.dedGrant = reader.uint32();
                    continue;
                }
                case 1020: {
                    if (tag !== 8160) {
                        break;
                    }
                    message.mobilityProactiveSlotChange = reader.uint32();
                    continue;
                }
                case 1021: {
                    if (tag !== 8168) {
                        break;
                    }
                    message.mobilityReactiveSlotChange = reader.uint32();
                    continue;
                }
                case 1022: {
                    if (tag !== 8176) {
                        break;
                    }
                    message.rfpTotalSynFailed = reader.uint32();
                    continue;
                }
                case 1023: {
                    if (tag !== 8184) {
                        break;
                    }
                    message.numOutOfSeq = reader.uint32();
                    continue;
                }
                case 1024: {
                    if (tag !== 8192) {
                        break;
                    }
                    message.numUlmapDrop = reader.uint32();
                    continue;
                }
                case 1025: {
                    if (tag !== 8205) {
                        break;
                    }
                    message.currentSecondsOfSchedule = reader.float();
                    continue;
                }
                case 1026: {
                    if (tag !== 8208) {
                        break;
                    }
                    message.sendLabelSwitchToGroundFailedCalls = reader.uint32();
                    continue;
                }
                case 1027: {
                    if (tag !== 8217) {
                        break;
                    }
                    message.emaVelocityX = reader.double();
                    continue;
                }
                case 1028: {
                    if (tag !== 8225) {
                        break;
                    }
                    message.emaVelocityY = reader.double();
                    continue;
                }
                case 1029: {
                    if (tag !== 8233) {
                        break;
                    }
                    message.emaVelocityZ = reader.double();
                    continue;
                }
                case 1030: {
                    if (tag !== 8245) {
                        break;
                    }
                    message.ceRssiDb = reader.float();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            antennaPointingMode: isSet(object.antennaPointingMode)
                ? globalThis.Number(object.antennaPointingMode)
                : isSet(object.antenna_pointing_mode)
                    ? globalThis.Number(object.antenna_pointing_mode)
                    : undefined,
            antennaPitch: isSet(object.antennaPitch)
                ? globalThis.Number(object.antennaPitch)
                : isSet(object.antenna_pitch)
                    ? globalThis.Number(object.antenna_pitch)
                    : undefined,
            antennaRoll: isSet(object.antennaRoll)
                ? globalThis.Number(object.antennaRoll)
                : isSet(object.antenna_roll)
                    ? globalThis.Number(object.antenna_roll)
                    : undefined,
            antennaRxTheta: isSet(object.antennaRxTheta)
                ? globalThis.Number(object.antennaRxTheta)
                : isSet(object.antenna_rx_theta)
                    ? globalThis.Number(object.antenna_rx_theta)
                    : undefined,
            antennaTrueHeading: isSet(object.antennaTrueHeading)
                ? globalThis.Number(object.antennaTrueHeading)
                : isSet(object.antenna_true_heading)
                    ? globalThis.Number(object.antenna_true_heading)
                    : undefined,
            rxChannel: isSet(object.rxChannel)
                ? globalThis.Number(object.rxChannel)
                : isSet(object.rx_channel)
                    ? globalThis.Number(object.rx_channel)
                    : undefined,
            currentCellId: isSet(object.currentCellId)
                ? globalThis.Number(object.currentCellId)
                : isSet(object.current_cell_id)
                    ? globalThis.Number(object.current_cell_id)
                    : undefined,
            secondsUntilSlotEnd: isSet(object.secondsUntilSlotEnd)
                ? globalThis.Number(object.secondsUntilSlotEnd)
                : isSet(object.seconds_until_slot_end)
                    ? globalThis.Number(object.seconds_until_slot_end)
                    : undefined,
            wbRssiPeakMagDb: isSet(object.wbRssiPeakMagDb)
                ? globalThis.Number(object.wbRssiPeakMagDb)
                : isSet(object.wb_rssi_peak_mag_db)
                    ? globalThis.Number(object.wb_rssi_peak_mag_db)
                    : undefined,
            popPingDropRate: isSet(object.popPingDropRate)
                ? globalThis.Number(object.popPingDropRate)
                : isSet(object.pop_ping_drop_rate)
                    ? globalThis.Number(object.pop_ping_drop_rate)
                    : undefined,
            snrDb: isSet(object.snrDb)
                ? globalThis.Number(object.snrDb)
                : isSet(object.snr_db)
                    ? globalThis.Number(object.snr_db)
                    : undefined,
            l1SnrAvgDb: isSet(object.l1SnrAvgDb)
                ? globalThis.Number(object.l1SnrAvgDb)
                : isSet(object.l1_snr_avg_db)
                    ? globalThis.Number(object.l1_snr_avg_db)
                    : undefined,
            l1SnrMinDb: isSet(object.l1SnrMinDb)
                ? globalThis.Number(object.l1SnrMinDb)
                : isSet(object.l1_snr_min_db)
                    ? globalThis.Number(object.l1_snr_min_db)
                    : undefined,
            l1SnrMaxDb: isSet(object.l1SnrMaxDb)
                ? globalThis.Number(object.l1SnrMaxDb)
                : isSet(object.l1_snr_max_db)
                    ? globalThis.Number(object.l1_snr_max_db)
                    : undefined,
            lmacSatelliteId: isSet(object.lmacSatelliteId)
                ? globalThis.Number(object.lmacSatelliteId)
                : isSet(object.lmac_satellite_id)
                    ? globalThis.Number(object.lmac_satellite_id)
                    : undefined,
            targetSatelliteId: isSet(object.targetSatelliteId)
                ? globalThis.Number(object.targetSatelliteId)
                : isSet(object.target_satellite_id)
                    ? globalThis.Number(object.target_satellite_id)
                    : undefined,
            grantMcs: isSet(object.grantMcs)
                ? globalThis.Number(object.grantMcs)
                : isSet(object.grant_mcs)
                    ? globalThis.Number(object.grant_mcs)
                    : undefined,
            grantSymbolsAvg: isSet(object.grantSymbolsAvg)
                ? globalThis.Number(object.grantSymbolsAvg)
                : isSet(object.grant_symbols_avg)
                    ? globalThis.Number(object.grant_symbols_avg)
                    : undefined,
            dedGrant: isSet(object.dedGrant)
                ? globalThis.Number(object.dedGrant)
                : isSet(object.ded_grant)
                    ? globalThis.Number(object.ded_grant)
                    : undefined,
            mobilityProactiveSlotChange: isSet(object.mobilityProactiveSlotChange)
                ? globalThis.Number(object.mobilityProactiveSlotChange)
                : isSet(object.mobility_proactive_slot_change)
                    ? globalThis.Number(object.mobility_proactive_slot_change)
                    : undefined,
            mobilityReactiveSlotChange: isSet(object.mobilityReactiveSlotChange)
                ? globalThis.Number(object.mobilityReactiveSlotChange)
                : isSet(object.mobility_reactive_slot_change)
                    ? globalThis.Number(object.mobility_reactive_slot_change)
                    : undefined,
            rfpTotalSynFailed: isSet(object.rfpTotalSynFailed)
                ? globalThis.Number(object.rfpTotalSynFailed)
                : isSet(object.rfp_total_syn_failed)
                    ? globalThis.Number(object.rfp_total_syn_failed)
                    : undefined,
            numOutOfSeq: isSet(object.numOutOfSeq)
                ? globalThis.Number(object.numOutOfSeq)
                : isSet(object.num_out_of_seq)
                    ? globalThis.Number(object.num_out_of_seq)
                    : undefined,
            numUlmapDrop: isSet(object.numUlmapDrop)
                ? globalThis.Number(object.numUlmapDrop)
                : isSet(object.num_ulmap_drop)
                    ? globalThis.Number(object.num_ulmap_drop)
                    : undefined,
            currentSecondsOfSchedule: isSet(object.currentSecondsOfSchedule)
                ? globalThis.Number(object.currentSecondsOfSchedule)
                : isSet(object.current_seconds_of_schedule)
                    ? globalThis.Number(object.current_seconds_of_schedule)
                    : undefined,
            sendLabelSwitchToGroundFailedCalls: isSet(object.sendLabelSwitchToGroundFailedCalls)
                ? globalThis.Number(object.sendLabelSwitchToGroundFailedCalls)
                : isSet(object.send_label_switch_to_ground_failed_calls)
                    ? globalThis.Number(object.send_label_switch_to_ground_failed_calls)
                    : undefined,
            emaVelocityX: isSet(object.emaVelocityX)
                ? globalThis.Number(object.emaVelocityX)
                : isSet(object.ema_velocity_x)
                    ? globalThis.Number(object.ema_velocity_x)
                    : undefined,
            emaVelocityY: isSet(object.emaVelocityY)
                ? globalThis.Number(object.emaVelocityY)
                : isSet(object.ema_velocity_y)
                    ? globalThis.Number(object.ema_velocity_y)
                    : undefined,
            emaVelocityZ: isSet(object.emaVelocityZ)
                ? globalThis.Number(object.emaVelocityZ)
                : isSet(object.ema_velocity_z)
                    ? globalThis.Number(object.ema_velocity_z)
                    : undefined,
            ceRssiDb: isSet(object.ceRssiDb)
                ? globalThis.Number(object.ceRssiDb)
                : isSet(object.ce_rssi_db)
                    ? globalThis.Number(object.ce_rssi_db)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.antennaPointingMode !== undefined) {
            obj.antennaPointingMode = Math.round(message.antennaPointingMode);
        }
        if (message.antennaPitch !== undefined) {
            obj.antennaPitch = message.antennaPitch;
        }
        if (message.antennaRoll !== undefined) {
            obj.antennaRoll = message.antennaRoll;
        }
        if (message.antennaRxTheta !== undefined) {
            obj.antennaRxTheta = message.antennaRxTheta;
        }
        if (message.antennaTrueHeading !== undefined) {
            obj.antennaTrueHeading = message.antennaTrueHeading;
        }
        if (message.rxChannel !== undefined) {
            obj.rxChannel = Math.round(message.rxChannel);
        }
        if (message.currentCellId !== undefined) {
            obj.currentCellId = Math.round(message.currentCellId);
        }
        if (message.secondsUntilSlotEnd !== undefined) {
            obj.secondsUntilSlotEnd = message.secondsUntilSlotEnd;
        }
        if (message.wbRssiPeakMagDb !== undefined) {
            obj.wbRssiPeakMagDb = message.wbRssiPeakMagDb;
        }
        if (message.popPingDropRate !== undefined) {
            obj.popPingDropRate = message.popPingDropRate;
        }
        if (message.snrDb !== undefined) {
            obj.snrDb = message.snrDb;
        }
        if (message.l1SnrAvgDb !== undefined) {
            obj.l1SnrAvgDb = message.l1SnrAvgDb;
        }
        if (message.l1SnrMinDb !== undefined) {
            obj.l1SnrMinDb = message.l1SnrMinDb;
        }
        if (message.l1SnrMaxDb !== undefined) {
            obj.l1SnrMaxDb = message.l1SnrMaxDb;
        }
        if (message.lmacSatelliteId !== undefined) {
            obj.lmacSatelliteId = Math.round(message.lmacSatelliteId);
        }
        if (message.targetSatelliteId !== undefined) {
            obj.targetSatelliteId = Math.round(message.targetSatelliteId);
        }
        if (message.grantMcs !== undefined) {
            obj.grantMcs = Math.round(message.grantMcs);
        }
        if (message.grantSymbolsAvg !== undefined) {
            obj.grantSymbolsAvg = message.grantSymbolsAvg;
        }
        if (message.dedGrant !== undefined) {
            obj.dedGrant = Math.round(message.dedGrant);
        }
        if (message.mobilityProactiveSlotChange !== undefined) {
            obj.mobilityProactiveSlotChange = Math.round(message.mobilityProactiveSlotChange);
        }
        if (message.mobilityReactiveSlotChange !== undefined) {
            obj.mobilityReactiveSlotChange = Math.round(message.mobilityReactiveSlotChange);
        }
        if (message.rfpTotalSynFailed !== undefined) {
            obj.rfpTotalSynFailed = Math.round(message.rfpTotalSynFailed);
        }
        if (message.numOutOfSeq !== undefined) {
            obj.numOutOfSeq = Math.round(message.numOutOfSeq);
        }
        if (message.numUlmapDrop !== undefined) {
            obj.numUlmapDrop = Math.round(message.numUlmapDrop);
        }
        if (message.currentSecondsOfSchedule !== undefined) {
            obj.currentSecondsOfSchedule = message.currentSecondsOfSchedule;
        }
        if (message.sendLabelSwitchToGroundFailedCalls !== undefined) {
            obj.sendLabelSwitchToGroundFailedCalls = Math.round(message.sendLabelSwitchToGroundFailedCalls);
        }
        if (message.emaVelocityX !== undefined) {
            obj.emaVelocityX = message.emaVelocityX;
        }
        if (message.emaVelocityY !== undefined) {
            obj.emaVelocityY = message.emaVelocityY;
        }
        if (message.emaVelocityZ !== undefined) {
            obj.emaVelocityZ = message.emaVelocityZ;
        }
        if (message.ceRssiDb !== undefined) {
            obj.ceRssiDb = message.ceRssiDb;
        }
        return obj;
    },
    create(base) {
        return exports.TransceiverGetTelemetryResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5;
        const message = createBaseTransceiverGetTelemetryResponse();
        message.antennaPointingMode = (_a = object.antennaPointingMode) !== null && _a !== void 0 ? _a : undefined;
        message.antennaPitch = (_b = object.antennaPitch) !== null && _b !== void 0 ? _b : undefined;
        message.antennaRoll = (_c = object.antennaRoll) !== null && _c !== void 0 ? _c : undefined;
        message.antennaRxTheta = (_d = object.antennaRxTheta) !== null && _d !== void 0 ? _d : undefined;
        message.antennaTrueHeading = (_e = object.antennaTrueHeading) !== null && _e !== void 0 ? _e : undefined;
        message.rxChannel = (_f = object.rxChannel) !== null && _f !== void 0 ? _f : undefined;
        message.currentCellId = (_g = object.currentCellId) !== null && _g !== void 0 ? _g : undefined;
        message.secondsUntilSlotEnd = (_h = object.secondsUntilSlotEnd) !== null && _h !== void 0 ? _h : undefined;
        message.wbRssiPeakMagDb = (_j = object.wbRssiPeakMagDb) !== null && _j !== void 0 ? _j : undefined;
        message.popPingDropRate = (_k = object.popPingDropRate) !== null && _k !== void 0 ? _k : undefined;
        message.snrDb = (_l = object.snrDb) !== null && _l !== void 0 ? _l : undefined;
        message.l1SnrAvgDb = (_m = object.l1SnrAvgDb) !== null && _m !== void 0 ? _m : undefined;
        message.l1SnrMinDb = (_o = object.l1SnrMinDb) !== null && _o !== void 0 ? _o : undefined;
        message.l1SnrMaxDb = (_p = object.l1SnrMaxDb) !== null && _p !== void 0 ? _p : undefined;
        message.lmacSatelliteId = (_q = object.lmacSatelliteId) !== null && _q !== void 0 ? _q : undefined;
        message.targetSatelliteId = (_r = object.targetSatelliteId) !== null && _r !== void 0 ? _r : undefined;
        message.grantMcs = (_s = object.grantMcs) !== null && _s !== void 0 ? _s : undefined;
        message.grantSymbolsAvg = (_t = object.grantSymbolsAvg) !== null && _t !== void 0 ? _t : undefined;
        message.dedGrant = (_u = object.dedGrant) !== null && _u !== void 0 ? _u : undefined;
        message.mobilityProactiveSlotChange = (_v = object.mobilityProactiveSlotChange) !== null && _v !== void 0 ? _v : undefined;
        message.mobilityReactiveSlotChange = (_w = object.mobilityReactiveSlotChange) !== null && _w !== void 0 ? _w : undefined;
        message.rfpTotalSynFailed = (_x = object.rfpTotalSynFailed) !== null && _x !== void 0 ? _x : undefined;
        message.numOutOfSeq = (_y = object.numOutOfSeq) !== null && _y !== void 0 ? _y : undefined;
        message.numUlmapDrop = (_z = object.numUlmapDrop) !== null && _z !== void 0 ? _z : undefined;
        message.currentSecondsOfSchedule = (_0 = object.currentSecondsOfSchedule) !== null && _0 !== void 0 ? _0 : undefined;
        message.sendLabelSwitchToGroundFailedCalls = (_1 = object.sendLabelSwitchToGroundFailedCalls) !== null && _1 !== void 0 ? _1 : undefined;
        message.emaVelocityX = (_2 = object.emaVelocityX) !== null && _2 !== void 0 ? _2 : undefined;
        message.emaVelocityY = (_3 = object.emaVelocityY) !== null && _3 !== void 0 ? _3 : undefined;
        message.emaVelocityZ = (_4 = object.emaVelocityZ) !== null && _4 !== void 0 ? _4 : undefined;
        message.ceRssiDb = (_5 = object.ceRssiDb) !== null && _5 !== void 0 ? _5 : undefined;
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
