// @gibme/starlink/dist/protobuf/spacex/api/device/wifi.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.11.6
//   protoc               v3.21.12
// source: spacex/api/device/wifi.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.WifiGetDiagnosticsResponse2_Network = exports.WifiGetDiagnosticsResponse2 = exports.WifiNetwork = exports.WifiScanResults_Network = exports.WifiScanResults = exports.WifiGetDiagnosticsResponse = exports.WifiGetDiagnosticsRequest = exports.WifiGetPingMetricsResponse = exports.WifiGetPingMetricsRequest = exports.PingMetrics = exports.WifiAccountBondingEvent = exports.WifiAuthenticateResponse = exports.WifiAuthenticateRequest = exports.WifiGetStatusResponse = exports.WifiSetupResponse = exports.WifiSetupRequest = exports.WifiClient_TxStats = exports.WifiClient_RxStats = exports.WifiClient = exports.WifiNewClientConnectedEvent = exports.WifiGetHistoryResponse = exports.WifiGetClientsResponse = exports.WifiGetClientsRequest = exports.WifiGetConfigResponse = exports.WifiGetConfigRequest = exports.WifiSetConfigResponse = exports.WifiSetConfigRequest = exports.WifiNetwork_Band = exports.WifiScanResults_Network_Source = exports.WifiClient_Interface = exports.protobufPackage = void 0;
exports.wifiClient_InterfaceFromJSON = wifiClient_InterfaceFromJSON;
exports.wifiClient_InterfaceToJSON = wifiClient_InterfaceToJSON;
exports.wifiScanResults_Network_SourceFromJSON = wifiScanResults_Network_SourceFromJSON;
exports.wifiScanResults_Network_SourceToJSON = wifiScanResults_Network_SourceToJSON;
exports.wifiNetwork_BandFromJSON = wifiNetwork_BandFromJSON;
exports.wifiNetwork_BandToJSON = wifiNetwork_BandToJSON;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
const common_1 = require("./common");
const wifi_config_1 = require("./wifi_config");
exports.protobufPackage = "SpaceX.API.Device";
var WifiClient_Interface;
(function (WifiClient_Interface) {
    WifiClient_Interface[WifiClient_Interface["UNKNOWN"] = 0] = "UNKNOWN";
    WifiClient_Interface[WifiClient_Interface["ETH"] = 1] = "ETH";
    WifiClient_Interface[WifiClient_Interface["RF_2GHZ"] = 2] = "RF_2GHZ";
    WifiClient_Interface[WifiClient_Interface["RF_5GHZ"] = 3] = "RF_5GHZ";
    WifiClient_Interface[WifiClient_Interface["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(WifiClient_Interface || (exports.WifiClient_Interface = WifiClient_Interface = {}));
function wifiClient_InterfaceFromJSON(object) {
    switch (object) {
        case 0:
        case "UNKNOWN":
            return WifiClient_Interface.UNKNOWN;
        case 1:
        case "ETH":
            return WifiClient_Interface.ETH;
        case 2:
        case "RF_2GHZ":
            return WifiClient_Interface.RF_2GHZ;
        case 3:
        case "RF_5GHZ":
            return WifiClient_Interface.RF_5GHZ;
        case -1:
        case "UNRECOGNIZED":
        default:
            return WifiClient_Interface.UNRECOGNIZED;
    }
}
function wifiClient_InterfaceToJSON(object) {
    switch (object) {
        case WifiClient_Interface.UNKNOWN:
            return "UNKNOWN";
        case WifiClient_Interface.ETH:
            return "ETH";
        case WifiClient_Interface.RF_2GHZ:
            return "RF_2GHZ";
        case WifiClient_Interface.RF_5GHZ:
            return "RF_5GHZ";
        case WifiClient_Interface.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
var WifiScanResults_Network_Source;
(function (WifiScanResults_Network_Source) {
    WifiScanResults_Network_Source[WifiScanResults_Network_Source["UNKNOWN"] = 0] = "UNKNOWN";
    WifiScanResults_Network_Source[WifiScanResults_Network_Source["SCAN_2_4GHZ"] = 1] = "SCAN_2_4GHZ";
    WifiScanResults_Network_Source[WifiScanResults_Network_Source["SCAN_5GHZ"] = 2] = "SCAN_5GHZ";
    WifiScanResults_Network_Source[WifiScanResults_Network_Source["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(WifiScanResults_Network_Source || (exports.WifiScanResults_Network_Source = WifiScanResults_Network_Source = {}));
function wifiScanResults_Network_SourceFromJSON(object) {
    switch (object) {
        case 0:
        case "UNKNOWN":
            return WifiScanResults_Network_Source.UNKNOWN;
        case 1:
        case "SCAN_2_4GHZ":
            return WifiScanResults_Network_Source.SCAN_2_4GHZ;
        case 2:
        case "SCAN_5GHZ":
            return WifiScanResults_Network_Source.SCAN_5GHZ;
        case -1:
        case "UNRECOGNIZED":
        default:
            return WifiScanResults_Network_Source.UNRECOGNIZED;
    }
}
function wifiScanResults_Network_SourceToJSON(object) {
    switch (object) {
        case WifiScanResults_Network_Source.UNKNOWN:
            return "UNKNOWN";
        case WifiScanResults_Network_Source.SCAN_2_4GHZ:
            return "SCAN_2_4GHZ";
        case WifiScanResults_Network_Source.SCAN_5GHZ:
            return "SCAN_5GHZ";
        case WifiScanResults_Network_Source.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
var WifiNetwork_Band;
(function (WifiNetwork_Band) {
    WifiNetwork_Band[WifiNetwork_Band["WIFI_UNKNOWN"] = 0] = "WIFI_UNKNOWN";
    WifiNetwork_Band[WifiNetwork_Band["WIFI_2_4GHZ"] = 1] = "WIFI_2_4GHZ";
    WifiNetwork_Band[WifiNetwork_Band["WIFI_5GHZ"] = 2] = "WIFI_5GHZ";
    WifiNetwork_Band[WifiNetwork_Band["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(WifiNetwork_Band || (exports.WifiNetwork_Band = WifiNetwork_Band = {}));
function wifiNetwork_BandFromJSON(object) {
    switch (object) {
        case 0:
        case "WIFI_UNKNOWN":
            return WifiNetwork_Band.WIFI_UNKNOWN;
        case 1:
        case "WIFI_2_4GHZ":
            return WifiNetwork_Band.WIFI_2_4GHZ;
        case 2:
        case "WIFI_5GHZ":
            return WifiNetwork_Band.WIFI_5GHZ;
        case -1:
        case "UNRECOGNIZED":
        default:
            return WifiNetwork_Band.UNRECOGNIZED;
    }
}
function wifiNetwork_BandToJSON(object) {
    switch (object) {
        case WifiNetwork_Band.WIFI_UNKNOWN:
            return "WIFI_UNKNOWN";
        case WifiNetwork_Band.WIFI_2_4GHZ:
            return "WIFI_2_4GHZ";
        case WifiNetwork_Band.WIFI_5GHZ:
            return "WIFI_5GHZ";
        case WifiNetwork_Band.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBaseWifiSetConfigRequest() {
    return { wifiConfig: undefined };
}
exports.WifiSetConfigRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.wifiConfig !== undefined) {
            wifi_config_1.WifiConfig.encode(message.wifiConfig, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiSetConfigRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.wifiConfig = wifi_config_1.WifiConfig.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            wifiConfig: isSet(object.wifiConfig)
                ? wifi_config_1.WifiConfig.fromJSON(object.wifiConfig)
                : isSet(object.wifi_config)
                    ? wifi_config_1.WifiConfig.fromJSON(object.wifi_config)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.wifiConfig !== undefined) {
            obj.wifiConfig = wifi_config_1.WifiConfig.toJSON(message.wifiConfig);
        }
        return obj;
    },
    create(base) {
        return exports.WifiSetConfigRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseWifiSetConfigRequest();
        message.wifiConfig = (object.wifiConfig !== undefined && object.wifiConfig !== null)
            ? wifi_config_1.WifiConfig.fromPartial(object.wifiConfig)
            : undefined;
        return message;
    },
};
function createBaseWifiSetConfigResponse() {
    return { updatedWifiConfig: undefined };
}
exports.WifiSetConfigResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.updatedWifiConfig !== undefined) {
            wifi_config_1.WifiConfig.encode(message.updatedWifiConfig, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiSetConfigResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.updatedWifiConfig = wifi_config_1.WifiConfig.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            updatedWifiConfig: isSet(object.updatedWifiConfig)
                ? wifi_config_1.WifiConfig.fromJSON(object.updatedWifiConfig)
                : isSet(object.updated_wifi_config)
                    ? wifi_config_1.WifiConfig.fromJSON(object.updated_wifi_config)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.updatedWifiConfig !== undefined) {
            obj.updatedWifiConfig = wifi_config_1.WifiConfig.toJSON(message.updatedWifiConfig);
        }
        return obj;
    },
    create(base) {
        return exports.WifiSetConfigResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseWifiSetConfigResponse();
        message.updatedWifiConfig = (object.updatedWifiConfig !== undefined && object.updatedWifiConfig !== null)
            ? wifi_config_1.WifiConfig.fromPartial(object.updatedWifiConfig)
            : undefined;
        return message;
    },
};
function createBaseWifiGetConfigRequest() {
    return {};
}
exports.WifiGetConfigRequest = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiGetConfigRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.WifiGetConfigRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseWifiGetConfigRequest();
        return message;
    },
};
function createBaseWifiGetConfigResponse() {
    return { wifiConfig: undefined };
}
exports.WifiGetConfigResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.wifiConfig !== undefined) {
            wifi_config_1.WifiConfig.encode(message.wifiConfig, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiGetConfigResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.wifiConfig = wifi_config_1.WifiConfig.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            wifiConfig: isSet(object.wifiConfig)
                ? wifi_config_1.WifiConfig.fromJSON(object.wifiConfig)
                : isSet(object.wifi_config)
                    ? wifi_config_1.WifiConfig.fromJSON(object.wifi_config)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.wifiConfig !== undefined) {
            obj.wifiConfig = wifi_config_1.WifiConfig.toJSON(message.wifiConfig);
        }
        return obj;
    },
    create(base) {
        return exports.WifiGetConfigResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseWifiGetConfigResponse();
        message.wifiConfig = (object.wifiConfig !== undefined && object.wifiConfig !== null)
            ? wifi_config_1.WifiConfig.fromPartial(object.wifiConfig)
            : undefined;
        return message;
    },
};
function createBaseWifiGetClientsRequest() {
    return {};
}
exports.WifiGetClientsRequest = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiGetClientsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.WifiGetClientsRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseWifiGetClientsRequest();
        return message;
    },
};
function createBaseWifiGetClientsResponse() {
    return { clients: [] };
}
exports.WifiGetClientsResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        for (const v of message.clients) {
            exports.WifiClient.encode(v, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiGetClientsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.clients.push(exports.WifiClient.decode(reader, reader.uint32()));
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            clients: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.clients) ? object.clients.map((e) => exports.WifiClient.fromJSON(e)) : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.clients) === null || _a === void 0 ? void 0 : _a.length) {
            obj.clients = message.clients.map((e) => exports.WifiClient.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.WifiGetClientsResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseWifiGetClientsResponse();
        message.clients = ((_a = object.clients) === null || _a === void 0 ? void 0 : _a.map((e) => exports.WifiClient.fromPartial(e))) || [];
        return message;
    },
};
function createBaseWifiGetHistoryResponse() {
    return { current: undefined, pingDropRate: [], pingLatencyMs: [] };
}
exports.WifiGetHistoryResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.current !== undefined) {
            writer.uint32(8).uint64(message.current);
        }
        writer.uint32(8010).fork();
        for (const v of message.pingDropRate) {
            writer.float(v);
        }
        writer.join();
        writer.uint32(8018).fork();
        for (const v of message.pingLatencyMs) {
            writer.float(v);
        }
        writer.join();
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiGetHistoryResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.current = longToNumber(reader.uint64());
                    continue;
                }
                case 1001: {
                    if (tag === 8013) {
                        message.pingDropRate.push(reader.float());
                        continue;
                    }
                    if (tag === 8010) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.pingDropRate.push(reader.float());
                        }
                        continue;
                    }
                    break;
                }
                case 1002: {
                    if (tag === 8021) {
                        message.pingLatencyMs.push(reader.float());
                        continue;
                    }
                    if (tag === 8018) {
                        const end2 = reader.uint32() + reader.pos;
                        while (reader.pos < end2) {
                            message.pingLatencyMs.push(reader.float());
                        }
                        continue;
                    }
                    break;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            current: isSet(object.current) ? globalThis.Number(object.current) : undefined,
            pingDropRate: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.pingDropRate)
                ? object.pingDropRate.map((e) => globalThis.Number(e))
                : globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.ping_drop_rate)
                    ? object.ping_drop_rate.map((e) => globalThis.Number(e))
                    : [],
            pingLatencyMs: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.pingLatencyMs)
                ? object.pingLatencyMs.map((e) => globalThis.Number(e))
                : globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.ping_latency_ms)
                    ? object.ping_latency_ms.map((e) => globalThis.Number(e))
                    : [],
        };
    },
    toJSON(message) {
        var _a, _b;
        const obj = {};
        if (message.current !== undefined) {
            obj.current = Math.round(message.current);
        }
        if ((_a = message.pingDropRate) === null || _a === void 0 ? void 0 : _a.length) {
            obj.pingDropRate = message.pingDropRate;
        }
        if ((_b = message.pingLatencyMs) === null || _b === void 0 ? void 0 : _b.length) {
            obj.pingLatencyMs = message.pingLatencyMs;
        }
        return obj;
    },
    create(base) {
        return exports.WifiGetHistoryResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseWifiGetHistoryResponse();
        message.current = (_a = object.current) !== null && _a !== void 0 ? _a : undefined;
        message.pingDropRate = ((_b = object.pingDropRate) === null || _b === void 0 ? void 0 : _b.map((e) => e)) || [];
        message.pingLatencyMs = ((_c = object.pingLatencyMs) === null || _c === void 0 ? void 0 : _c.map((e) => e)) || [];
        return message;
    },
};
function createBaseWifiNewClientConnectedEvent() {
    return { client: undefined };
}
exports.WifiNewClientConnectedEvent = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.client !== undefined) {
            exports.WifiClient.encode(message.client, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiNewClientConnectedEvent();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.client = exports.WifiClient.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { client: isSet(object.client) ? exports.WifiClient.fromJSON(object.client) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.client !== undefined) {
            obj.client = exports.WifiClient.toJSON(message.client);
        }
        return obj;
    },
    create(base) {
        return exports.WifiNewClientConnectedEvent.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseWifiNewClientConnectedEvent();
        message.client = (object.client !== undefined && object.client !== null)
            ? exports.WifiClient.fromPartial(object.client)
            : undefined;
        return message;
    },
};
function createBaseWifiClient() {
    return {
        name: undefined,
        macAddress: undefined,
        ipAddress: undefined,
        signalStrength: undefined,
        rxStats: undefined,
        txStats: undefined,
        associatedTimeS: undefined,
        modeStr: undefined,
        iface: undefined,
        snr: undefined,
        psmode: undefined,
    };
}
exports.WifiClient = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.name !== undefined) {
            writer.uint32(10).string(message.name);
        }
        if (message.macAddress !== undefined) {
            writer.uint32(18).string(message.macAddress);
        }
        if (message.ipAddress !== undefined) {
            writer.uint32(26).string(message.ipAddress);
        }
        if (message.signalStrength !== undefined) {
            writer.uint32(37).float(message.signalStrength);
        }
        if (message.rxStats !== undefined) {
            exports.WifiClient_RxStats.encode(message.rxStats, writer.uint32(42).fork()).join();
        }
        if (message.txStats !== undefined) {
            exports.WifiClient_TxStats.encode(message.txStats, writer.uint32(50).fork()).join();
        }
        if (message.associatedTimeS !== undefined) {
            writer.uint32(56).uint32(message.associatedTimeS);
        }
        if (message.modeStr !== undefined) {
            writer.uint32(66).string(message.modeStr);
        }
        if (message.iface !== undefined) {
            writer.uint32(72).int32(message.iface);
        }
        if (message.snr !== undefined) {
            writer.uint32(85).float(message.snr);
        }
        if (message.psmode !== undefined) {
            writer.uint32(88).int32(message.psmode);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiClient();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.macAddress = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.ipAddress = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 37) {
                        break;
                    }
                    message.signalStrength = reader.float();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.rxStats = exports.WifiClient_RxStats.decode(reader, reader.uint32());
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.txStats = exports.WifiClient_TxStats.decode(reader, reader.uint32());
                    continue;
                }
                case 7: {
                    if (tag !== 56) {
                        break;
                    }
                    message.associatedTimeS = reader.uint32();
                    continue;
                }
                case 8: {
                    if (tag !== 66) {
                        break;
                    }
                    message.modeStr = reader.string();
                    continue;
                }
                case 9: {
                    if (tag !== 72) {
                        break;
                    }
                    message.iface = reader.int32();
                    continue;
                }
                case 10: {
                    if (tag !== 85) {
                        break;
                    }
                    message.snr = reader.float();
                    continue;
                }
                case 11: {
                    if (tag !== 88) {
                        break;
                    }
                    message.psmode = reader.int32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            name: isSet(object.name) ? globalThis.String(object.name) : undefined,
            macAddress: isSet(object.macAddress)
                ? globalThis.String(object.macAddress)
                : isSet(object.mac_address)
                    ? globalThis.String(object.mac_address)
                    : undefined,
            ipAddress: isSet(object.ipAddress)
                ? globalThis.String(object.ipAddress)
                : isSet(object.ip_address)
                    ? globalThis.String(object.ip_address)
                    : undefined,
            signalStrength: isSet(object.signalStrength)
                ? globalThis.Number(object.signalStrength)
                : isSet(object.signal_strength)
                    ? globalThis.Number(object.signal_strength)
                    : undefined,
            rxStats: isSet(object.rxStats)
                ? exports.WifiClient_RxStats.fromJSON(object.rxStats)
                : isSet(object.rx_stats)
                    ? exports.WifiClient_RxStats.fromJSON(object.rx_stats)
                    : undefined,
            txStats: isSet(object.txStats)
                ? exports.WifiClient_TxStats.fromJSON(object.txStats)
                : isSet(object.tx_stats)
                    ? exports.WifiClient_TxStats.fromJSON(object.tx_stats)
                    : undefined,
            associatedTimeS: isSet(object.associatedTimeS)
                ? globalThis.Number(object.associatedTimeS)
                : isSet(object.associated_time_s)
                    ? globalThis.Number(object.associated_time_s)
                    : undefined,
            modeStr: isSet(object.modeStr)
                ? globalThis.String(object.modeStr)
                : isSet(object.mode_str)
                    ? globalThis.String(object.mode_str)
                    : undefined,
            iface: isSet(object.iface) ? wifiClient_InterfaceFromJSON(object.iface) : undefined,
            snr: isSet(object.snr) ? globalThis.Number(object.snr) : undefined,
            psmode: isSet(object.psmode) ? globalThis.Number(object.psmode) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== undefined) {
            obj.name = message.name;
        }
        if (message.macAddress !== undefined) {
            obj.macAddress = message.macAddress;
        }
        if (message.ipAddress !== undefined) {
            obj.ipAddress = message.ipAddress;
        }
        if (message.signalStrength !== undefined) {
            obj.signalStrength = message.signalStrength;
        }
        if (message.rxStats !== undefined) {
            obj.rxStats = exports.WifiClient_RxStats.toJSON(message.rxStats);
        }
        if (message.txStats !== undefined) {
            obj.txStats = exports.WifiClient_TxStats.toJSON(message.txStats);
        }
        if (message.associatedTimeS !== undefined) {
            obj.associatedTimeS = Math.round(message.associatedTimeS);
        }
        if (message.modeStr !== undefined) {
            obj.modeStr = message.modeStr;
        }
        if (message.iface !== undefined) {
            obj.iface = wifiClient_InterfaceToJSON(message.iface);
        }
        if (message.snr !== undefined) {
            obj.snr = message.snr;
        }
        if (message.psmode !== undefined) {
            obj.psmode = Math.round(message.psmode);
        }
        return obj;
    },
    create(base) {
        return exports.WifiClient.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        const message = createBaseWifiClient();
        message.name = (_a = object.name) !== null && _a !== void 0 ? _a : undefined;
        message.macAddress = (_b = object.macAddress) !== null && _b !== void 0 ? _b : undefined;
        message.ipAddress = (_c = object.ipAddress) !== null && _c !== void 0 ? _c : undefined;
        message.signalStrength = (_d = object.signalStrength) !== null && _d !== void 0 ? _d : undefined;
        message.rxStats = (object.rxStats !== undefined && object.rxStats !== null)
            ? exports.WifiClient_RxStats.fromPartial(object.rxStats)
            : undefined;
        message.txStats = (object.txStats !== undefined && object.txStats !== null)
            ? exports.WifiClient_TxStats.fromPartial(object.txStats)
            : undefined;
        message.associatedTimeS = (_e = object.associatedTimeS) !== null && _e !== void 0 ? _e : undefined;
        message.modeStr = (_f = object.modeStr) !== null && _f !== void 0 ? _f : undefined;
        message.iface = (_g = object.iface) !== null && _g !== void 0 ? _g : undefined;
        message.snr = (_h = object.snr) !== null && _h !== void 0 ? _h : undefined;
        message.psmode = (_j = object.psmode) !== null && _j !== void 0 ? _j : undefined;
        return message;
    },
};
function createBaseWifiClient_RxStats() {
    return { bytes: undefined, countErrors: undefined, nss: undefined };
}
exports.WifiClient_RxStats = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.bytes !== undefined) {
            writer.uint32(8).uint64(message.bytes);
        }
        if (message.countErrors !== undefined) {
            writer.uint32(16).uint64(message.countErrors);
        }
        if (message.nss !== undefined) {
            writer.uint32(24).int32(message.nss);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiClient_RxStats();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.bytes = longToNumber(reader.uint64());
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.countErrors = longToNumber(reader.uint64());
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.nss = reader.int32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            bytes: isSet(object.bytes) ? globalThis.Number(object.bytes) : undefined,
            countErrors: isSet(object.countErrors)
                ? globalThis.Number(object.countErrors)
                : isSet(object.count_errors)
                    ? globalThis.Number(object.count_errors)
                    : undefined,
            nss: isSet(object.nss) ? globalThis.Number(object.nss) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.bytes !== undefined) {
            obj.bytes = Math.round(message.bytes);
        }
        if (message.countErrors !== undefined) {
            obj.countErrors = Math.round(message.countErrors);
        }
        if (message.nss !== undefined) {
            obj.nss = Math.round(message.nss);
        }
        return obj;
    },
    create(base) {
        return exports.WifiClient_RxStats.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseWifiClient_RxStats();
        message.bytes = (_a = object.bytes) !== null && _a !== void 0 ? _a : undefined;
        message.countErrors = (_b = object.countErrors) !== null && _b !== void 0 ? _b : undefined;
        message.nss = (_c = object.nss) !== null && _c !== void 0 ? _c : undefined;
        return message;
    },
};
function createBaseWifiClient_TxStats() {
    return { bytes: undefined, successBytes: undefined, nss: undefined };
}
exports.WifiClient_TxStats = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.bytes !== undefined) {
            writer.uint32(8).uint64(message.bytes);
        }
        if (message.successBytes !== undefined) {
            writer.uint32(16).uint64(message.successBytes);
        }
        if (message.nss !== undefined) {
            writer.uint32(24).int32(message.nss);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiClient_TxStats();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.bytes = longToNumber(reader.uint64());
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.successBytes = longToNumber(reader.uint64());
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.nss = reader.int32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            bytes: isSet(object.bytes) ? globalThis.Number(object.bytes) : undefined,
            successBytes: isSet(object.successBytes)
                ? globalThis.Number(object.successBytes)
                : isSet(object.success_bytes)
                    ? globalThis.Number(object.success_bytes)
                    : undefined,
            nss: isSet(object.nss) ? globalThis.Number(object.nss) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.bytes !== undefined) {
            obj.bytes = Math.round(message.bytes);
        }
        if (message.successBytes !== undefined) {
            obj.successBytes = Math.round(message.successBytes);
        }
        if (message.nss !== undefined) {
            obj.nss = Math.round(message.nss);
        }
        return obj;
    },
    create(base) {
        return exports.WifiClient_TxStats.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseWifiClient_TxStats();
        message.bytes = (_a = object.bytes) !== null && _a !== void 0 ? _a : undefined;
        message.successBytes = (_b = object.successBytes) !== null && _b !== void 0 ? _b : undefined;
        message.nss = (_c = object.nss) !== null && _c !== void 0 ? _c : undefined;
        return message;
    },
};
function createBaseWifiSetupRequest() {
    return { skip: undefined, networkName: undefined, networkPassword: undefined };
}
exports.WifiSetupRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.skip !== undefined) {
            writer.uint32(8).bool(message.skip);
        }
        if (message.networkName !== undefined) {
            writer.uint32(18).string(message.networkName);
        }
        if (message.networkPassword !== undefined) {
            writer.uint32(26).string(message.networkPassword);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiSetupRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.skip = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.networkName = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.networkPassword = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            skip: isSet(object.skip) ? globalThis.Boolean(object.skip) : undefined,
            networkName: isSet(object.networkName)
                ? globalThis.String(object.networkName)
                : isSet(object.network_name)
                    ? globalThis.String(object.network_name)
                    : undefined,
            networkPassword: isSet(object.networkPassword)
                ? globalThis.String(object.networkPassword)
                : isSet(object.network_password)
                    ? globalThis.String(object.network_password)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.skip !== undefined) {
            obj.skip = message.skip;
        }
        if (message.networkName !== undefined) {
            obj.networkName = message.networkName;
        }
        if (message.networkPassword !== undefined) {
            obj.networkPassword = message.networkPassword;
        }
        return obj;
    },
    create(base) {
        return exports.WifiSetupRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseWifiSetupRequest();
        message.skip = (_a = object.skip) !== null && _a !== void 0 ? _a : undefined;
        message.networkName = (_b = object.networkName) !== null && _b !== void 0 ? _b : undefined;
        message.networkPassword = (_c = object.networkPassword) !== null && _c !== void 0 ? _c : undefined;
        return message;
    },
};
function createBaseWifiSetupResponse() {
    return {};
}
exports.WifiSetupResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiSetupResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.WifiSetupResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseWifiSetupResponse();
        return message;
    },
};
function createBaseWifiGetStatusResponse() {
    return {
        deviceInfo: undefined,
        deviceState: undefined,
        captivePortalEnabled: undefined,
        clients: [],
        serialNumber: undefined,
        sku: undefined,
        ipv4WanAddress: undefined,
        pingDropRate: undefined,
        pingLatencyMs: undefined,
    };
}
exports.WifiGetStatusResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.deviceInfo !== undefined) {
            common_1.DeviceInfo.encode(message.deviceInfo, writer.uint32(26).fork()).join();
        }
        if (message.deviceState !== undefined) {
            common_1.DeviceState.encode(message.deviceState, writer.uint32(34).fork()).join();
        }
        if (message.captivePortalEnabled !== undefined) {
            writer.uint32(8).bool(message.captivePortalEnabled);
        }
        for (const v of message.clients) {
            exports.WifiClient.encode(v, writer.uint32(18).fork()).join();
        }
        if (message.serialNumber !== undefined) {
            writer.uint32(8010).string(message.serialNumber);
        }
        if (message.sku !== undefined) {
            writer.uint32(8018).string(message.sku);
        }
        if (message.ipv4WanAddress !== undefined) {
            writer.uint32(8026).string(message.ipv4WanAddress);
        }
        if (message.pingDropRate !== undefined) {
            writer.uint32(8037).float(message.pingDropRate);
        }
        if (message.pingLatencyMs !== undefined) {
            writer.uint32(8045).float(message.pingLatencyMs);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiGetStatusResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.deviceInfo = common_1.DeviceInfo.decode(reader, reader.uint32());
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.deviceState = common_1.DeviceState.decode(reader, reader.uint32());
                    continue;
                }
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.captivePortalEnabled = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.clients.push(exports.WifiClient.decode(reader, reader.uint32()));
                    continue;
                }
                case 1001: {
                    if (tag !== 8010) {
                        break;
                    }
                    message.serialNumber = reader.string();
                    continue;
                }
                case 1002: {
                    if (tag !== 8018) {
                        break;
                    }
                    message.sku = reader.string();
                    continue;
                }
                case 1003: {
                    if (tag !== 8026) {
                        break;
                    }
                    message.ipv4WanAddress = reader.string();
                    continue;
                }
                case 1004: {
                    if (tag !== 8037) {
                        break;
                    }
                    message.pingDropRate = reader.float();
                    continue;
                }
                case 1005: {
                    if (tag !== 8045) {
                        break;
                    }
                    message.pingLatencyMs = reader.float();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            deviceInfo: isSet(object.deviceInfo)
                ? common_1.DeviceInfo.fromJSON(object.deviceInfo)
                : isSet(object.device_info)
                    ? common_1.DeviceInfo.fromJSON(object.device_info)
                    : undefined,
            deviceState: isSet(object.deviceState)
                ? common_1.DeviceState.fromJSON(object.deviceState)
                : isSet(object.device_state)
                    ? common_1.DeviceState.fromJSON(object.device_state)
                    : undefined,
            captivePortalEnabled: isSet(object.captivePortalEnabled)
                ? globalThis.Boolean(object.captivePortalEnabled)
                : isSet(object.captive_portal_enabled)
                    ? globalThis.Boolean(object.captive_portal_enabled)
                    : undefined,
            clients: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.clients) ? object.clients.map((e) => exports.WifiClient.fromJSON(e)) : [],
            serialNumber: isSet(object.serialNumber)
                ? globalThis.String(object.serialNumber)
                : isSet(object.serial_number)
                    ? globalThis.String(object.serial_number)
                    : undefined,
            sku: isSet(object.sku) ? globalThis.String(object.sku) : undefined,
            ipv4WanAddress: isSet(object.ipv4WanAddress)
                ? globalThis.String(object.ipv4WanAddress)
                : isSet(object.ipv4_wan_address)
                    ? globalThis.String(object.ipv4_wan_address)
                    : undefined,
            pingDropRate: isSet(object.pingDropRate)
                ? globalThis.Number(object.pingDropRate)
                : isSet(object.ping_drop_rate)
                    ? globalThis.Number(object.ping_drop_rate)
                    : undefined,
            pingLatencyMs: isSet(object.pingLatencyMs)
                ? globalThis.Number(object.pingLatencyMs)
                : isSet(object.ping_latency_ms)
                    ? globalThis.Number(object.ping_latency_ms)
                    : undefined,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.deviceInfo !== undefined) {
            obj.deviceInfo = common_1.DeviceInfo.toJSON(message.deviceInfo);
        }
        if (message.deviceState !== undefined) {
            obj.deviceState = common_1.DeviceState.toJSON(message.deviceState);
        }
        if (message.captivePortalEnabled !== undefined) {
            obj.captivePortalEnabled = message.captivePortalEnabled;
        }
        if ((_a = message.clients) === null || _a === void 0 ? void 0 : _a.length) {
            obj.clients = message.clients.map((e) => exports.WifiClient.toJSON(e));
        }
        if (message.serialNumber !== undefined) {
            obj.serialNumber = message.serialNumber;
        }
        if (message.sku !== undefined) {
            obj.sku = message.sku;
        }
        if (message.ipv4WanAddress !== undefined) {
            obj.ipv4WanAddress = message.ipv4WanAddress;
        }
        if (message.pingDropRate !== undefined) {
            obj.pingDropRate = message.pingDropRate;
        }
        if (message.pingLatencyMs !== undefined) {
            obj.pingLatencyMs = message.pingLatencyMs;
        }
        return obj;
    },
    create(base) {
        return exports.WifiGetStatusResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g;
        const message = createBaseWifiGetStatusResponse();
        message.deviceInfo = (object.deviceInfo !== undefined && object.deviceInfo !== null)
            ? common_1.DeviceInfo.fromPartial(object.deviceInfo)
            : undefined;
        message.deviceState = (object.deviceState !== undefined && object.deviceState !== null)
            ? common_1.DeviceState.fromPartial(object.deviceState)
            : undefined;
        message.captivePortalEnabled = (_a = object.captivePortalEnabled) !== null && _a !== void 0 ? _a : undefined;
        message.clients = ((_b = object.clients) === null || _b === void 0 ? void 0 : _b.map((e) => exports.WifiClient.fromPartial(e))) || [];
        message.serialNumber = (_c = object.serialNumber) !== null && _c !== void 0 ? _c : undefined;
        message.sku = (_d = object.sku) !== null && _d !== void 0 ? _d : undefined;
        message.ipv4WanAddress = (_e = object.ipv4WanAddress) !== null && _e !== void 0 ? _e : undefined;
        message.pingDropRate = (_f = object.pingDropRate) !== null && _f !== void 0 ? _f : undefined;
        message.pingLatencyMs = (_g = object.pingLatencyMs) !== null && _g !== void 0 ? _g : undefined;
        return message;
    },
};
function createBaseWifiAuthenticateRequest() {
    return { challenge: undefined };
}
exports.WifiAuthenticateRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.challenge !== undefined) {
            common_1.SignedData.encode(message.challenge, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiAuthenticateRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.challenge = common_1.SignedData.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { challenge: isSet(object.challenge) ? common_1.SignedData.fromJSON(object.challenge) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.challenge !== undefined) {
            obj.challenge = common_1.SignedData.toJSON(message.challenge);
        }
        return obj;
    },
    create(base) {
        return exports.WifiAuthenticateRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseWifiAuthenticateRequest();
        message.challenge = (object.challenge !== undefined && object.challenge !== null)
            ? common_1.SignedData.fromPartial(object.challenge)
            : undefined;
        return message;
    },
};
function createBaseWifiAuthenticateResponse() {
    return { wifi: undefined, dish: undefined };
}
exports.WifiAuthenticateResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.wifi !== undefined) {
            common_1.ChallengeResponse.encode(message.wifi, writer.uint32(10).fork()).join();
        }
        if (message.dish !== undefined) {
            common_1.ChallengeResponse.encode(message.dish, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiAuthenticateResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.wifi = common_1.ChallengeResponse.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.dish = common_1.ChallengeResponse.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            wifi: isSet(object.wifi) ? common_1.ChallengeResponse.fromJSON(object.wifi) : undefined,
            dish: isSet(object.dish) ? common_1.ChallengeResponse.fromJSON(object.dish) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.wifi !== undefined) {
            obj.wifi = common_1.ChallengeResponse.toJSON(message.wifi);
        }
        if (message.dish !== undefined) {
            obj.dish = common_1.ChallengeResponse.toJSON(message.dish);
        }
        return obj;
    },
    create(base) {
        return exports.WifiAuthenticateResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseWifiAuthenticateResponse();
        message.wifi = (object.wifi !== undefined && object.wifi !== null)
            ? common_1.ChallengeResponse.fromPartial(object.wifi)
            : undefined;
        message.dish = (object.dish !== undefined && object.dish !== null)
            ? common_1.ChallengeResponse.fromPartial(object.dish)
            : undefined;
        return message;
    },
};
function createBaseWifiAccountBondingEvent() {
    return { dishId: undefined };
}
exports.WifiAccountBondingEvent = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.dishId !== undefined) {
            writer.uint32(10).string(message.dishId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiAccountBondingEvent();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.dishId = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            dishId: isSet(object.dishId)
                ? globalThis.String(object.dishId)
                : isSet(object.dish_id)
                    ? globalThis.String(object.dish_id)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.dishId !== undefined) {
            obj.dishId = message.dishId;
        }
        return obj;
    },
    create(base) {
        return exports.WifiAccountBondingEvent.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseWifiAccountBondingEvent();
        message.dishId = (_a = object.dishId) !== null && _a !== void 0 ? _a : undefined;
        return message;
    },
};
function createBasePingMetrics() {
    return {
        latencyMeanMs: undefined,
        latencyStddevMs: undefined,
        latencyMeanMs5m: undefined,
        latencyMeanMs1h: undefined,
        latencyMeanMs1d: undefined,
        dropRate: undefined,
        dropRate5m: undefined,
        dropRate1h: undefined,
        dropRate1d: undefined,
        secondsSinceLastSuccess: undefined,
        secondsSinceLast1sOutage: undefined,
        secondsSinceLast2sOutage: undefined,
        secondsSinceLast5sOutage: undefined,
        secondsSinceLast15sOutage: undefined,
        secondsSinceLast60sOutage: undefined,
        secondsSinceLast300sOutage: undefined,
        happyHours1s1d: undefined,
        happyHours2s1d: undefined,
        happyHours5s1d: undefined,
    };
}
exports.PingMetrics = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.latencyMeanMs !== undefined) {
            writer.uint32(13).float(message.latencyMeanMs);
        }
        if (message.latencyStddevMs !== undefined) {
            writer.uint32(21).float(message.latencyStddevMs);
        }
        if (message.latencyMeanMs5m !== undefined) {
            writer.uint32(29).float(message.latencyMeanMs5m);
        }
        if (message.latencyMeanMs1h !== undefined) {
            writer.uint32(37).float(message.latencyMeanMs1h);
        }
        if (message.latencyMeanMs1d !== undefined) {
            writer.uint32(45).float(message.latencyMeanMs1d);
        }
        if (message.dropRate !== undefined) {
            writer.uint32(53).float(message.dropRate);
        }
        if (message.dropRate5m !== undefined) {
            writer.uint32(61).float(message.dropRate5m);
        }
        if (message.dropRate1h !== undefined) {
            writer.uint32(69).float(message.dropRate1h);
        }
        if (message.dropRate1d !== undefined) {
            writer.uint32(77).float(message.dropRate1d);
        }
        if (message.secondsSinceLastSuccess !== undefined) {
            writer.uint32(85).float(message.secondsSinceLastSuccess);
        }
        if (message.secondsSinceLast1sOutage !== undefined) {
            writer.uint32(93).float(message.secondsSinceLast1sOutage);
        }
        if (message.secondsSinceLast2sOutage !== undefined) {
            writer.uint32(125).float(message.secondsSinceLast2sOutage);
        }
        if (message.secondsSinceLast5sOutage !== undefined) {
            writer.uint32(101).float(message.secondsSinceLast5sOutage);
        }
        if (message.secondsSinceLast15sOutage !== undefined) {
            writer.uint32(149).float(message.secondsSinceLast15sOutage);
        }
        if (message.secondsSinceLast60sOutage !== undefined) {
            writer.uint32(157).float(message.secondsSinceLast60sOutage);
        }
        if (message.secondsSinceLast300sOutage !== undefined) {
            writer.uint32(165).float(message.secondsSinceLast300sOutage);
        }
        if (message.happyHours1s1d !== undefined) {
            writer.uint32(109).float(message.happyHours1s1d);
        }
        if (message.happyHours2s1d !== undefined) {
            writer.uint32(133).float(message.happyHours2s1d);
        }
        if (message.happyHours5s1d !== undefined) {
            writer.uint32(117).float(message.happyHours5s1d);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePingMetrics();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 13) {
                        break;
                    }
                    message.latencyMeanMs = reader.float();
                    continue;
                }
                case 2: {
                    if (tag !== 21) {
                        break;
                    }
                    message.latencyStddevMs = reader.float();
                    continue;
                }
                case 3: {
                    if (tag !== 29) {
                        break;
                    }
                    message.latencyMeanMs5m = reader.float();
                    continue;
                }
                case 4: {
                    if (tag !== 37) {
                        break;
                    }
                    message.latencyMeanMs1h = reader.float();
                    continue;
                }
                case 5: {
                    if (tag !== 45) {
                        break;
                    }
                    message.latencyMeanMs1d = reader.float();
                    continue;
                }
                case 6: {
                    if (tag !== 53) {
                        break;
                    }
                    message.dropRate = reader.float();
                    continue;
                }
                case 7: {
                    if (tag !== 61) {
                        break;
                    }
                    message.dropRate5m = reader.float();
                    continue;
                }
                case 8: {
                    if (tag !== 69) {
                        break;
                    }
                    message.dropRate1h = reader.float();
                    continue;
                }
                case 9: {
                    if (tag !== 77) {
                        break;
                    }
                    message.dropRate1d = reader.float();
                    continue;
                }
                case 10: {
                    if (tag !== 85) {
                        break;
                    }
                    message.secondsSinceLastSuccess = reader.float();
                    continue;
                }
                case 11: {
                    if (tag !== 93) {
                        break;
                    }
                    message.secondsSinceLast1sOutage = reader.float();
                    continue;
                }
                case 15: {
                    if (tag !== 125) {
                        break;
                    }
                    message.secondsSinceLast2sOutage = reader.float();
                    continue;
                }
                case 12: {
                    if (tag !== 101) {
                        break;
                    }
                    message.secondsSinceLast5sOutage = reader.float();
                    continue;
                }
                case 18: {
                    if (tag !== 149) {
                        break;
                    }
                    message.secondsSinceLast15sOutage = reader.float();
                    continue;
                }
                case 19: {
                    if (tag !== 157) {
                        break;
                    }
                    message.secondsSinceLast60sOutage = reader.float();
                    continue;
                }
                case 20: {
                    if (tag !== 165) {
                        break;
                    }
                    message.secondsSinceLast300sOutage = reader.float();
                    continue;
                }
                case 13: {
                    if (tag !== 109) {
                        break;
                    }
                    message.happyHours1s1d = reader.float();
                    continue;
                }
                case 16: {
                    if (tag !== 133) {
                        break;
                    }
                    message.happyHours2s1d = reader.float();
                    continue;
                }
                case 14: {
                    if (tag !== 117) {
                        break;
                    }
                    message.happyHours5s1d = reader.float();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            latencyMeanMs: isSet(object.latencyMeanMs)
                ? globalThis.Number(object.latencyMeanMs)
                : isSet(object.latency_mean_ms)
                    ? globalThis.Number(object.latency_mean_ms)
                    : undefined,
            latencyStddevMs: isSet(object.latencyStddevMs)
                ? globalThis.Number(object.latencyStddevMs)
                : isSet(object.latency_stddev_ms)
                    ? globalThis.Number(object.latency_stddev_ms)
                    : undefined,
            latencyMeanMs5m: isSet(object.latencyMeanMs5m)
                ? globalThis.Number(object.latencyMeanMs5m)
                : isSet(object.latency_mean_ms_5m)
                    ? globalThis.Number(object.latency_mean_ms_5m)
                    : undefined,
            latencyMeanMs1h: isSet(object.latencyMeanMs1h)
                ? globalThis.Number(object.latencyMeanMs1h)
                : isSet(object.latency_mean_ms_1h)
                    ? globalThis.Number(object.latency_mean_ms_1h)
                    : undefined,
            latencyMeanMs1d: isSet(object.latencyMeanMs1d)
                ? globalThis.Number(object.latencyMeanMs1d)
                : isSet(object.latency_mean_ms_1d)
                    ? globalThis.Number(object.latency_mean_ms_1d)
                    : undefined,
            dropRate: isSet(object.dropRate)
                ? globalThis.Number(object.dropRate)
                : isSet(object.drop_rate)
                    ? globalThis.Number(object.drop_rate)
                    : undefined,
            dropRate5m: isSet(object.dropRate5m)
                ? globalThis.Number(object.dropRate5m)
                : isSet(object.drop_rate_5m)
                    ? globalThis.Number(object.drop_rate_5m)
                    : undefined,
            dropRate1h: isSet(object.dropRate1h)
                ? globalThis.Number(object.dropRate1h)
                : isSet(object.drop_rate_1h)
                    ? globalThis.Number(object.drop_rate_1h)
                    : undefined,
            dropRate1d: isSet(object.dropRate1d)
                ? globalThis.Number(object.dropRate1d)
                : isSet(object.drop_rate_1d)
                    ? globalThis.Number(object.drop_rate_1d)
                    : undefined,
            secondsSinceLastSuccess: isSet(object.secondsSinceLastSuccess)
                ? globalThis.Number(object.secondsSinceLastSuccess)
                : isSet(object.seconds_since_last_success)
                    ? globalThis.Number(object.seconds_since_last_success)
                    : undefined,
            secondsSinceLast1sOutage: isSet(object.secondsSinceLast1sOutage)
                ? globalThis.Number(object.secondsSinceLast1sOutage)
                : isSet(object.seconds_since_last_1s_outage)
                    ? globalThis.Number(object.seconds_since_last_1s_outage)
                    : undefined,
            secondsSinceLast2sOutage: isSet(object.secondsSinceLast2sOutage)
                ? globalThis.Number(object.secondsSinceLast2sOutage)
                : isSet(object.seconds_since_last_2s_outage)
                    ? globalThis.Number(object.seconds_since_last_2s_outage)
                    : undefined,
            secondsSinceLast5sOutage: isSet(object.secondsSinceLast5sOutage)
                ? globalThis.Number(object.secondsSinceLast5sOutage)
                : isSet(object.seconds_since_last_5s_outage)
                    ? globalThis.Number(object.seconds_since_last_5s_outage)
                    : undefined,
            secondsSinceLast15sOutage: isSet(object.secondsSinceLast15sOutage)
                ? globalThis.Number(object.secondsSinceLast15sOutage)
                : isSet(object.seconds_since_last_15s_outage)
                    ? globalThis.Number(object.seconds_since_last_15s_outage)
                    : undefined,
            secondsSinceLast60sOutage: isSet(object.secondsSinceLast60sOutage)
                ? globalThis.Number(object.secondsSinceLast60sOutage)
                : isSet(object.seconds_since_last_60s_outage)
                    ? globalThis.Number(object.seconds_since_last_60s_outage)
                    : undefined,
            secondsSinceLast300sOutage: isSet(object.secondsSinceLast300sOutage)
                ? globalThis.Number(object.secondsSinceLast300sOutage)
                : isSet(object.seconds_since_last_300s_outage)
                    ? globalThis.Number(object.seconds_since_last_300s_outage)
                    : undefined,
            happyHours1s1d: isSet(object.happyHours1s1d)
                ? globalThis.Number(object.happyHours1s1d)
                : isSet(object.happy_hours_1s_1d)
                    ? globalThis.Number(object.happy_hours_1s_1d)
                    : undefined,
            happyHours2s1d: isSet(object.happyHours2s1d)
                ? globalThis.Number(object.happyHours2s1d)
                : isSet(object.happy_hours_2s_1d)
                    ? globalThis.Number(object.happy_hours_2s_1d)
                    : undefined,
            happyHours5s1d: isSet(object.happyHours5s1d)
                ? globalThis.Number(object.happyHours5s1d)
                : isSet(object.happy_hours_5s_1d)
                    ? globalThis.Number(object.happy_hours_5s_1d)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.latencyMeanMs !== undefined) {
            obj.latencyMeanMs = message.latencyMeanMs;
        }
        if (message.latencyStddevMs !== undefined) {
            obj.latencyStddevMs = message.latencyStddevMs;
        }
        if (message.latencyMeanMs5m !== undefined) {
            obj.latencyMeanMs5m = message.latencyMeanMs5m;
        }
        if (message.latencyMeanMs1h !== undefined) {
            obj.latencyMeanMs1h = message.latencyMeanMs1h;
        }
        if (message.latencyMeanMs1d !== undefined) {
            obj.latencyMeanMs1d = message.latencyMeanMs1d;
        }
        if (message.dropRate !== undefined) {
            obj.dropRate = message.dropRate;
        }
        if (message.dropRate5m !== undefined) {
            obj.dropRate5m = message.dropRate5m;
        }
        if (message.dropRate1h !== undefined) {
            obj.dropRate1h = message.dropRate1h;
        }
        if (message.dropRate1d !== undefined) {
            obj.dropRate1d = message.dropRate1d;
        }
        if (message.secondsSinceLastSuccess !== undefined) {
            obj.secondsSinceLastSuccess = message.secondsSinceLastSuccess;
        }
        if (message.secondsSinceLast1sOutage !== undefined) {
            obj.secondsSinceLast1sOutage = message.secondsSinceLast1sOutage;
        }
        if (message.secondsSinceLast2sOutage !== undefined) {
            obj.secondsSinceLast2sOutage = message.secondsSinceLast2sOutage;
        }
        if (message.secondsSinceLast5sOutage !== undefined) {
            obj.secondsSinceLast5sOutage = message.secondsSinceLast5sOutage;
        }
        if (message.secondsSinceLast15sOutage !== undefined) {
            obj.secondsSinceLast15sOutage = message.secondsSinceLast15sOutage;
        }
        if (message.secondsSinceLast60sOutage !== undefined) {
            obj.secondsSinceLast60sOutage = message.secondsSinceLast60sOutage;
        }
        if (message.secondsSinceLast300sOutage !== undefined) {
            obj.secondsSinceLast300sOutage = message.secondsSinceLast300sOutage;
        }
        if (message.happyHours1s1d !== undefined) {
            obj.happyHours1s1d = message.happyHours1s1d;
        }
        if (message.happyHours2s1d !== undefined) {
            obj.happyHours2s1d = message.happyHours2s1d;
        }
        if (message.happyHours5s1d !== undefined) {
            obj.happyHours5s1d = message.happyHours5s1d;
        }
        return obj;
    },
    create(base) {
        return exports.PingMetrics.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u;
        const message = createBasePingMetrics();
        message.latencyMeanMs = (_a = object.latencyMeanMs) !== null && _a !== void 0 ? _a : undefined;
        message.latencyStddevMs = (_b = object.latencyStddevMs) !== null && _b !== void 0 ? _b : undefined;
        message.latencyMeanMs5m = (_c = object.latencyMeanMs5m) !== null && _c !== void 0 ? _c : undefined;
        message.latencyMeanMs1h = (_d = object.latencyMeanMs1h) !== null && _d !== void 0 ? _d : undefined;
        message.latencyMeanMs1d = (_e = object.latencyMeanMs1d) !== null && _e !== void 0 ? _e : undefined;
        message.dropRate = (_f = object.dropRate) !== null && _f !== void 0 ? _f : undefined;
        message.dropRate5m = (_g = object.dropRate5m) !== null && _g !== void 0 ? _g : undefined;
        message.dropRate1h = (_h = object.dropRate1h) !== null && _h !== void 0 ? _h : undefined;
        message.dropRate1d = (_j = object.dropRate1d) !== null && _j !== void 0 ? _j : undefined;
        message.secondsSinceLastSuccess = (_k = object.secondsSinceLastSuccess) !== null && _k !== void 0 ? _k : undefined;
        message.secondsSinceLast1sOutage = (_l = object.secondsSinceLast1sOutage) !== null && _l !== void 0 ? _l : undefined;
        message.secondsSinceLast2sOutage = (_m = object.secondsSinceLast2sOutage) !== null && _m !== void 0 ? _m : undefined;
        message.secondsSinceLast5sOutage = (_o = object.secondsSinceLast5sOutage) !== null && _o !== void 0 ? _o : undefined;
        message.secondsSinceLast15sOutage = (_p = object.secondsSinceLast15sOutage) !== null && _p !== void 0 ? _p : undefined;
        message.secondsSinceLast60sOutage = (_q = object.secondsSinceLast60sOutage) !== null && _q !== void 0 ? _q : undefined;
        message.secondsSinceLast300sOutage = (_r = object.secondsSinceLast300sOutage) !== null && _r !== void 0 ? _r : undefined;
        message.happyHours1s1d = (_s = object.happyHours1s1d) !== null && _s !== void 0 ? _s : undefined;
        message.happyHours2s1d = (_t = object.happyHours2s1d) !== null && _t !== void 0 ? _t : undefined;
        message.happyHours5s1d = (_u = object.happyHours5s1d) !== null && _u !== void 0 ? _u : undefined;
        return message;
    },
};
function createBaseWifiGetPingMetricsRequest() {
    return {};
}
exports.WifiGetPingMetricsRequest = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiGetPingMetricsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.WifiGetPingMetricsRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseWifiGetPingMetricsRequest();
        return message;
    },
};
function createBaseWifiGetPingMetricsResponse() {
    return { internet: undefined };
}
exports.WifiGetPingMetricsResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.internet !== undefined) {
            exports.PingMetrics.encode(message.internet, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiGetPingMetricsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.internet = exports.PingMetrics.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { internet: isSet(object.internet) ? exports.PingMetrics.fromJSON(object.internet) : undefined };
    },
    toJSON(message) {
        const obj = {};
        if (message.internet !== undefined) {
            obj.internet = exports.PingMetrics.toJSON(message.internet);
        }
        return obj;
    },
    create(base) {
        return exports.WifiGetPingMetricsResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        const message = createBaseWifiGetPingMetricsResponse();
        message.internet = (object.internet !== undefined && object.internet !== null)
            ? exports.PingMetrics.fromPartial(object.internet)
            : undefined;
        return message;
    },
};
function createBaseWifiGetDiagnosticsRequest() {
    return {};
}
exports.WifiGetDiagnosticsRequest = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiGetDiagnosticsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.WifiGetDiagnosticsRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(_) {
        const message = createBaseWifiGetDiagnosticsRequest();
        return message;
    },
};
function createBaseWifiGetDiagnosticsResponse() {
    return { networkScan: undefined, wifiNetworks: [] };
}
exports.WifiGetDiagnosticsResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.networkScan !== undefined) {
            exports.WifiScanResults.encode(message.networkScan, writer.uint32(10).fork()).join();
        }
        for (const v of message.wifiNetworks) {
            exports.WifiNetwork.encode(v, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiGetDiagnosticsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.networkScan = exports.WifiScanResults.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.wifiNetworks.push(exports.WifiNetwork.decode(reader, reader.uint32()));
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            networkScan: isSet(object.networkScan)
                ? exports.WifiScanResults.fromJSON(object.networkScan)
                : isSet(object.network_scan)
                    ? exports.WifiScanResults.fromJSON(object.network_scan)
                    : undefined,
            wifiNetworks: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.wifiNetworks)
                ? object.wifiNetworks.map((e) => exports.WifiNetwork.fromJSON(e))
                : globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.wifi_networks)
                    ? object.wifi_networks.map((e) => exports.WifiNetwork.fromJSON(e))
                    : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.networkScan !== undefined) {
            obj.networkScan = exports.WifiScanResults.toJSON(message.networkScan);
        }
        if ((_a = message.wifiNetworks) === null || _a === void 0 ? void 0 : _a.length) {
            obj.wifiNetworks = message.wifiNetworks.map((e) => exports.WifiNetwork.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.WifiGetDiagnosticsResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseWifiGetDiagnosticsResponse();
        message.networkScan = (object.networkScan !== undefined && object.networkScan !== null)
            ? exports.WifiScanResults.fromPartial(object.networkScan)
            : undefined;
        message.wifiNetworks = ((_a = object.wifiNetworks) === null || _a === void 0 ? void 0 : _a.map((e) => exports.WifiNetwork.fromPartial(e))) || [];
        return message;
    },
};
function createBaseWifiScanResults() {
    return { networks: [] };
}
exports.WifiScanResults = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        for (const v of message.networks) {
            exports.WifiScanResults_Network.encode(v, writer.uint32(10).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiScanResults();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.networks.push(exports.WifiScanResults_Network.decode(reader, reader.uint32()));
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            networks: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.networks)
                ? object.networks.map((e) => exports.WifiScanResults_Network.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.networks) === null || _a === void 0 ? void 0 : _a.length) {
            obj.networks = message.networks.map((e) => exports.WifiScanResults_Network.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.WifiScanResults.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseWifiScanResults();
        message.networks = ((_a = object.networks) === null || _a === void 0 ? void 0 : _a.map((e) => exports.WifiScanResults_Network.fromPartial(e))) || [];
        return message;
    },
};
function createBaseWifiScanResults_Network() {
    return {
        source: undefined,
        ssid: undefined,
        bssid: undefined,
        frequencyGhz: undefined,
        channel: undefined,
        signalLevelDbm: undefined,
        noiseLevelDbm: undefined,
        hasEncryptionKey: undefined,
        phyModeStr: undefined,
    };
}
exports.WifiScanResults_Network = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.source !== undefined) {
            writer.uint32(8).int32(message.source);
        }
        if (message.ssid !== undefined) {
            writer.uint32(18).string(message.ssid);
        }
        if (message.bssid !== undefined) {
            writer.uint32(26).string(message.bssid);
        }
        if (message.frequencyGhz !== undefined) {
            writer.uint32(34).string(message.frequencyGhz);
        }
        if (message.channel !== undefined) {
            writer.uint32(40).int32(message.channel);
        }
        if (message.signalLevelDbm !== undefined) {
            writer.uint32(48).int32(message.signalLevelDbm);
        }
        if (message.noiseLevelDbm !== undefined) {
            writer.uint32(56).int32(message.noiseLevelDbm);
        }
        if (message.hasEncryptionKey !== undefined) {
            writer.uint32(64).bool(message.hasEncryptionKey);
        }
        if (message.phyModeStr !== undefined) {
            writer.uint32(74).string(message.phyModeStr);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiScanResults_Network();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.source = reader.int32();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.ssid = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.bssid = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.frequencyGhz = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 40) {
                        break;
                    }
                    message.channel = reader.int32();
                    continue;
                }
                case 6: {
                    if (tag !== 48) {
                        break;
                    }
                    message.signalLevelDbm = reader.int32();
                    continue;
                }
                case 7: {
                    if (tag !== 56) {
                        break;
                    }
                    message.noiseLevelDbm = reader.int32();
                    continue;
                }
                case 8: {
                    if (tag !== 64) {
                        break;
                    }
                    message.hasEncryptionKey = reader.bool();
                    continue;
                }
                case 9: {
                    if (tag !== 74) {
                        break;
                    }
                    message.phyModeStr = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            source: isSet(object.source) ? wifiScanResults_Network_SourceFromJSON(object.source) : undefined,
            ssid: isSet(object.ssid) ? globalThis.String(object.ssid) : undefined,
            bssid: isSet(object.bssid) ? globalThis.String(object.bssid) : undefined,
            frequencyGhz: isSet(object.frequencyGhz)
                ? globalThis.String(object.frequencyGhz)
                : isSet(object.frequency_ghz)
                    ? globalThis.String(object.frequency_ghz)
                    : undefined,
            channel: isSet(object.channel) ? globalThis.Number(object.channel) : undefined,
            signalLevelDbm: isSet(object.signalLevelDbm)
                ? globalThis.Number(object.signalLevelDbm)
                : isSet(object.signal_level_dbm)
                    ? globalThis.Number(object.signal_level_dbm)
                    : undefined,
            noiseLevelDbm: isSet(object.noiseLevelDbm)
                ? globalThis.Number(object.noiseLevelDbm)
                : isSet(object.noise_level_dbm)
                    ? globalThis.Number(object.noise_level_dbm)
                    : undefined,
            hasEncryptionKey: isSet(object.hasEncryptionKey)
                ? globalThis.Boolean(object.hasEncryptionKey)
                : isSet(object.has_encryption_key)
                    ? globalThis.Boolean(object.has_encryption_key)
                    : undefined,
            phyModeStr: isSet(object.phyModeStr)
                ? globalThis.String(object.phyModeStr)
                : isSet(object.phy_mode_str)
                    ? globalThis.String(object.phy_mode_str)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.source !== undefined) {
            obj.source = wifiScanResults_Network_SourceToJSON(message.source);
        }
        if (message.ssid !== undefined) {
            obj.ssid = message.ssid;
        }
        if (message.bssid !== undefined) {
            obj.bssid = message.bssid;
        }
        if (message.frequencyGhz !== undefined) {
            obj.frequencyGhz = message.frequencyGhz;
        }
        if (message.channel !== undefined) {
            obj.channel = Math.round(message.channel);
        }
        if (message.signalLevelDbm !== undefined) {
            obj.signalLevelDbm = Math.round(message.signalLevelDbm);
        }
        if (message.noiseLevelDbm !== undefined) {
            obj.noiseLevelDbm = Math.round(message.noiseLevelDbm);
        }
        if (message.hasEncryptionKey !== undefined) {
            obj.hasEncryptionKey = message.hasEncryptionKey;
        }
        if (message.phyModeStr !== undefined) {
            obj.phyModeStr = message.phyModeStr;
        }
        return obj;
    },
    create(base) {
        return exports.WifiScanResults_Network.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        const message = createBaseWifiScanResults_Network();
        message.source = (_a = object.source) !== null && _a !== void 0 ? _a : undefined;
        message.ssid = (_b = object.ssid) !== null && _b !== void 0 ? _b : undefined;
        message.bssid = (_c = object.bssid) !== null && _c !== void 0 ? _c : undefined;
        message.frequencyGhz = (_d = object.frequencyGhz) !== null && _d !== void 0 ? _d : undefined;
        message.channel = (_e = object.channel) !== null && _e !== void 0 ? _e : undefined;
        message.signalLevelDbm = (_f = object.signalLevelDbm) !== null && _f !== void 0 ? _f : undefined;
        message.noiseLevelDbm = (_g = object.noiseLevelDbm) !== null && _g !== void 0 ? _g : undefined;
        message.hasEncryptionKey = (_h = object.hasEncryptionKey) !== null && _h !== void 0 ? _h : undefined;
        message.phyModeStr = (_j = object.phyModeStr) !== null && _j !== void 0 ? _j : undefined;
        return message;
    },
};
function createBaseWifiNetwork() {
    return { band: undefined, ssid: undefined, channel: undefined, encryptionTypeStr: undefined };
}
exports.WifiNetwork = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.band !== undefined) {
            writer.uint32(8).int32(message.band);
        }
        if (message.ssid !== undefined) {
            writer.uint32(18).string(message.ssid);
        }
        if (message.channel !== undefined) {
            writer.uint32(24).uint32(message.channel);
        }
        if (message.encryptionTypeStr !== undefined) {
            writer.uint32(34).string(message.encryptionTypeStr);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiNetwork();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.band = reader.int32();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.ssid = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.channel = reader.uint32();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.encryptionTypeStr = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            band: isSet(object.band) ? wifiNetwork_BandFromJSON(object.band) : undefined,
            ssid: isSet(object.ssid) ? globalThis.String(object.ssid) : undefined,
            channel: isSet(object.channel) ? globalThis.Number(object.channel) : undefined,
            encryptionTypeStr: isSet(object.encryptionTypeStr)
                ? globalThis.String(object.encryptionTypeStr)
                : isSet(object.encryption_type_str)
                    ? globalThis.String(object.encryption_type_str)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.band !== undefined) {
            obj.band = wifiNetwork_BandToJSON(message.band);
        }
        if (message.ssid !== undefined) {
            obj.ssid = message.ssid;
        }
        if (message.channel !== undefined) {
            obj.channel = Math.round(message.channel);
        }
        if (message.encryptionTypeStr !== undefined) {
            obj.encryptionTypeStr = message.encryptionTypeStr;
        }
        return obj;
    },
    create(base) {
        return exports.WifiNetwork.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseWifiNetwork();
        message.band = (_a = object.band) !== null && _a !== void 0 ? _a : undefined;
        message.ssid = (_b = object.ssid) !== null && _b !== void 0 ? _b : undefined;
        message.channel = (_c = object.channel) !== null && _c !== void 0 ? _c : undefined;
        message.encryptionTypeStr = (_d = object.encryptionTypeStr) !== null && _d !== void 0 ? _d : undefined;
        return message;
    },
};
function createBaseWifiGetDiagnosticsResponse2() {
    return { id: "", hardwareVersion: "", softwareVersion: "", networks: [] };
}
exports.WifiGetDiagnosticsResponse2 = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.hardwareVersion !== "") {
            writer.uint32(18).string(message.hardwareVersion);
        }
        if (message.softwareVersion !== "") {
            writer.uint32(26).string(message.softwareVersion);
        }
        for (const v of message.networks) {
            exports.WifiGetDiagnosticsResponse2_Network.encode(v, writer.uint32(34).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiGetDiagnosticsResponse2();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.hardwareVersion = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.softwareVersion = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.networks.push(exports.WifiGetDiagnosticsResponse2_Network.decode(reader, reader.uint32()));
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? globalThis.String(object.id) : "",
            hardwareVersion: isSet(object.hardwareVersion)
                ? globalThis.String(object.hardwareVersion)
                : isSet(object.hardware_version)
                    ? globalThis.String(object.hardware_version)
                    : "",
            softwareVersion: isSet(object.softwareVersion)
                ? globalThis.String(object.softwareVersion)
                : isSet(object.software_version)
                    ? globalThis.String(object.software_version)
                    : "",
            networks: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.networks)
                ? object.networks.map((e) => exports.WifiGetDiagnosticsResponse2_Network.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.id !== "") {
            obj.id = message.id;
        }
        if (message.hardwareVersion !== "") {
            obj.hardwareVersion = message.hardwareVersion;
        }
        if (message.softwareVersion !== "") {
            obj.softwareVersion = message.softwareVersion;
        }
        if ((_a = message.networks) === null || _a === void 0 ? void 0 : _a.length) {
            obj.networks = message.networks.map((e) => exports.WifiGetDiagnosticsResponse2_Network.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.WifiGetDiagnosticsResponse2.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseWifiGetDiagnosticsResponse2();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : "";
        message.hardwareVersion = (_b = object.hardwareVersion) !== null && _b !== void 0 ? _b : "";
        message.softwareVersion = (_c = object.softwareVersion) !== null && _c !== void 0 ? _c : "";
        message.networks = ((_d = object.networks) === null || _d === void 0 ? void 0 : _d.map((e) => exports.WifiGetDiagnosticsResponse2_Network.fromPartial(e))) || [];
        return message;
    },
};
function createBaseWifiGetDiagnosticsResponse2_Network() {
    return { domain: "", ipv4: "", ipv6: [], clientsEthernet: 0, clients2ghz: 0, clients5ghz: 0 };
}
exports.WifiGetDiagnosticsResponse2_Network = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.domain !== "") {
            writer.uint32(10).string(message.domain);
        }
        if (message.ipv4 !== "") {
            writer.uint32(18).string(message.ipv4);
        }
        for (const v of message.ipv6) {
            writer.uint32(26).string(v);
        }
        if (message.clientsEthernet !== 0) {
            writer.uint32(80).uint32(message.clientsEthernet);
        }
        if (message.clients2ghz !== 0) {
            writer.uint32(88).uint32(message.clients2ghz);
        }
        if (message.clients5ghz !== 0) {
            writer.uint32(96).uint32(message.clients5ghz);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiGetDiagnosticsResponse2_Network();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.domain = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.ipv4 = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.ipv6.push(reader.string());
                    continue;
                }
                case 10: {
                    if (tag !== 80) {
                        break;
                    }
                    message.clientsEthernet = reader.uint32();
                    continue;
                }
                case 11: {
                    if (tag !== 88) {
                        break;
                    }
                    message.clients2ghz = reader.uint32();
                    continue;
                }
                case 12: {
                    if (tag !== 96) {
                        break;
                    }
                    message.clients5ghz = reader.uint32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            domain: isSet(object.domain) ? globalThis.String(object.domain) : "",
            ipv4: isSet(object.ipv4) ? globalThis.String(object.ipv4) : "",
            ipv6: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.ipv6) ? object.ipv6.map((e) => globalThis.String(e)) : [],
            clientsEthernet: isSet(object.clientsEthernet)
                ? globalThis.Number(object.clientsEthernet)
                : isSet(object.clients_ethernet)
                    ? globalThis.Number(object.clients_ethernet)
                    : 0,
            clients2ghz: isSet(object.clients2ghz)
                ? globalThis.Number(object.clients2ghz)
                : isSet(object.clients_2ghz)
                    ? globalThis.Number(object.clients_2ghz)
                    : 0,
            clients5ghz: isSet(object.clients5ghz)
                ? globalThis.Number(object.clients5ghz)
                : isSet(object.clients_5ghz)
                    ? globalThis.Number(object.clients_5ghz)
                    : 0,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.domain !== "") {
            obj.domain = message.domain;
        }
        if (message.ipv4 !== "") {
            obj.ipv4 = message.ipv4;
        }
        if ((_a = message.ipv6) === null || _a === void 0 ? void 0 : _a.length) {
            obj.ipv6 = message.ipv6;
        }
        if (message.clientsEthernet !== 0) {
            obj.clientsEthernet = Math.round(message.clientsEthernet);
        }
        if (message.clients2ghz !== 0) {
            obj.clients2ghz = Math.round(message.clients2ghz);
        }
        if (message.clients5ghz !== 0) {
            obj.clients5ghz = Math.round(message.clients5ghz);
        }
        return obj;
    },
    create(base) {
        return exports.WifiGetDiagnosticsResponse2_Network.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f;
        const message = createBaseWifiGetDiagnosticsResponse2_Network();
        message.domain = (_a = object.domain) !== null && _a !== void 0 ? _a : "";
        message.ipv4 = (_b = object.ipv4) !== null && _b !== void 0 ? _b : "";
        message.ipv6 = ((_c = object.ipv6) === null || _c === void 0 ? void 0 : _c.map((e) => e)) || [];
        message.clientsEthernet = (_d = object.clientsEthernet) !== null && _d !== void 0 ? _d : 0;
        message.clients2ghz = (_e = object.clients2ghz) !== null && _e !== void 0 ? _e : 0;
        message.clients5ghz = (_f = object.clients5ghz) !== null && _f !== void 0 ? _f : 0;
        return message;
    },
};
function longToNumber(int64) {
    const num = globalThis.Number(int64.toString());
    if (num > globalThis.Number.MAX_SAFE_INTEGER) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    if (num < globalThis.Number.MIN_SAFE_INTEGER) {
        throw new globalThis.Error("Value is smaller than Number.MIN_SAFE_INTEGER");
    }
    return num;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
