// @gibme/starlink/dist/protobuf/spacex/api/device/wifi_config.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.11.6
//   protoc               v3.21.12
// source: spacex/api/device/wifi_config.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.WifiConfig = exports.WifiConfig_Security = exports.protobufPackage = void 0;
exports.wifiConfig_SecurityFromJSON = wifiConfig_SecurityFromJSON;
exports.wifiConfig_SecurityToJSON = wifiConfig_SecurityToJSON;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
exports.protobufPackage = "SpaceX.API.Device";
var WifiConfig_Security;
(function (WifiConfig_Security) {
    WifiConfig_Security[WifiConfig_Security["UNKNOWN"] = 0] = "UNKNOWN";
    WifiConfig_Security[WifiConfig_Security["WPA2"] = 1] = "WPA2";
    WifiConfig_Security[WifiConfig_Security["WPA3"] = 2] = "WPA3";
    WifiConfig_Security[WifiConfig_Security["WPA2WPA3"] = 3] = "WPA2WPA3";
    WifiConfig_Security[WifiConfig_Security["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(WifiConfig_Security || (exports.WifiConfig_Security = WifiConfig_Security = {}));
function wifiConfig_SecurityFromJSON(object) {
    switch (object) {
        case 0:
        case "UNKNOWN":
            return WifiConfig_Security.UNKNOWN;
        case 1:
        case "WPA2":
            return WifiConfig_Security.WPA2;
        case 2:
        case "WPA3":
            return WifiConfig_Security.WPA3;
        case 3:
        case "WPA2WPA3":
            return WifiConfig_Security.WPA2WPA3;
        case -1:
        case "UNRECOGNIZED":
        default:
            return WifiConfig_Security.UNRECOGNIZED;
    }
}
function wifiConfig_SecurityToJSON(object) {
    switch (object) {
        case WifiConfig_Security.UNKNOWN:
            return "UNKNOWN";
        case WifiConfig_Security.WPA2:
            return "WPA2";
        case WifiConfig_Security.WPA3:
            return "WPA3";
        case WifiConfig_Security.WPA2WPA3:
            return "WPA2WPA3";
        case WifiConfig_Security.UNRECOGNIZED:
        default:
            return "UNRECOGNIZED";
    }
}
function createBaseWifiConfig() {
    return {
        networkName: undefined,
        applyNetworkName: undefined,
        networkPassword: undefined,
        applyNetworkPassword: undefined,
        countryCode: undefined,
        wifiDisabled: undefined,
        applyWifiDisabled: undefined,
        lanIpv4: undefined,
        lanIpv4SubnetMask: undefined,
        setupComplete: undefined,
        applySetupComplete: undefined,
        factoryResetTicker: undefined,
        version: undefined,
        wifiSecurity: undefined,
        applyWifiSecurity: undefined,
        networkName5ghz: undefined,
        applyNetworkName5ghz: undefined,
        macWan: undefined,
        applyMacWan: undefined,
        macLan: undefined,
        applyMacLan: undefined,
        macLan2ghz: undefined,
        applyMacLan2ghz: undefined,
        macLan5ghz: undefined,
        applyMacLan5ghz: undefined,
        deviceId: undefined,
    };
}
exports.WifiConfig = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.networkName !== undefined) {
            writer.uint32(10).string(message.networkName);
        }
        if (message.applyNetworkName !== undefined) {
            writer.uint32(8008).bool(message.applyNetworkName);
        }
        if (message.networkPassword !== undefined) {
            writer.uint32(18).string(message.networkPassword);
        }
        if (message.applyNetworkPassword !== undefined) {
            writer.uint32(8016).bool(message.applyNetworkPassword);
        }
        if (message.countryCode !== undefined) {
            writer.uint32(26).string(message.countryCode);
        }
        if (message.wifiDisabled !== undefined) {
            writer.uint32(32).bool(message.wifiDisabled);
        }
        if (message.applyWifiDisabled !== undefined) {
            writer.uint32(8024).bool(message.applyWifiDisabled);
        }
        if (message.lanIpv4 !== undefined) {
            writer.uint32(42).string(message.lanIpv4);
        }
        if (message.lanIpv4SubnetMask !== undefined) {
            writer.uint32(50).string(message.lanIpv4SubnetMask);
        }
        if (message.setupComplete !== undefined) {
            writer.uint32(56).bool(message.setupComplete);
        }
        if (message.applySetupComplete !== undefined) {
            writer.uint32(8080).bool(message.applySetupComplete);
        }
        if (message.factoryResetTicker !== undefined) {
            writer.uint32(64).uint32(message.factoryResetTicker);
        }
        if (message.version !== undefined) {
            writer.uint32(72).uint32(message.version);
        }
        if (message.wifiSecurity !== undefined) {
            writer.uint32(80).int32(message.wifiSecurity);
        }
        if (message.applyWifiSecurity !== undefined) {
            writer.uint32(8032).bool(message.applyWifiSecurity);
        }
        if (message.networkName5ghz !== undefined) {
            writer.uint32(90).string(message.networkName5ghz);
        }
        if (message.applyNetworkName5ghz !== undefined) {
            writer.uint32(8040).bool(message.applyNetworkName5ghz);
        }
        if (message.macWan !== undefined) {
            writer.uint32(98).string(message.macWan);
        }
        if (message.applyMacWan !== undefined) {
            writer.uint32(8048).bool(message.applyMacWan);
        }
        if (message.macLan !== undefined) {
            writer.uint32(106).string(message.macLan);
        }
        if (message.applyMacLan !== undefined) {
            writer.uint32(8056).bool(message.applyMacLan);
        }
        if (message.macLan2ghz !== undefined) {
            writer.uint32(114).string(message.macLan2ghz);
        }
        if (message.applyMacLan2ghz !== undefined) {
            writer.uint32(8064).bool(message.applyMacLan2ghz);
        }
        if (message.macLan5ghz !== undefined) {
            writer.uint32(122).string(message.macLan5ghz);
        }
        if (message.applyMacLan5ghz !== undefined) {
            writer.uint32(8072).bool(message.applyMacLan5ghz);
        }
        if (message.deviceId !== undefined) {
            writer.uint32(130).string(message.deviceId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseWifiConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.networkName = reader.string();
                    continue;
                }
                case 1001: {
                    if (tag !== 8008) {
                        break;
                    }
                    message.applyNetworkName = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.networkPassword = reader.string();
                    continue;
                }
                case 1002: {
                    if (tag !== 8016) {
                        break;
                    }
                    message.applyNetworkPassword = reader.bool();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.countryCode = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.wifiDisabled = reader.bool();
                    continue;
                }
                case 1003: {
                    if (tag !== 8024) {
                        break;
                    }
                    message.applyWifiDisabled = reader.bool();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.lanIpv4 = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.lanIpv4SubnetMask = reader.string();
                    continue;
                }
                case 7: {
                    if (tag !== 56) {
                        break;
                    }
                    message.setupComplete = reader.bool();
                    continue;
                }
                case 1010: {
                    if (tag !== 8080) {
                        break;
                    }
                    message.applySetupComplete = reader.bool();
                    continue;
                }
                case 8: {
                    if (tag !== 64) {
                        break;
                    }
                    message.factoryResetTicker = reader.uint32();
                    continue;
                }
                case 9: {
                    if (tag !== 72) {
                        break;
                    }
                    message.version = reader.uint32();
                    continue;
                }
                case 10: {
                    if (tag !== 80) {
                        break;
                    }
                    message.wifiSecurity = reader.int32();
                    continue;
                }
                case 1004: {
                    if (tag !== 8032) {
                        break;
                    }
                    message.applyWifiSecurity = reader.bool();
                    continue;
                }
                case 11: {
                    if (tag !== 90) {
                        break;
                    }
                    message.networkName5ghz = reader.string();
                    continue;
                }
                case 1005: {
                    if (tag !== 8040) {
                        break;
                    }
                    message.applyNetworkName5ghz = reader.bool();
                    continue;
                }
                case 12: {
                    if (tag !== 98) {
                        break;
                    }
                    message.macWan = reader.string();
                    continue;
                }
                case 1006: {
                    if (tag !== 8048) {
                        break;
                    }
                    message.applyMacWan = reader.bool();
                    continue;
                }
                case 13: {
                    if (tag !== 106) {
                        break;
                    }
                    message.macLan = reader.string();
                    continue;
                }
                case 1007: {
                    if (tag !== 8056) {
                        break;
                    }
                    message.applyMacLan = reader.bool();
                    continue;
                }
                case 14: {
                    if (tag !== 114) {
                        break;
                    }
                    message.macLan2ghz = reader.string();
                    continue;
                }
                case 1008: {
                    if (tag !== 8064) {
                        break;
                    }
                    message.applyMacLan2ghz = reader.bool();
                    continue;
                }
                case 15: {
                    if (tag !== 122) {
                        break;
                    }
                    message.macLan5ghz = reader.string();
                    continue;
                }
                case 1009: {
                    if (tag !== 8072) {
                        break;
                    }
                    message.applyMacLan5ghz = reader.bool();
                    continue;
                }
                case 16: {
                    if (tag !== 130) {
                        break;
                    }
                    message.deviceId = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            networkName: isSet(object.networkName)
                ? globalThis.String(object.networkName)
                : isSet(object.network_name)
                    ? globalThis.String(object.network_name)
                    : undefined,
            applyNetworkName: isSet(object.applyNetworkName)
                ? globalThis.Boolean(object.applyNetworkName)
                : isSet(object.apply_network_name)
                    ? globalThis.Boolean(object.apply_network_name)
                    : undefined,
            networkPassword: isSet(object.networkPassword)
                ? globalThis.String(object.networkPassword)
                : isSet(object.network_password)
                    ? globalThis.String(object.network_password)
                    : undefined,
            applyNetworkPassword: isSet(object.applyNetworkPassword)
                ? globalThis.Boolean(object.applyNetworkPassword)
                : isSet(object.apply_network_password)
                    ? globalThis.Boolean(object.apply_network_password)
                    : undefined,
            countryCode: isSet(object.countryCode)
                ? globalThis.String(object.countryCode)
                : isSet(object.country_code)
                    ? globalThis.String(object.country_code)
                    : undefined,
            wifiDisabled: isSet(object.wifiDisabled)
                ? globalThis.Boolean(object.wifiDisabled)
                : isSet(object.wifi_disabled)
                    ? globalThis.Boolean(object.wifi_disabled)
                    : undefined,
            applyWifiDisabled: isSet(object.applyWifiDisabled)
                ? globalThis.Boolean(object.applyWifiDisabled)
                : isSet(object.apply_wifi_disabled)
                    ? globalThis.Boolean(object.apply_wifi_disabled)
                    : undefined,
            lanIpv4: isSet(object.lanIpv4)
                ? globalThis.String(object.lanIpv4)
                : isSet(object.lan_ipv4)
                    ? globalThis.String(object.lan_ipv4)
                    : undefined,
            lanIpv4SubnetMask: isSet(object.lanIpv4SubnetMask)
                ? globalThis.String(object.lanIpv4SubnetMask)
                : isSet(object.lan_ipv4_subnet_mask)
                    ? globalThis.String(object.lan_ipv4_subnet_mask)
                    : undefined,
            setupComplete: isSet(object.setupComplete)
                ? globalThis.Boolean(object.setupComplete)
                : isSet(object.setup_complete)
                    ? globalThis.Boolean(object.setup_complete)
                    : undefined,
            applySetupComplete: isSet(object.applySetupComplete)
                ? globalThis.Boolean(object.applySetupComplete)
                : isSet(object.apply_setup_complete)
                    ? globalThis.Boolean(object.apply_setup_complete)
                    : undefined,
            factoryResetTicker: isSet(object.factoryResetTicker)
                ? globalThis.Number(object.factoryResetTicker)
                : isSet(object.factory_reset_ticker)
                    ? globalThis.Number(object.factory_reset_ticker)
                    : undefined,
            version: isSet(object.version) ? globalThis.Number(object.version) : undefined,
            wifiSecurity: isSet(object.wifiSecurity)
                ? wifiConfig_SecurityFromJSON(object.wifiSecurity)
                : isSet(object.wifi_security)
                    ? wifiConfig_SecurityFromJSON(object.wifi_security)
                    : undefined,
            applyWifiSecurity: isSet(object.applyWifiSecurity)
                ? globalThis.Boolean(object.applyWifiSecurity)
                : isSet(object.apply_wifi_security)
                    ? globalThis.Boolean(object.apply_wifi_security)
                    : undefined,
            networkName5ghz: isSet(object.networkName5ghz)
                ? globalThis.String(object.networkName5ghz)
                : isSet(object.network_name_5ghz)
                    ? globalThis.String(object.network_name_5ghz)
                    : undefined,
            applyNetworkName5ghz: isSet(object.applyNetworkName5ghz)
                ? globalThis.Boolean(object.applyNetworkName5ghz)
                : isSet(object.apply_network_name_5ghz)
                    ? globalThis.Boolean(object.apply_network_name_5ghz)
                    : undefined,
            macWan: isSet(object.macWan)
                ? globalThis.String(object.macWan)
                : isSet(object.mac_wan)
                    ? globalThis.String(object.mac_wan)
                    : undefined,
            applyMacWan: isSet(object.applyMacWan)
                ? globalThis.Boolean(object.applyMacWan)
                : isSet(object.apply_mac_wan)
                    ? globalThis.Boolean(object.apply_mac_wan)
                    : undefined,
            macLan: isSet(object.macLan)
                ? globalThis.String(object.macLan)
                : isSet(object.mac_lan)
                    ? globalThis.String(object.mac_lan)
                    : undefined,
            applyMacLan: isSet(object.applyMacLan)
                ? globalThis.Boolean(object.applyMacLan)
                : isSet(object.apply_mac_lan)
                    ? globalThis.Boolean(object.apply_mac_lan)
                    : undefined,
            macLan2ghz: isSet(object.macLan2ghz)
                ? globalThis.String(object.macLan2ghz)
                : isSet(object.mac_lan_2ghz)
                    ? globalThis.String(object.mac_lan_2ghz)
                    : undefined,
            applyMacLan2ghz: isSet(object.applyMacLan2ghz)
                ? globalThis.Boolean(object.applyMacLan2ghz)
                : isSet(object.apply_mac_lan_2ghz)
                    ? globalThis.Boolean(object.apply_mac_lan_2ghz)
                    : undefined,
            macLan5ghz: isSet(object.macLan5ghz)
                ? globalThis.String(object.macLan5ghz)
                : isSet(object.mac_lan_5ghz)
                    ? globalThis.String(object.mac_lan_5ghz)
                    : undefined,
            applyMacLan5ghz: isSet(object.applyMacLan5ghz)
                ? globalThis.Boolean(object.applyMacLan5ghz)
                : isSet(object.apply_mac_lan_5ghz)
                    ? globalThis.Boolean(object.apply_mac_lan_5ghz)
                    : undefined,
            deviceId: isSet(object.deviceId)
                ? globalThis.String(object.deviceId)
                : isSet(object.device_id)
                    ? globalThis.String(object.device_id)
                    : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.networkName !== undefined) {
            obj.networkName = message.networkName;
        }
        if (message.applyNetworkName !== undefined) {
            obj.applyNetworkName = message.applyNetworkName;
        }
        if (message.networkPassword !== undefined) {
            obj.networkPassword = message.networkPassword;
        }
        if (message.applyNetworkPassword !== undefined) {
            obj.applyNetworkPassword = message.applyNetworkPassword;
        }
        if (message.countryCode !== undefined) {
            obj.countryCode = message.countryCode;
        }
        if (message.wifiDisabled !== undefined) {
            obj.wifiDisabled = message.wifiDisabled;
        }
        if (message.applyWifiDisabled !== undefined) {
            obj.applyWifiDisabled = message.applyWifiDisabled;
        }
        if (message.lanIpv4 !== undefined) {
            obj.lanIpv4 = message.lanIpv4;
        }
        if (message.lanIpv4SubnetMask !== undefined) {
            obj.lanIpv4SubnetMask = message.lanIpv4SubnetMask;
        }
        if (message.setupComplete !== undefined) {
            obj.setupComplete = message.setupComplete;
        }
        if (message.applySetupComplete !== undefined) {
            obj.applySetupComplete = message.applySetupComplete;
        }
        if (message.factoryResetTicker !== undefined) {
            obj.factoryResetTicker = Math.round(message.factoryResetTicker);
        }
        if (message.version !== undefined) {
            obj.version = Math.round(message.version);
        }
        if (message.wifiSecurity !== undefined) {
            obj.wifiSecurity = wifiConfig_SecurityToJSON(message.wifiSecurity);
        }
        if (message.applyWifiSecurity !== undefined) {
            obj.applyWifiSecurity = message.applyWifiSecurity;
        }
        if (message.networkName5ghz !== undefined) {
            obj.networkName5ghz = message.networkName5ghz;
        }
        if (message.applyNetworkName5ghz !== undefined) {
            obj.applyNetworkName5ghz = message.applyNetworkName5ghz;
        }
        if (message.macWan !== undefined) {
            obj.macWan = message.macWan;
        }
        if (message.applyMacWan !== undefined) {
            obj.applyMacWan = message.applyMacWan;
        }
        if (message.macLan !== undefined) {
            obj.macLan = message.macLan;
        }
        if (message.applyMacLan !== undefined) {
            obj.applyMacLan = message.applyMacLan;
        }
        if (message.macLan2ghz !== undefined) {
            obj.macLan2ghz = message.macLan2ghz;
        }
        if (message.applyMacLan2ghz !== undefined) {
            obj.applyMacLan2ghz = message.applyMacLan2ghz;
        }
        if (message.macLan5ghz !== undefined) {
            obj.macLan5ghz = message.macLan5ghz;
        }
        if (message.applyMacLan5ghz !== undefined) {
            obj.applyMacLan5ghz = message.applyMacLan5ghz;
        }
        if (message.deviceId !== undefined) {
            obj.deviceId = message.deviceId;
        }
        return obj;
    },
    create(base) {
        return exports.WifiConfig.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1;
        const message = createBaseWifiConfig();
        message.networkName = (_a = object.networkName) !== null && _a !== void 0 ? _a : undefined;
        message.applyNetworkName = (_b = object.applyNetworkName) !== null && _b !== void 0 ? _b : undefined;
        message.networkPassword = (_c = object.networkPassword) !== null && _c !== void 0 ? _c : undefined;
        message.applyNetworkPassword = (_d = object.applyNetworkPassword) !== null && _d !== void 0 ? _d : undefined;
        message.countryCode = (_e = object.countryCode) !== null && _e !== void 0 ? _e : undefined;
        message.wifiDisabled = (_f = object.wifiDisabled) !== null && _f !== void 0 ? _f : undefined;
        message.applyWifiDisabled = (_g = object.applyWifiDisabled) !== null && _g !== void 0 ? _g : undefined;
        message.lanIpv4 = (_h = object.lanIpv4) !== null && _h !== void 0 ? _h : undefined;
        message.lanIpv4SubnetMask = (_j = object.lanIpv4SubnetMask) !== null && _j !== void 0 ? _j : undefined;
        message.setupComplete = (_k = object.setupComplete) !== null && _k !== void 0 ? _k : undefined;
        message.applySetupComplete = (_l = object.applySetupComplete) !== null && _l !== void 0 ? _l : undefined;
        message.factoryResetTicker = (_m = object.factoryResetTicker) !== null && _m !== void 0 ? _m : undefined;
        message.version = (_o = object.version) !== null && _o !== void 0 ? _o : undefined;
        message.wifiSecurity = (_p = object.wifiSecurity) !== null && _p !== void 0 ? _p : undefined;
        message.applyWifiSecurity = (_q = object.applyWifiSecurity) !== null && _q !== void 0 ? _q : undefined;
        message.networkName5ghz = (_r = object.networkName5ghz) !== null && _r !== void 0 ? _r : undefined;
        message.applyNetworkName5ghz = (_s = object.applyNetworkName5ghz) !== null && _s !== void 0 ? _s : undefined;
        message.macWan = (_t = object.macWan) !== null && _t !== void 0 ? _t : undefined;
        message.applyMacWan = (_u = object.applyMacWan) !== null && _u !== void 0 ? _u : undefined;
        message.macLan = (_v = object.macLan) !== null && _v !== void 0 ? _v : undefined;
        message.applyMacLan = (_w = object.applyMacLan) !== null && _w !== void 0 ? _w : undefined;
        message.macLan2ghz = (_x = object.macLan2ghz) !== null && _x !== void 0 ? _x : undefined;
        message.applyMacLan2ghz = (_y = object.applyMacLan2ghz) !== null && _y !== void 0 ? _y : undefined;
        message.macLan5ghz = (_z = object.macLan5ghz) !== null && _z !== void 0 ? _z : undefined;
        message.applyMacLan5ghz = (_0 = object.applyMacLan5ghz) !== null && _0 !== void 0 ? _0 : undefined;
        message.deviceId = (_1 = object.deviceId) !== null && _1 !== void 0 ? _1 : undefined;
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
