// @gibme/sql/dist/database.js
"use strict";
// Copyright (c) 2016-2025, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Database = void 0;
const events_1 = require("events");
const pg_format_1 = __importDefault(require("pg-format"));
const sqlstring_1 = require("sqlstring");
class Database extends events_1.EventEmitter {
    type;
    tableOptions;
    constructor(type, tableOptions = '') {
        super();
        this.type = type;
        this.tableOptions = tableOptions;
    }
    get typeName() {
        switch (this.type) {
            case Database.Type.MYSQL:
                return 'MySQL';
            case Database.Type.POSTGRES:
                return 'Postgres';
            case Database.Type.SQLITE:
                return 'SQLite';
            case Database.Type.MARIADB:
                return 'MariaDB';
        }
    }
    /**
     * Drop the tables from the database
     *
     * @param tables
     */
    async dropTable(tables) {
        if (!Array.isArray(tables)) {
            tables = [tables];
        }
        const queries = [];
        for (const table of tables) {
            queries.push({
                query: `DROP TABLE IF EXISTS ${this.escapeId(table)}`
            });
        }
        return this.transaction(queries);
    }
    /**
     * Escapes the value for SQL
     *
     * @param value
     */
    escape(value) {
        if (this.type === Database.Type.POSTGRES) {
            return (0, pg_format_1.default)('%L', value);
        }
        return (0, sqlstring_1.escape)(value);
    }
    /**
     * Escapes the ID for sql
     *
     * @param id
     */
    escapeId(id) {
        if (this.type === Database.Type.POSTGRES) {
            return (0, pg_format_1.default)('%I', id);
        }
        return (0, sqlstring_1.escapeId)(id);
    }
    /**
     * Prepares a query to perform a multi-insert statement which is far
     * faster than a bunch of individual insert statements
     *
     * @param table
     * @param columns
     * @param values
     */
    prepareMultiInsert(table, columns = [], values) {
        return this._prepareMultiInsert(this.type, table, columns, values);
    }
    /**
     * Prepares a query to perform a multi-update statement which is
     * based upon a multi-insert statement that performs an UPSERT
     * and this is a lot faster than a bunch of individual
     * update statements
     *
     * @param table
     * @param primaryKey
     * @param columns
     * @param values
     */
    prepareMultiUpdate(table, primaryKey, columns, values) {
        return this._prepareMultiUpdate(this.type, table, primaryKey, columns, values);
    }
    /**
     * Prepares the creation of an index on the specified table
     *
     * @param table
     * @param fields
     * @param type
     */
    prepareCreateIndex(table, fields, type = Database.Table.IndexType.NONE) {
        return [this._prepareCreateIndex(this.type, table, fields, type)];
    }
    /**
     * Prepares the creation of a table including the relevant indexes and constraints
     *
     * @param name
     * @param fields
     * @param primaryKey
     * @param tableOptions
     */
    prepareCreateTable(name, fields, primaryKey = [], tableOptions = this.tableOptions) {
        return this._prepareCreateTable(this.type, name, fields, primaryKey, tableOptions);
    }
    /**
     * Prepares and executes the creation of a table including the relevant indexes and
     * constraints
     *
     * @param name
     * @param fields
     * @param primaryKey
     * @param tableOptions
     * @param useTransaction
     */
    async createTable(name, fields, primaryKey = [], tableOptions = this.tableOptions, useTransaction = true) {
        const queries = this.prepareCreateTable(name, fields, primaryKey, tableOptions);
        if (useTransaction) {
            await this.transaction(queries);
        }
        else {
            for (const query of queries) {
                await this.query(query);
            }
        }
    }
    /**
     * Prepares and executes the creation of an index on the table specified
     *
     * @param table
     * @param fields
     * @param type
     * @param useTransaction
     */
    async createIndex(table, fields, type = Database.Table.IndexType.NONE, useTransaction = true) {
        const queries = this.prepareCreateIndex(table, fields, type);
        if (useTransaction) {
            await this.transaction(queries);
        }
        else {
            for (const query of queries) {
                await this.query(query);
            }
        }
    }
    /**
     * Prepares and performs a query that performs a multi-insert statement
     * which is far faster than a bunch of individual insert statements
     *
     * @param table
     * @param columns
     * @param values
     * @param useTransaction
     */
    async multiInsert(table, columns = [], values, useTransaction = true) {
        const queries = this.prepareMultiInsert(table, columns, values);
        return this._executeMulti(queries, useTransaction);
    }
    /**
     * Prepares and executes a query to that performs  a multi-update statement
     * which is based upon a multi-insert statement that performs an UPSERT
     * which is a lot faster than a bunch of update statements
     *
     * @param table
     * @param primaryKey
     * @param columns
     * @param values
     * @param useTransaction
     */
    async multiUpdate(table, primaryKey, columns, values, useTransaction = true) {
        const queries = this.prepareMultiUpdate(table, primaryKey, columns, values);
        return this._executeMulti(queries, useTransaction);
    }
    /**
     * Truncates the specified table(s)
     *
     * @param table
     * @param useTransaction
     */
    async truncate(table, useTransaction = true) {
        if (!Array.isArray(table)) {
            table = [table];
        }
        const queries = [];
        switch (this.type) {
            case Database.Type.SQLITE:
                table.forEach(table => queries.push({ query: `DELETE FROM ${this.escapeId(table)}` }));
                break;
            default:
                table.forEach(table => queries.push({ query: `TRUNCATE TABLE ${this.escapeId(table)}` }));
                break;
        }
        try {
            if (useTransaction) {
                await this.transaction(queries);
            }
            else {
                for (const query of queries) {
                    await this.query(query);
                }
            }
            return true;
        }
        catch {
            return false;
        }
    }
    /**
     * Prepares any foreign key constraints based upon the field supplied
     *
     * @param table
     * @param fields
     * @protected
     */
    _prepareConstraints(table, fields) {
        table = table.trim();
        const constraints = [];
        for (const field of fields) {
            if (field.foreignKey) {
                field.name = field.name.trim();
                field.foreignKey.table = field.foreignKey.table.trim();
                field.foreignKey.column = field.foreignKey.column.trim();
                const key_name = [
                    'fk',
                    table,
                    field.name,
                    field.foreignKey.table,
                    field.foreignKey.column
                ].join('_');
                let constraint = `CONSTRAINT ${key_name} ` +
                    `FOREIGN KEY (${this.escapeId(field.name)}) ` +
                    `REFERENCES ${this.escapeId(field.foreignKey.table)} ` +
                    `(${this.escapeId(field.foreignKey.column)})`;
                if (field.foreignKey.onDelete) {
                    constraint += ` ON DELETE ${field.foreignKey.onDelete}`;
                }
                if (field.foreignKey.onUpdate) {
                    constraint += ` ON UPDATE ${field.foreignKey.onUpdate}`;
                }
                constraints.push(constraint.trim());
            }
        }
        return constraints;
    }
    /**
     * Prepares the creation of an index on a table
     *
     * @param databaseType
     * @param table
     * @param fields
     * @param type
     * @protected
     */
    _prepareCreateIndex(databaseType, table, fields, type = Database.Table.IndexType.NONE) {
        table = table.trim();
        fields = fields.map(field => field.trim());
        const can_if_not_exists = databaseType !== Database.Type.MYSQL;
        const if_not_exists = can_if_not_exists ? ' IF NOT EXISTS' : '';
        return {
            query: `CREATE ${type} INDEX${if_not_exists} ${type.toLowerCase()}_${table}_${fields.join('_')} ` +
                `ON ${this.escapeId(table)} (${fields.map(elem => this.escapeId(elem)).join(',')})`,
            noError: !can_if_not_exists
        };
    }
    /**
     * Prepares statements to create indexes during a CREATE table statement
     *
     * @param databaseType
     * @param table
     * @param fields
     * @param type
     * @protected
     */
    _prepareCreateIndexes(databaseType, table, fields, type = Database.Table.IndexType.NONE) {
        table = table.trim();
        return fields.filter(column => column.unique === true)
            .map(column => this._prepareCreateIndex(databaseType, table, [column.name], type));
    }
    /**
     * Prepares the creation of a table including the relevant indexes and constraints
     *
     * @param databaseType
     * @param table
     * @param columns
     * @param primaryKey
     * @param tableOptions
     */
    _prepareCreateTable(databaseType, table, columns, primaryKey = [], tableOptions = '') {
        table = table.trim();
        primaryKey = primaryKey.map(column => this.escapeId(column));
        const [fields, values] = this._prepareColumns(columns);
        const constraints = this._prepareConstraints(table, columns);
        const primary_key = primaryKey.length !== 0
            ? `, PRIMARY KEY (${primaryKey.join(', ')})`
            : '';
        const query = [
            'CREATE TABLE IF NOT EXISTS',
            this.escapeId(table),
            '(',
            fields.join(', '),
            primary_key,
            constraints.length !== 0 ? `, ${constraints.join(', ')}` : '',
            ')',
            tableOptions
        ].filter(elem => elem.length !== 0)
            .map(elem => elem.trim())
            .join(' ')
            .trim();
        const unique = this._prepareCreateIndexes(databaseType, table, columns, Database.Table.IndexType.UNIQUE);
        return [
            {
                query,
                values
            },
            ...unique
        ];
    }
    /**
     * Prepares the columns for a CREATE statement
     *
     * @param fields
     * @protected
     */
    _prepareColumns(fields) {
        const values = [];
        const columns = [];
        for (const column of fields) {
            if (typeof column.default !== 'undefined') {
                values.push(column.default);
            }
            columns.push(`${this.escapeId(column.name)} ${column.type.toUpperCase()} ` +
                `${!column.nullable ? 'NOT NULL' : 'NULL'} ` +
                `${typeof column.default !== 'undefined' ? 'DEFAULT ?' : ''}`
                    .trim());
        }
        return [columns, values];
    }
    /**
     * Prepares a query to perform a multi-insert statement which is far faster than
     * a bunch of individual insert statements
     *
     * @param databaseType
     * @param table
     * @param columns
     * @param values
     */
    _prepareMultiInsert(databaseType, table, columns = [], values) {
        const toPlaceholders = (arr) => {
            return arr.map(() => '?')
                .join(',');
        };
        if (values.length === 0) {
            throw new Error('Must supply values');
        }
        if (columns.length !== 0) {
            for (const _values of values) {
                if (_values.length !== columns.length) {
                    throw new Error('Column count does not match values count');
                }
            }
        }
        const _columns = columns.length !== 0 ? ` (${columns.map(elem => this.escapeId(elem)).join(',')})` : '';
        // SQLite handles things a bit differently
        if (databaseType === Database.Type.SQLITE) {
            const queries = [];
            for (const _values of values) {
                queries.push({
                    query: `INSERT INTO ${this.escapeId(table)}${_columns} VALUES (${toPlaceholders(_values)})`,
                    values: _values
                });
            }
            return queries;
        }
        const placeholders = [];
        const parameters = [];
        const placeholder = columns.length !== 0 ? `(${toPlaceholders(columns)})` : `(${toPlaceholders(values[0])})`;
        for (const _values of values) {
            placeholders.push(placeholder);
            parameters.push(..._values);
        }
        return [{
                query: `INSERT INTO ${this.escapeId(table)}${_columns} VALUES ${placeholders.join(',')}`.trim(),
                values: parameters
            }];
    }
    /**
     * Prepares a query to perform a multi-update statement which is based upon a multi-insert statement
     * that performs an UPSERT and this is a lot faster than a bunch of individual
     * update statements
     *
     * @param databaseType
     * @param table
     * @param primaryKey
     * @param columns
     * @param values
     */
    _prepareMultiUpdate(databaseType, table, primaryKey, columns, values) {
        if (columns.length === 0) {
            throw new Error('Must specify columns for multi-update');
        }
        if (primaryKey.length === 0) {
            throw new Error('Must specify primary key column(s) for multi-update');
        }
        const queries = this._prepareMultiInsert(databaseType, table, columns, values);
        const updates = [];
        if (databaseType === Database.Type.MYSQL || databaseType === Database.Type.MARIADB) {
            for (const column of columns) {
                if (primaryKey.includes(column)) {
                    continue;
                }
                updates.push(`${this.escapeId(column)} = VALUES(${this.escapeId(column)})`);
            }
            return queries.map(query => {
                query.query += ` ON DUPLICATE KEY UPDATE ${updates.join(',')}`;
                return query;
            });
        }
        for (const column of columns) {
            if (primaryKey.includes(column)) {
                continue;
            }
            updates.push(`${this.escapeId(column)} = excluded.${this.escapeId(column)}`);
        }
        return queries.map(query => {
            query.query += ` ON CONFLICT (${primaryKey.map(elem => this.escapeId(elem)).join(',')}) ` +
                `DO UPDATE SET ${updates.join(',')}`;
            return query;
        });
    }
    /**
     * Executes multiple statements and returns their results in a single Database.Query.Result
     *
     * @param queries
     * @param useTransaction
     * @protected
     */
    async _executeMulti(queries, useTransaction) {
        if (useTransaction) {
            const results = await this.transaction(queries);
            return [
                [],
                {
                    affectedRows: results.map(result => result[1].affectedRows)
                        .reduce((previous, current) => previous + current),
                    changedRows: results.map(result => result[1].changedRows)
                        .reduce((previous, current) => previous + current),
                    length: results.map(result => result[1].length)
                        .reduce((previous, current) => previous + current)
                },
                {
                    query: queries.map(query => query.query).join(';')
                }
            ];
        }
        else {
            let affectedRows = 0;
            let changedRows = 0;
            let length = 0;
            for (const query of queries) {
                const [, meta] = await this.query(query);
                affectedRows += meta.affectedRows;
                changedRows += meta.changedRows;
                length += meta.length;
            }
            return [
                [],
                {
                    affectedRows,
                    changedRows,
                    length
                }, {
                    query: queries.map(query => query.query).join(';')
                }
            ];
        }
    }
    /** @ignore */
    make_error(error) {
        if (error instanceof Error) {
            return error;
        }
        return new Error(error.toString());
    }
    ;
}
exports.Database = Database;
(function (Database) {
    let Type;
    (function (Type) {
        Type[Type["MYSQL"] = 0] = "MYSQL";
        Type[Type["POSTGRES"] = 1] = "POSTGRES";
        Type[Type["SQLITE"] = 2] = "SQLITE";
        Type[Type["MARIADB"] = 4] = "MARIADB";
    })(Type = Database.Type || (Database.Type = {}));
    let Table;
    (function (Table) {
        let IndexType;
        (function (IndexType) {
            IndexType["NONE"] = "";
            IndexType["UNIQUE"] = "UNIQUE";
        })(IndexType = Table.IndexType || (Table.IndexType = {}));
        let ForeignKeyConstraint;
        (function (ForeignKeyConstraint) {
            ForeignKeyConstraint["RESTRICT"] = "RESTRICT";
            ForeignKeyConstraint["CASCADE"] = "CASCADE";
            ForeignKeyConstraint["NULL"] = "SET NULL";
            ForeignKeyConstraint["DEFAULT"] = "SET DEFAULT";
            ForeignKeyConstraint["NA"] = "NO ACTION";
        })(ForeignKeyConstraint = Table.ForeignKeyConstraint || (Table.ForeignKeyConstraint = {}));
    })(Table = Database.Table || (Database.Table = {}));
})(Database || (exports.Database = Database = {}));
exports.default = Database;
