// @gibme/sql/dist/index.js
"use strict";
// Copyright (c) 2016-2025, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Database = exports.SQLite = exports.MariaDB = exports.MySQL = exports.Postgres = exports.createConnection = void 0;
const postgres_1 = __importDefault(require("./postgres"));
exports.Postgres = postgres_1.default;
const mysql_1 = __importDefault(require("./mysql"));
exports.MySQL = mysql_1.default;
const sqlite_1 = __importDefault(require("./sqlite"));
exports.SQLite = sqlite_1.default;
const database_1 = __importDefault(require("./database"));
exports.Database = database_1.default;
const mariadb_1 = __importDefault(require("./mariadb"));
exports.MariaDB = mariadb_1.default;
const dotenv_1 = require("dotenv");
(0, dotenv_1.config)();
const createConnection = (database_type = parseInt(process.env.SQL_TYPE || '2') || database_1.default.Type.SQLITE, options = {}) => {
    switch (database_type) {
        case database_1.default.Type.SQLITE:
            return new sqlite_1.default({
                filename: process.env.SQL_FILENAME,
                ...options
            });
        case database_1.default.Type.MYSQL:
            return new mysql_1.default({
                host: process.env.SQL_HOST,
                port: parseInt(process.env.SQL_PORT || '') || undefined,
                user: process.env.SQL_USERNAME,
                password: process.env.SQL_PASSWORD,
                database: process.env.SQL_DATABASE,
                rejectUnauthorized: false,
                useSSL: process.env.SQL_SSL === 'true',
                ...options
            });
        case database_1.default.Type.MARIADB:
            return new mariadb_1.default({
                host: process.env.SQL_HOST,
                port: parseInt(process.env.SQL_PORT || '') || undefined,
                user: process.env.SQL_USERNAME,
                password: process.env.SQL_PASSWORD,
                database: process.env.SQL_DATABASE,
                rejectUnauthorized: false,
                useSSL: process.env.SQL_SSL === 'true',
                ...options
            });
        case database_1.default.Type.POSTGRES:
            return new postgres_1.default({
                host: process.env.SQL_HOST,
                port: parseInt(process.env.SQL_PORT || '') || undefined,
                user: process.env.SQL_USERNAME,
                password: process.env.SQL_PASSWORD,
                database: process.env.SQL_DATABASE,
                ssl: process.env.SQL_SSL === 'true',
                rejectUnauthorized: false,
                ...options
            });
        default:
            throw new Error('Invalid database type specified');
    }
};
exports.createConnection = createConnection;
exports.default = {
    Postgres: postgres_1.default,
    MySQL: mysql_1.default,
    SQLite: sqlite_1.default,
    MariaDB: mariadb_1.default,
    Database: database_1.default,
    createConnection: exports.createConnection
};
