// @gibme/sql/dist/mysql.js
"use strict";
// Copyright (c) 2016-2025, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MySQL = exports.Database = void 0;
const mariadb_1 = require("mariadb");
const database_1 = __importDefault(require("./database"));
exports.Database = database_1.default;
/** @ignore */
const isOkPacket = (x) => {
    return !!x && typeof x === 'object' &&
        'affectedRows' in x &&
        'insertId' in x &&
        'warningStatus' in x;
};
class MySQL extends database_1.default {
    config;
    pool;
    /**
     * Creates a new instance of the class
     *
     * @param config
     * @param tableOptions
     * @param override_type
     */
    constructor(config = {}, tableOptions = 'ENGINE=InnoDB PACK_KEYS=1 ROW_FORMAT=COMPRESSED', override_type = database_1.default.Type.MYSQL) {
        super(override_type, tableOptions);
        this.config = config;
        this.config.host ??= '127.0.0.1';
        this.config.port ??= 3306;
        this.config.user ??= '';
        this.config.connectTimeout ??= 30_000;
        this.config.useSSL ??= false;
        this.config.rejectUnauthorized ??= false;
        this.config.allowPublicKeyRetrieval ??= true;
        this.config.multipleStatements = false;
        if (this.config.useSSL) {
            this.config.ssl ||= {
                rejectUnauthorized: this.config.rejectUnauthorized
            };
        }
        this.pool = (0, mariadb_1.createPool)(this.config);
        this.pool.on('acquire', connection => this.emit('acquire', connection));
        this.pool.on('connection', connection => this.emit('connection', connection));
        this.pool.on('enqueue', () => this.emit('enqueue'));
        this.pool.on('release', connection => this.emit('release', connection));
    }
    static get type() {
        return database_1.default.Type.MYSQL;
    }
    on(event, listener) {
        return super.on(event, listener);
    }
    /**
     * Gets the total number of idle connections in the pool
     */
    get idleConnections() {
        return this.pool.idleConnections();
    }
    /**
     * Gets the total number of connections in the pool
     */
    get totalConnections() {
        return this.pool.totalConnections();
    }
    /**
     * Closes all the pooled connections
     */
    async close() {
        return this.pool.end();
    }
    /**
     * Switches the default database referenced by the connection
     *
     * @param database
     */
    async use(database) {
        await this.query(`USE ${this.escapeId(database)}`);
        this.config.database = database;
        return this;
    }
    /**
     * Lists the tables in the specified database
     *
     * @param database
     */
    async listTables(database = this.config.database) {
        if (!database) {
            throw new Error('No database specified');
        }
        const [rows] = await this.query(`SHOW TABLES FROM ${this.escapeId(database)}`);
        return rows.map(row => {
            const key = Object.keys(row)[0];
            return row[key];
        });
    }
    /**
     * Performs an individual query and returns the result
     *
     * @param query
     * @param values
     */
    async query(query, ...values) {
        return this._query(query, values);
    }
    /**
     * Performs the specified queries in a transaction
     *
     * @param queries
     */
    async transaction(queries) {
        const connection = await this.connection();
        const results = [];
        try {
            await this.beginTransaction(connection);
        }
        catch (error) {
            await connection.release();
            throw this.make_error(error);
        }
        for (const query of queries) {
            try {
                const result = await this._query(query.query, query.values, connection);
                results.push(result);
            }
            catch (error) {
                if (!query.noError) {
                    await this.rollbackTransaction(connection);
                    await connection.release();
                    throw this.make_error(error);
                }
            }
        }
        try {
            await this.commitTransaction(connection);
        }
        catch (error) {
            throw this.make_error(error);
        }
        finally {
            await connection.release();
        }
        return results;
    }
    /**
     * Returns a database connection from the pool
     *
     * @protected
     */
    async connection() {
        return this.pool.getConnection();
    }
    /**
     * Starts a transaction on the specified connection
     *
     * @param connection
     * @protected
     */
    async beginTransaction(connection) {
        return connection.beginTransaction();
    }
    /**
     * Commits a transaction on the specified connection
     *
     * @param connection
     * @protected
     */
    async commitTransaction(connection) {
        return connection.commit();
    }
    /**
     * Rolls back a transaction on the specified connection
     *
     * @param connection
     * @protected
     */
    async rollbackTransaction(connection) {
        return connection.rollback();
    }
    /**
     * Performs an individual query and returns the result
     *
     * @param query
     * @param values
     * @param connection
     */
    async _query(query, values = [], connection = this.pool) {
        if (typeof query === 'object') {
            if (query.values) {
                values = query.values;
            }
            query = query.query;
        }
        const result = await connection.query(query, values);
        if (isOkPacket(result)) {
            return [[], {
                    changedRows: result.changedRows ?? 0,
                    affectedRows: result.affectedRows ?? 0,
                    insertId: result.insertId ?? 0,
                    length: 0
                }, { query, values }];
        }
        const results = result;
        return [results, {
                changedRows: 0,
                affectedRows: 0,
                insertId: undefined,
                length: results.length
            }, { query, values }];
    }
}
exports.MySQL = MySQL;
exports.default = MySQL;
