// @gibme/sql/dist/postgres.js
"use strict";
// Copyright (c) 2016-2025, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Postgres = exports.Database = void 0;
const pg_1 = require("pg");
const database_1 = __importDefault(require("./database"));
exports.Database = database_1.default;
class Postgres extends database_1.default {
    config;
    pool;
    /**
     * Creates a new instance of the class
     *
     * @param config
     */
    constructor(config = {}) {
        super(database_1.default.Type.POSTGRES);
        this.config = config;
        this.config.host ??= '127.0.0.1';
        this.config.port ??= 5432;
        this.config.user ??= '';
        this.config.rejectUnauthorized ??= false;
        if (this.config.ssl === true) {
            this.config.ssl = {
                rejectUnauthorized: this.config.rejectUnauthorized
            };
        }
        else if (this.config.ssl === false) {
            this.config.ssl = undefined;
        }
        else if (typeof this.config.ssl === 'object') {
            this.config.ssl = {
                ...this.config.ssl,
                rejectUnauthorized: this.config.rejectUnauthorized
            };
        }
        this.pool = new pg_1.Pool(this.config);
        this.pool.on('connect', client => this.emit('connect', client));
        this.pool.on('acquire', client => this.emit('acquire', client));
        this.pool.on('remove', client => this.emit('remove', client));
        this.pool.on('error', (error, client) => this.emit('error', error, client));
    }
    static get type() {
        return database_1.default.Type.POSTGRES;
    }
    on(event, listener) {
        return super.on(event, listener);
    }
    /**
     * Gets the total number of idle connections in the pool
     */
    get idleConnections() {
        return this.pool.idleCount;
    }
    /**
     * Gets the total number of connections in the pool
     */
    get totalConnections() {
        return this.pool.totalCount;
    }
    /**
     * Closes all pooled connections
     */
    async close() {
        return this.pool.end();
    }
    /**
     * Creates a new instance connected to the specified database using
     * the same configuration options
     *
     * @param database
     * @param user
     * @param password
     */
    async use(database, user, password) {
        return new Postgres({
            ...this.config,
            database,
            user: user || this.config.user,
            password: password || this.config.password
        });
    }
    /**
     * Lists the tables in the current schema
     */
    async listTables(database = this.config.database) {
        if (!database) {
            throw new Error('No database specified');
        }
        const instance = database !== this.config.database ? await this.use(database) : this;
        const [rows] = await instance.query('SELECT tablename FROM pg_catalog.pg_tables ' +
            'WHERE schemaname != \'pg_catalog\' AND schemaname != \'information_schema\' ' +
            'ORDER BY tablename');
        return rows.map(row => row.tablename);
    }
    /**
     * Performs an individual query and returns the results
     *
     * @param query
     * @param values
     */
    async query(query, ...values) {
        return this._query(query, values);
    }
    /**
     * Performs the specified queries in a transaction
     *
     * @param queries
     */
    async transaction(queries) {
        const connection = await this.connection();
        const results = [];
        try {
            await this.beginTransaction(connection);
        }
        catch (error) {
            await this.release(connection);
            throw this.make_error(error);
        }
        for (const query of queries) {
            try {
                const result = await this._query(query.query, query.values, connection);
                results.push(result);
            }
            catch (error) {
                if (!query.noError) {
                    await this.rollbackTransaction(connection);
                    await this.release(connection);
                    throw this.make_error(error);
                }
            }
        }
        try {
            await this.commitTransaction(connection);
        }
        catch (error) {
            throw this.make_error(error);
        }
        finally {
            await this.release(connection);
        }
        return results;
    }
    /**
     * Returns a database connection from the pool
     *
     * @protected
     */
    async connection() {
        return this.pool.connect();
    }
    /**
     * Starts a transaction on the specified connection
     *
     * @param connection
     * @protected
     */
    async beginTransaction(connection) {
        await connection.query('BEGIN');
    }
    /**
     * Commits a transaction on the specified connection
     *
     * @param connection
     * @protected
     */
    async commitTransaction(connection) {
        await connection.query('COMMIT');
    }
    /**
     * Rolls back a transaction on the specified connection
     *
     * @param connection
     * @protected
     */
    async rollbackTransaction(connection) {
        await connection.query('ROLLBACK');
    }
    /**
     * Transforms a query from using ?-based placeholders to pgsql placeholders
     *
     * @param query
     * @protected
     */
    transformQuery(query) {
        let counter = 1;
        while (query.includes('?')) {
            query = query.replace(/\?/, `$${counter++}`);
        }
        return query;
    }
    /**
     * Assists with releasing connections once we are done with them
     *
     * @param connection
     * @protected
     */
    async release(connection) {
        try {
            connection.release(false);
        }
        catch (error) {
            if (!error.toString()
                .toLowerCase()
                .includes('already been released')) {
                throw this.make_error(error);
            }
        }
    }
    /**
     * Performs an individual query and returns the results
     *
     * @param query
     * @param values
     * @param connection
     */
    async _query(query, values = [], connection = this.pool) {
        if (typeof query === 'object') {
            if (query.values) {
                values = query.values;
            }
            query = query.query;
        }
        query = this.transformQuery(query);
        const result = await connection.query(query, values);
        return [result.rows, {
                changedRows: result.rows.length === 0 ? result.rowCount || 0 : 0,
                affectedRows: result.rows.length === 0 ? result.rowCount || 0 : 0,
                length: result.rows.length
            }, {
                query,
                values
            }];
    }
}
exports.Postgres = Postgres;
exports.default = Postgres;
