// @gibme/sql/dist/sqlite_instance.js
"use strict";
// Copyright (c) 2016-2025, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLiteInstance = void 0;
const better_sqlite3_1 = __importDefault(require("better-sqlite3"));
const async_mutex_1 = require("async-mutex");
/** @ignore */
const pragmaFunctionCalls = [
    'quick_check',
    'integrity_check',
    'incremental_vacuum',
    'foreign_key_check',
    'foreign_key_list',
    'index_info',
    'index_list',
    'index_xinfo',
    'table_info',
    'table_xinfo',
    'optimize'
];
class SQLiteInstance {
    database;
    mutex = new async_mutex_1.Mutex();
    constructor(database) {
        this.database = database;
    }
    /**
     * Gets the total number of idle connections in the pool
     */
    get idleConnections() {
        return this.mutex.isLocked() ? 0 : 1;
    }
    /**
     * Gets the total number of connections in the pool
     */
    get totalConnections() {
        return 1;
    }
    /**
     * Returns if the database is currently open
     */
    get open() {
        return this.database.open;
    }
    /**
     * Returns if the database is currently in a transaction
     */
    get inTransaction() {
        return this.database.inTransaction;
    }
    /**
     * Returns the filename of the database
     */
    get name() {
        return this.database.name;
    }
    /**
     * Returns whether this database is in-memory only
     */
    get memory() {
        return this.database.memory;
    }
    /**
     * Returns whether this database is in readonly mode
     */
    get readonly() {
        return this.database.readonly;
    }
    /**
     * Loads a new instance of an SQLite database or loads from memory if already opened
     *
     * @param filename
     * @param readonly
     */
    static async load(filename, readonly = false) {
        const db = await SQLiteInstance.open(filename, readonly);
        return new SQLiteInstance(db);
    }
    /**
     * Opens an SQLite database
     *
     * @param filename
     * @param readonly
     * @private
     */
    static async open(filename, readonly = false) {
        return new better_sqlite3_1.default(filename, { readonly });
    }
    /**
     * Closes the database
     */
    async close() {
        this.database.close();
    }
    /**
     * Retrieves the current PRAGMA setting
     * @param option
     */
    async getPragma(option) {
        option = option.toLowerCase();
        const [rows] = await this.allAsync({
            query: `PRAGMA ${option}`
        });
        if (rows.length === 1) {
            return rows[0][option];
        }
        else {
            if (rows[0][option]) {
                return rows.map(elem => elem[option]);
            }
        }
        return rows;
    }
    /**
     * Sets the given PRAGMA setting
     * @param option
     * @param value
     */
    async setPragma(option, value) {
        option = option.toLowerCase();
        value = pragmaFunctionCalls.includes(option) ? `(${value})` : ` = ${value}`;
        /**
         * Execute this call via the low-level calling system outside the normal
         * queuing provided so that we do not mistakenly block the connection
         */
        await this.runAsync({ query: `PRAGMA ${option}${value}` });
    }
    /**
     * Performs an individual query and returns the result
     *
     * @param query
     * @param values
     */
    async query(query, ...values) {
        if (typeof query === 'object') {
            if (query.values) {
                values = query.values;
            }
            query = query.query;
        }
        if (query.toLowerCase().startsWith('select')) {
            return this.allAsync({
                query,
                values
            });
        }
        else {
            return this.runAsync({
                query,
                values
            });
        }
    }
    /**
     * Performs the specified queries in a transaction
     *
     * @param queries
     */
    async transaction(queries) {
        return this.transactionAsync(queries);
    }
    /**
     * Executes a low-level-all call against the SQLite database connection
     * that is wrapped in a mutex to prevent multiple queries from being
     * executed at the same time
     *
     * @param query
     * @private
     */
    async allAsync(query) {
        return this.mutex.runExclusive(() => {
            return this.allSync(query);
        });
    }
    /**
     * Executes a low-level-all call against the SQLite database connection
     *
     * @param query
     * @private
     */
    allSync(query) {
        const stmt = this.database.prepare(query.query);
        const rows = stmt.all(...query.values ?? []);
        return [
            rows,
            {
                changedRows: 0,
                affectedRows: 0,
                length: rows.length
            },
            query
        ];
    }
    /**
     * Executes a low-level run call against the SQLite database connection
     * that is wrapped in a mutex to prevent multiple queries from being
     * executed at the same time
     *
     * @param query
     * @private
     */
    async runAsync(query) {
        return this.mutex.runExclusive(() => {
            return this.runSync(query);
        });
    }
    /**
     * Executes a low-level run call against the SQLite database connection
     *
     * @param query
     * @private
     */
    runSync(query) {
        const stmt = this.database.prepare(query.query);
        const info = stmt.run(...query.values ?? []);
        const insertId = parseInt(info.lastInsertRowid.toString() ?? '') || 0;
        return [
            [],
            {
                changedRows: info.changes ?? 0,
                affectedRows: info.changes ?? 0,
                insertId,
                length: 0
            },
            query
        ];
    }
    /**
     * Executes a low-level transaction call against the SQLite database connection
     *
     * @param queries
     * @protected
     */
    async transactionAsync(queries) {
        return this.mutex.runExclusive(() => {
            const results = [];
            const tx = this.database.transaction(() => {
                for (const query of queries) {
                    if (query.query.toLowerCase().startsWith('select')) {
                        results.push(this.allSync(query));
                    }
                    else {
                        results.push(this.runSync(query));
                    }
                }
            });
            try {
                tx();
            }
            catch (error) {
                if (queries.some(q => !q.noError)) {
                    if (error instanceof Error) {
                        throw error;
                    }
                    throw new Error(error.toString());
                }
            }
            return results;
        });
    }
}
exports.SQLiteInstance = SQLiteInstance;
