// @gibme/sql/dist/sqlite.js
"use strict";
// Copyright (c) 2016-2025, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLite = exports.Database = void 0;
const path_1 = require("path");
const sqlite_instance_manager_1 = require("./sqlite_instance_manager");
const database_1 = __importDefault(require("./database"));
exports.Database = database_1.default;
class SQLite extends database_1.default {
    config;
    database;
    /**
     * Creates a new instance of the class
     *
     * @param config
     */
    constructor(config = {}) {
        super(database_1.default.Type.SQLITE);
        this.config = config;
        this.config.filename ??= ':memory:';
        this.config.readonly ??= false;
        this.config.foreignKeys ??= true;
        this.config.WALmode ??= true;
        if (this.config.filename.toLowerCase() !== ':memory:') {
            this.config.filename = (0, path_1.resolve)(this.config.filename);
        }
    }
    static get type() {
        return database_1.default.Type.SQLITE;
    }
    /**
     * Gets the total number of idle connections in the pool
     */
    get idleConnections() {
        return this.database?.idleConnections ?? 0;
    }
    /**
     * Gets the total number of connections in the pool
     */
    get totalConnections() {
        return this.database?.totalConnections ?? 0;
    }
    /**
     * Closes the database instance
     *
     * Note: closing the db using this method will kill the instance
     * for all references to the instance
     */
    async close() {
        const instance = await this.getInstance();
        return instance.close();
    }
    /**
     * Creates a new instance connected to the specified database using
     * the same configuration options
     *
     * @param filename
     */
    async use(filename) {
        return new SQLite({
            ...this.config,
            filename
        });
    }
    /**
     * Retrieves the current PRAGMA setting
     * @param option
     */
    async getPragma(option) {
        const instance = await this.getInstance();
        return instance.getPragma(option);
    }
    /**
     * Sets the given PRAGMA setting
     * @param option
     * @param value
     */
    async setPragma(option, value) {
        const instance = await this.getInstance();
        return instance.setPragma(option, value);
    }
    /**
     * Lists the tables in the specified database
     *
     * @param filename
     */
    async listTables(filename) {
        const instance = filename && filename !== this.config.filename ? await this.use(filename) : this;
        const [rows] = await instance.query('SELECT name FROM sqlite_schema WHERE type = \'table\' AND name NOT LIKE \'sqlite_%\'');
        return rows.map(row => row.name);
    }
    /**
     * Performs an individual query and returns the result
     *
     * @param query
     * @param values
     */
    async query(query, ...values) {
        const instance = await this.getInstance();
        return instance.query(query, ...values);
    }
    /**
     * Performs the specified queries in a transaction
     *
     * @param queries
     */
    async transaction(queries) {
        const instance = await this.getInstance();
        return instance.transaction(queries);
    }
    /**
     * This does nothing as the underlying instance handles this internally
     *
     * @param _connection
     * @protected
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    beginTransaction(_connection) {
        return Promise.resolve();
    }
    /**
     * This does nothing as the underlying instance handles this internally
     *
     * @param _connection
     * @protected
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    commitTransaction(_connection) {
        return Promise.resolve();
    }
    /**
     * This does nothing as the underlying instance handles this internally
     *
     * @param _connection
     * @protected
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    rollbackTransaction(_connection) {
        return Promise.resolve();
    }
    /**
     * This is not required for this database type
     *
     * @protected
     */
    connection() {
        return Promise.resolve();
    }
    /**
     * Retrieves the SQLite database instance singleton
     *
     * @private
     */
    async getInstance() {
        if (this.database) {
            return this.database;
        }
        this.database = await sqlite_instance_manager_1.SQLiteInstanceManager.get(this.config.filename ?? ':memory:', this.config.readonly);
        if (this.config.WALmode) {
            await this.database.setPragma('journal_mode', 'WAL');
        }
        if (this.config.foreignKeys) {
            await this.database.setPragma('foreign_keys', true);
        }
        return this.database;
    }
}
exports.SQLite = SQLite;
exports.default = SQLite;
