// @gibme/cache/dist/index.js
"use strict";
// Copyright (c) 2018-2025, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Database = exports.Redis = exports.Memory = exports.Cache = exports.SQLite = exports.Postgres = exports.MariaDB = exports.MySQL = void 0;
const memory_1 = __importDefault(require("./memory"));
exports.Memory = memory_1.default;
const redis_1 = __importDefault(require("./redis"));
exports.Redis = redis_1.default;
const database_1 = __importDefault(require("./database"));
exports.Database = database_1.default;
var database_2 = require("./database");
Object.defineProperty(exports, "MySQL", { enumerable: true, get: function () { return database_2.MySQL; } });
Object.defineProperty(exports, "MariaDB", { enumerable: true, get: function () { return database_2.MariaDB; } });
Object.defineProperty(exports, "Postgres", { enumerable: true, get: function () { return database_2.Postgres; } });
Object.defineProperty(exports, "SQLite", { enumerable: true, get: function () { return database_2.SQLite; } });
var cache_1 = require("./cache");
Object.defineProperty(exports, "Cache", { enumerable: true, get: function () { return cache_1.Cache; } });
exports.default = {
    Memory: memory_1.default,
    Redis: redis_1.default,
    Database: database_1.default
};
