// @gibme/cache/dist/memory.js
"use strict";
// Copyright (c) 2018-2025, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Memory = exports.Cache = void 0;
const node_cache_1 = __importDefault(require("node-cache"));
const cache_1 = __importDefault(require("./cache"));
exports.Cache = cache_1.default;
class Memory extends cache_1.default {
    options;
    client;
    ready = false;
    /**
     * Constructs a new instance of a memory-based cache
     *
     * @param options
     */
    constructor(options = {}) {
        super();
        this.options = options;
        this.options.stdTTL ??= 300;
        this.options.checkperiod ??= 30;
        this.client = new node_cache_1.default(this.options);
        this.client.on('set', (key, value) => this.emit('set', this.unstringify(key), this.unstringify(value)));
        this.client.on('expired', (key, value) => this.emit('expired', this.unstringify(key), this.unstringify(value)));
        this.client.on('del', (key, value) => this.emit('del', this.unstringify(key), this.unstringify(value)));
        this.client.on('flush', () => this.emit('flush'));
        this.client.on('flush_stats', () => this.emit('flush_stats'));
        this.client.on('error', error => this.emit('error', error));
        this.ready = true;
    }
    on(event, listener) {
        return super.on(event, listener);
    }
    /**
     * Returns in the instance is ready
     */
    get isReady() {
        return this.ready;
    }
    static get type() {
        return 'Memory';
    }
    /**
     * Connects to the memory storage
     */
    async connect() {
        return undefined;
    }
    /**
     * Closes the storage connection and halts the check interval
     */
    async disconnect() {
        this.ready = false;
        return this.client.close();
    }
    /**
     * Sets the value for the specified key
     *
     * @param key
     * @param value
     * @param ttl
     */
    async set(key, value, ttl = this.options.stdTTL ?? this.defaultTTL) {
        const _key = this.stringify(key);
        return this.client.set(_key, this.stringify(value), ttl);
    }
    /**
     * Retrieves the value of the specified key if it exists
     *
     * @param key
     */
    async get(key) {
        const _key = this.stringify(key);
        const response = this.client.get(_key);
        if (response) {
            return this.unstringify(response);
        }
    }
    /**
     * Checks if the specified key exists in the cache
     *
     * @param key
     */
    async includes(key) {
        const _key = this.stringify(key);
        return this.client.has(_key);
    }
    /**
     * Deletes the key (including the value) from the cache
     *
     * @param key
     */
    async del(key) {
        const _key = this.stringify(key);
        return this.client.del(_key);
    }
    /**
     * Deletes all keys (and values) from the cache
     */
    async clear() {
        this.client.flushAll();
        return true;
    }
    /**
     * Retrieves a list of all keys
     */
    async keys() {
        return this.client.keys()
            .map(key => this.unstringify(key));
    }
    /**
     * Mass retrieves values with the specified keys
     *
     * @param keys
     */
    async mget(keys) {
        const fetch_keys = keys.map(key => this.stringify(key));
        const data = this.client.mget(fetch_keys);
        const values = new Map();
        for (const key of Object.keys(data)) {
            if (data[key]) {
                values.set(this.unstringify(key), this.unstringify(data[key]));
            }
        }
        return values;
    }
    /**
     * Mass deletes values with the specified keys
     *
     * @param keys
     */
    async mdel(keys) {
        const _keys = keys.map(key => this.stringify(key));
        return this.client.del(_keys);
    }
    /**
     * Mass sets values with the specified keys
     *
     * @param keys
     * @param values
     * @param ttl
     */
    async mset(keys, values, ttl = this.options.stdTTL ?? this.defaultTTL) {
        if (keys.length !== values.length) {
            throw new Error('Keys and values array lengths must match!');
        }
        const _keys = keys.map(key => this.stringify(key));
        const _values = values.map(value => this.stringify(value));
        const set = [];
        for (let i = 0; i < _keys.length; i++) {
            set.push({
                key: _keys[i],
                val: _values[i],
                ttl
            });
        }
        return this.client.mset(set);
    }
    /**
     * Retrieves a map of all the keys and values in the cache
     */
    async list() {
        const keys = await this.keys();
        return this.mget(keys);
    }
    /**
     * Updates the TTL for the specified key
     *
     * @param key
     * @param ttl
     */
    async ttl(key, ttl = this.options.stdTTL ?? this.defaultTTL) {
        return this.client.ttl(this.stringify(key), ttl);
    }
    /**
     * Retrieves the current TTL of the specified key
     *
     * @param key
     */
    async getTtl(key) {
        return this.client.getTtl(this.stringify(key));
    }
    /**
     * Retrieves the value of the specified key and immediately deletes it from the cache
     *
     * @param key
     */
    async take(key) {
        const value = this.client.take(this.stringify(key));
        if (value) {
            return this.unstringify(value);
        }
    }
}
exports.Memory = Memory;
exports.default = Memory;
