// @gibme/cache/dist/redis.js
"use strict";
// Copyright (c) 2018-2025, Brandon Lehmann <brandonlehmann@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Redis = exports.Cache = void 0;
const redis_1 = require("redis");
const cache_1 = __importDefault(require("./cache"));
exports.Cache = cache_1.default;
const dotenv_1 = require("dotenv");
(0, dotenv_1.config)();
class Redis extends cache_1.default {
    options;
    client = (0, redis_1.createClient)();
    /**
     * Constructs a new instance of a redis-based cache
     *
     * @param options
     */
    constructor(options = {}) {
        super();
        this.options = options;
        this.options.host ??= process.env.REDIS_HOST ?? 'localhost';
        this.options.port ??= parseInt(process.env.REDIS_PORT || '6379') ?? 6379;
        this.options.username ??= process.env.REDIS_USERNAME;
        this.options.password ??= process.env.REDIS_PASSWORD;
        this.options.stdTTL ??= 300;
        this.options.url = 'redis://';
        if (this.options.username) {
            this.options.url += this.options.username;
        }
        if (this.options.password) {
            if (this.options.username) {
                this.options.url += ':';
            }
            this.options.url += this.options.password;
        }
        if (this.options.username || this.options.password) {
            this.options.url += '@';
        }
        delete this.options.username;
        delete this.options.password;
        this.options.url += `${this.options.host}:${this.options.port}`;
        this.client = (0, redis_1.createClient)(this.options);
        this.client.on('connect', () => this.emit('connect'));
        this.client.on('ready', () => this.emit('ready'));
        this.client.on('end', () => this.emit('end'));
        this.client.on('error', error => this.emit('error', error));
        this.client.on('reconnecting', () => this.emit('reconnecting'));
    }
    on(event, listener) {
        return super.on(event, listener);
    }
    /**
     * Returns if the instance is ready
     */
    get isReady() {
        return this.client.isReady;
    }
    static get type() {
        return 'Redis';
    }
    /**
     * Connects to the redis server
     */
    async connect() {
        await this.client.connect();
    }
    /**
     * Disconnects from the redis server
     */
    async disconnect() {
        return this.client.disconnect();
    }
    /**
     * Sets the value for the specified key
     *
     * @param key
     * @param value
     * @param ttl
     */
    async set(key, value, ttl = this.options.stdTTL ?? this.defaultTTL) {
        if (!this.client.isReady) {
            await this.connect();
        }
        const _key = this.stringify(key);
        const response = await this.client.set(_key, this.stringify(value), {
            EX: ttl
        });
        return response?.toLowerCase() === 'ok';
    }
    /**
     * Retrieves the value of the specified key if it exists
     *
     * @param key
     */
    async get(key) {
        if (!this.client.isReady) {
            await this.connect();
        }
        const _key = this.stringify(key);
        const response = await this.client.get(_key);
        if (response) {
            return this.unstringify(response);
        }
    }
    /**
     * Checks if the specified key exists in the cache
     *
     * @param key
     */
    async includes(key) {
        if (!this.client.isReady) {
            await this.connect();
        }
        const _key = this.stringify(key);
        return (await this.client.exists(_key)) === 1;
    }
    /**
     * Deletes the key (including the value) from the cache
     *
     * @param key
     */
    async del(key) {
        if (!this.client.isReady) {
            await this.connect();
        }
        const _key = this.stringify(key);
        return this.client.del(_key);
    }
    /**
     * Deletes all keys (and values) from the cache
     */
    async clear() {
        if (!this.client.isReady) {
            await this.connect();
        }
        return (await this.client.flushAll(redis_1.REDIS_FLUSH_MODES.SYNC)).toLowerCase() === 'ok';
    }
    /**
     * Retrieves a list of all keys
     */
    async keys() {
        if (!this.client.isReady) {
            await this.connect();
        }
        return (await this.client.keys('*'))
            .map(key => this.unstringify(key));
    }
    /**
     * Mass retrieves values with the specified keys
     *
     * @param keys
     */
    async mget(keys) {
        if (!this.client.isReady) {
            await this.connect();
        }
        const fetch_keys = keys.map(key => this.stringify(key));
        const data = await this.client.mGet(fetch_keys);
        const values = new Map();
        for (let i = 0; i < keys.length; i++) {
            if (data[i]) {
                values.set(keys[i], this.unstringify(data[i] ?? ''));
            }
        }
        return values;
    }
    /**
     * Mass deletes values with the specified keys
     *
     * @param keys
     */
    async mdel(keys) {
        if (!this.client.isReady) {
            await this.connect();
        }
        const _keys = keys.map(key => this.stringify(key));
        return this.client.del(_keys);
    }
    /**
     * Mass sets values with the specified keys
     *
     * @param keys
     * @param values
     * @param ttl
     */
    async mset(keys, values, ttl = this.options.stdTTL ?? this.defaultTTL) {
        if (!this.client.isReady) {
            await this.connect();
        }
        if (keys.length !== values.length) {
            throw new Error('Keys and values array lengths must match!');
        }
        const _keys = keys.map(key => this.stringify(key));
        const _values = values.map(value => this.stringify(value));
        const set = [];
        for (let i = 0; i < _keys.length; i++) {
            set.push([_keys[i], _values[i]]);
        }
        if ((await this.client.mSet(set)).toLowerCase() !== 'ok') {
            return false;
        }
        const promises = [];
        for (const key of keys) {
            promises.push(this.ttl(key, ttl));
        }
        return (await Promise.all(promises))
            .filter(result => result)
            .length === keys.length;
    }
    /**
     * Retrieves a map of all the keys and values in the cache
     */
    async list() {
        if (!this.client.isReady) {
            await this.connect();
        }
        const keys = await this.keys();
        return this.mget(keys);
    }
    /**
     * Updates the TTL for the specified key
     *
     * @param key
     * @param ttl
     */
    async ttl(key, ttl = this.options.stdTTL ?? this.defaultTTL) {
        if (!this.client.isReady) {
            await this.connect();
        }
        const _key = this.stringify(key);
        try {
            await this.client.expire(_key, ttl);
            return true;
        }
        catch {
            return false;
        }
    }
    /**
     * Retrieves the current TTL of the specified key
     *
     * @param key
     */
    async getTtl(key) {
        if (!this.client.isReady) {
            await this.connect();
        }
        const result = await this.client.ttl(this.stringify(key));
        if (result >= 0) {
            return result;
        }
    }
    /**
     * Retrieves the value of the specified key and immediately deletes it from the cache
     *
     * @param key
     */
    async take(key) {
        if (!this.client.isReady) {
            await this.connect();
        }
        const value = await this.get(key);
        if (value) {
            await this.del(key);
            return value;
        }
    }
}
exports.Redis = Redis;
exports.default = Redis;
